{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "4f172cdc-4994-41c5-8955-b7a3105482bc",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "fe00c1e4-71c3-464f-a743-8ca26f71256a",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "# all_muts_df = pd.read_pickle(\"./all_aledb_MG1655_ep_max_freq_mut_df.pkl\")\n",
    "# display(all_muts_df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "46b0e145",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # only care about public mutations for this.\n",
    "# all_muts_df = all_muts_df[all_muts_df['ALEdb source'] == 'public']\n",
    "# len(all_muts_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "0537ba2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_muts_df = pd.read_pickle(\"./pub_aledb_MG1655_ep_max_freq_mut_df.pkl\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "0652e98e-dfaa-490e-ad85-dc977d170367",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "# Renaming the \"Gene\" column to \"mutation target annotation\" because this is more accurate.\n",
    "all_muts_df = all_muts_df.rename(index=str, columns={\"Gene\": \"mutation target annotation\"})"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fe0fda0d-6813-4778-9111-697b684ddc83",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "# Build genetic feature annotation\n",
    "Need to build genetic feature annotations before genomic features annotations since the genetic annotations are used to fill gaps with unknown genomic targets."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "001ef62f-13d1-4731-81ab-c1a1951e97e9",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENE_ID</th>\n",
       "      <th>GENE_NAME</th>\n",
       "      <th>GENE_POSLEFT</th>\n",
       "      <th>GENE_POSRIGHT</th>\n",
       "      <th>GENE_STRAND</th>\n",
       "      <th>GENE_SEQUENCE</th>\n",
       "      <th>GC_CONTENT</th>\n",
       "      <th>CRI_SCORE</th>\n",
       "      <th>GENE_NOTE</th>\n",
       "      <th>GENE_INTERNAL_COMMENT</th>\n",
       "      <th>KEY_ID_ORG</th>\n",
       "      <th>GENE_TYPE</th>\n",
       "      <th>range</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ECK120000001</td>\n",
       "      <td>alr</td>\n",
       "      <td>4265782.0</td>\n",
       "      <td>4266861.0</td>\n",
       "      <td>forward</td>\n",
       "      <td>ATGCAAGCGGCAACTGTTGTGATTAACCGCCGCGCTCTGCGACACA...</td>\n",
       "      <td>55.93</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(4265782, 4266861)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ECK120000002</td>\n",
       "      <td>modB</td>\n",
       "      <td>795862.0</td>\n",
       "      <td>796551.0</td>\n",
       "      <td>forward</td>\n",
       "      <td>ATGATACTGACCGATCCAGAATGGCAGGCAGTTTTATTAAGCCTGA...</td>\n",
       "      <td>54.06</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(795862, 796551)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ECK120000003</td>\n",
       "      <td>cysZ</td>\n",
       "      <td>2531463.0</td>\n",
       "      <td>2532224.0</td>\n",
       "      <td>forward</td>\n",
       "      <td>ATGGTTTCATCATTCACATCTGCCCCACGCAGCGGTTTTTACTATT...</td>\n",
       "      <td>50.13</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(2531463, 2532224)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ECK120000004</td>\n",
       "      <td>dfp</td>\n",
       "      <td>3812731.0</td>\n",
       "      <td>3813951.0</td>\n",
       "      <td>forward</td>\n",
       "      <td>ATGAGCCTGGCCGGTAAAAAAATCGTTCTCGGCGTTAGCGGCGGTA...</td>\n",
       "      <td>53.64</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(3812731, 3813951)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ECK120000005</td>\n",
       "      <td>dcuB</td>\n",
       "      <td>4347404.0</td>\n",
       "      <td>4348744.0</td>\n",
       "      <td>reverse</td>\n",
       "      <td>ATGTTATTTACTATCCAACTTATCATAATACTGATATGTCTGTTTT...</td>\n",
       "      <td>52.27</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ECK12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>(4347404, 4348744)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        GENE_ID GENE_NAME  GENE_POSLEFT  GENE_POSRIGHT GENE_STRAND  \\\n",
       "0  ECK120000001       alr     4265782.0      4266861.0     forward   \n",
       "1  ECK120000002      modB      795862.0       796551.0     forward   \n",
       "2  ECK120000003      cysZ     2531463.0      2532224.0     forward   \n",
       "3  ECK120000004       dfp     3812731.0      3813951.0     forward   \n",
       "4  ECK120000005      dcuB     4347404.0      4348744.0     reverse   \n",
       "\n",
       "                                       GENE_SEQUENCE  GC_CONTENT  CRI_SCORE  \\\n",
       "0  ATGCAAGCGGCAACTGTTGTGATTAACCGCCGCGCTCTGCGACACA...       55.93        NaN   \n",
       "1  ATGATACTGACCGATCCAGAATGGCAGGCAGTTTTATTAAGCCTGA...       54.06        NaN   \n",
       "2  ATGGTTTCATCATTCACATCTGCCCCACGCAGCGGTTTTTACTATT...       50.13        NaN   \n",
       "3  ATGAGCCTGGCCGGTAAAAAAATCGTTCTCGGCGTTAGCGGCGGTA...       53.64        NaN   \n",
       "4  ATGTTATTTACTATCCAACTTATCATAATACTGATATGTCTGTTTT...       52.27        NaN   \n",
       "\n",
       "  GENE_NOTE  GENE_INTERNAL_COMMENT KEY_ID_ORG GENE_TYPE               range  \n",
       "0       NaN                    NaN      ECK12       NaN  (4265782, 4266861)  \n",
       "1       NaN                    NaN      ECK12       NaN    (795862, 796551)  \n",
       "2       NaN                    NaN      ECK12       NaN  (2531463, 2532224)  \n",
       "3       NaN                    NaN      ECK12       NaN  (3812731, 3813951)  \n",
       "4       NaN                    NaN      ECK12       NaN  (4347404, 4348744)  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_df = pd.read_pickle(\"./data/gene_df.pkl\")\n",
    "gene_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "d247b4b2-247f-4575-9e51-d99745b57dc9",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exp</th>\n",
       "      <th>ale</th>\n",
       "      <th>flask</th>\n",
       "      <th>isolate</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>presence</th>\n",
       "      <th>Position</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>Details</th>\n",
       "      <th>mutation target annotation</th>\n",
       "      <th>Reference Seq</th>\n",
       "      <th>ALEdb source</th>\n",
       "      <th>sample</th>\n",
       "      <th>coding</th>\n",
       "      <th>exp ALE</th>\n",
       "      <th>range</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>23836</th>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>702352</td>\n",
       "      <td>DEL</td>\n",
       "      <td>Δ21 bp</td>\n",
       "      <td>coding (380‑400/1149 nt)</td>\n",
       "      <td>nagA</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>public</td>\n",
       "      <td>1 124 1 1</td>\n",
       "      <td>True</td>\n",
       "      <td>42C 1</td>\n",
       "      <td>(702352, 702372)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>183980</th>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>702352</td>\n",
       "      <td>DEL</td>\n",
       "      <td>Δ21 bp</td>\n",
       "      <td>coding (380‑400/1149 nt)</td>\n",
       "      <td>nagA</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>public</td>\n",
       "      <td>1 124 1 1</td>\n",
       "      <td>True</td>\n",
       "      <td>42C 1</td>\n",
       "      <td>(702352, 702372)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23837</th>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1308318</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G→C</td>\n",
       "      <td>R110G (CGT→GGT)</td>\n",
       "      <td>clsA</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>public</td>\n",
       "      <td>1 124 1 1</td>\n",
       "      <td>True</td>\n",
       "      <td>42C 1</td>\n",
       "      <td>(1308318, 1308318)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>183981</th>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1308318</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G→C</td>\n",
       "      <td>R110G (CGT→GGT)</td>\n",
       "      <td>clsA</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>public</td>\n",
       "      <td>1 124 1 1</td>\n",
       "      <td>True</td>\n",
       "      <td>42C 1</td>\n",
       "      <td>(1308318, 1308318)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>183979</th>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2173364</td>\n",
       "      <td>DEL</td>\n",
       "      <td>Δ1 bp</td>\n",
       "      <td>intergenic (‑2/+1)</td>\n",
       "      <td>gatC, gatC</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>public</td>\n",
       "      <td>1 124 1 1</td>\n",
       "      <td>False</td>\n",
       "      <td>42C 1</td>\n",
       "      <td>(2173364, 2173364)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        exp  ale  flask  isolate  tech_rep  presence  Position Mutation Type  \\\n",
       "23836   42C    1    124        1         1       1.0    702352           DEL   \n",
       "183980  42C    1    124        1         1       1.0    702352           DEL   \n",
       "23837   42C    1    124        1         1       1.0   1308318           SNP   \n",
       "183981  42C    1    124        1         1       1.0   1308318           SNP   \n",
       "183979  42C    1    124        1         1       1.0   2173364           DEL   \n",
       "\n",
       "       Sequence Change                   Details mutation target annotation  \\\n",
       "23836           Δ21 bp  coding (380‑400/1149 nt)                       nagA   \n",
       "183980          Δ21 bp  coding (380‑400/1149 nt)                       nagA   \n",
       "23837              G→C           R110G (CGT→GGT)                       clsA   \n",
       "183981             G→C          R110G (CGT→GGT)                        clsA   \n",
       "183979           Δ1 bp        intergenic (‑2/+1)                 gatC, gatC   \n",
       "\n",
       "       Reference Seq ALEdb source     sample  coding exp ALE  \\\n",
       "23836      NC_000913       public  1 124 1 1    True   42C 1   \n",
       "183980     NC_000913       public  1 124 1 1    True   42C 1   \n",
       "23837      NC_000913       public  1 124 1 1    True   42C 1   \n",
       "183981     NC_000913       public  1 124 1 1    True   42C 1   \n",
       "183979     NC_000913       public  1 124 1 1   False   42C 1   \n",
       "\n",
       "                     range  \n",
       "23836     (702352, 702372)  \n",
       "183980    (702352, 702372)  \n",
       "23837   (1308318, 1308318)  \n",
       "183981  (1308318, 1308318)  \n",
       "183979  (2173364, 2173364)  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os, sys\n",
    "module_path = os.path.abspath(os.path.join('..'))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "from aledbmutil.mut import get_original_nuc_mut_range\n",
    "\n",
    "all_muts_df[\"range\"] = all_muts_df.apply(get_original_nuc_mut_range, axis=1)\n",
    "all_muts_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "ef2b55a5-d48d-45f9-9d9c-c7f229e1c49e",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "import pandarallel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "2f0649e9-6aef-41fe-85e9-6ae28e21f769",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO: Pandarallel will run on 4 workers.\n",
      "INFO: Pandarallel will use Memory file system to transfer data between the main process and workers.\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "1325d060a5ca4dd0895bfbcc89f6c87f",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "VBox(children=(HBox(children=(IntProgress(value=0, description='0.00%', max=1918), Label(value='0 / 1918'))), …"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exp</th>\n",
       "      <th>ale</th>\n",
       "      <th>flask</th>\n",
       "      <th>isolate</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>presence</th>\n",
       "      <th>Position</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>Details</th>\n",
       "      <th>mutation target annotation</th>\n",
       "      <th>Reference Seq</th>\n",
       "      <th>ALEdb source</th>\n",
       "      <th>sample</th>\n",
       "      <th>coding</th>\n",
       "      <th>exp ALE</th>\n",
       "      <th>range</th>\n",
       "      <th>gene RegulonDB ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>23836</th>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>702352</td>\n",
       "      <td>DEL</td>\n",
       "      <td>Δ21 bp</td>\n",
       "      <td>coding (380‑400/1149 nt)</td>\n",
       "      <td>nagA</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>public</td>\n",
       "      <td>1 124 1 1</td>\n",
       "      <td>True</td>\n",
       "      <td>42C 1</td>\n",
       "      <td>(702352, 702372)</td>\n",
       "      <td>{ECK120000625}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>183980</th>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>702352</td>\n",
       "      <td>DEL</td>\n",
       "      <td>Δ21 bp</td>\n",
       "      <td>coding (380‑400/1149 nt)</td>\n",
       "      <td>nagA</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>public</td>\n",
       "      <td>1 124 1 1</td>\n",
       "      <td>True</td>\n",
       "      <td>42C 1</td>\n",
       "      <td>(702352, 702372)</td>\n",
       "      <td>{ECK120000625}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23837</th>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1308318</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G→C</td>\n",
       "      <td>R110G (CGT→GGT)</td>\n",
       "      <td>clsA</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>public</td>\n",
       "      <td>1 124 1 1</td>\n",
       "      <td>True</td>\n",
       "      <td>42C 1</td>\n",
       "      <td>(1308318, 1308318)</td>\n",
       "      <td>{ECK120001556}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>183981</th>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1308318</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G→C</td>\n",
       "      <td>R110G (CGT→GGT)</td>\n",
       "      <td>clsA</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>public</td>\n",
       "      <td>1 124 1 1</td>\n",
       "      <td>True</td>\n",
       "      <td>42C 1</td>\n",
       "      <td>(1308318, 1308318)</td>\n",
       "      <td>{ECK120001556}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>183979</th>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2173364</td>\n",
       "      <td>DEL</td>\n",
       "      <td>Δ1 bp</td>\n",
       "      <td>intergenic (‑2/+1)</td>\n",
       "      <td>gatC, gatC</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>public</td>\n",
       "      <td>1 124 1 1</td>\n",
       "      <td>False</td>\n",
       "      <td>42C 1</td>\n",
       "      <td>(2173364, 2173364)</td>\n",
       "      <td>{ECK120002276}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        exp  ale  flask  isolate  tech_rep  presence  Position Mutation Type  \\\n",
       "23836   42C    1    124        1         1       1.0    702352           DEL   \n",
       "183980  42C    1    124        1         1       1.0    702352           DEL   \n",
       "23837   42C    1    124        1         1       1.0   1308318           SNP   \n",
       "183981  42C    1    124        1         1       1.0   1308318           SNP   \n",
       "183979  42C    1    124        1         1       1.0   2173364           DEL   \n",
       "\n",
       "       Sequence Change                   Details mutation target annotation  \\\n",
       "23836           Δ21 bp  coding (380‑400/1149 nt)                       nagA   \n",
       "183980          Δ21 bp  coding (380‑400/1149 nt)                       nagA   \n",
       "23837              G→C           R110G (CGT→GGT)                       clsA   \n",
       "183981             G→C          R110G (CGT→GGT)                        clsA   \n",
       "183979           Δ1 bp        intergenic (‑2/+1)                 gatC, gatC   \n",
       "\n",
       "       Reference Seq ALEdb source     sample  coding exp ALE  \\\n",
       "23836      NC_000913       public  1 124 1 1    True   42C 1   \n",
       "183980     NC_000913       public  1 124 1 1    True   42C 1   \n",
       "23837      NC_000913       public  1 124 1 1    True   42C 1   \n",
       "183981     NC_000913       public  1 124 1 1    True   42C 1   \n",
       "183979     NC_000913       public  1 124 1 1   False   42C 1   \n",
       "\n",
       "                     range gene RegulonDB ID  \n",
       "23836     (702352, 702372)    {ECK120000625}  \n",
       "183980    (702352, 702372)    {ECK120000625}  \n",
       "23837   (1308318, 1308318)    {ECK120001556}  \n",
       "183981  (1308318, 1308318)    {ECK120001556}  \n",
       "183979  (2173364, 2173364)    {ECK120002276}  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from pandarallel import pandarallel\n",
    "pandarallel.initialize(progress_bar=True)\n",
    "\n",
    "\n",
    "from aledbmutil.genome import NON_K12_EXP_L, get_feature_hit_set\n",
    "\n",
    "def _f(r):  # parallel_apply doesn't do lambda functions, therefore making temporary function\n",
    "    if r.exp not in NON_K12_EXP_L:\n",
    "        return get_feature_hit_set(r[\"range\"], gene_df, \"range\", \"GENE_ID\")\n",
    "    else:\n",
    "        return set()\n",
    "\n",
    "all_muts_df[\"gene RegulonDB ID\"] = all_muts_df.parallel_apply(_f, axis=1)\n",
    "all_muts_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "cfcae3af-664b-4f47-8bd5-3d4797eba328",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>RegulonDB ID</th>\n",
       "      <th>range</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>thrL/thrA</td>\n",
       "      <td>ECK120001251/ECK120000987</td>\n",
       "      <td>(256, 336)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>thrA/thrB</td>\n",
       "      <td>ECK120000987/ECK120000988</td>\n",
       "      <td>(2800, 2800)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>thrC/yaaX</td>\n",
       "      <td>ECK120000989/ECK120002701</td>\n",
       "      <td>(5021, 5233)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>yaaX/yaaA</td>\n",
       "      <td>ECK120002701/ECK120000009</td>\n",
       "      <td>(5531, 5682)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>yaaA/yaaJ</td>\n",
       "      <td>ECK120000009/ECK120001508</td>\n",
       "      <td>(6460, 6528)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        name               RegulonDB ID         range\n",
       "0  thrL/thrA  ECK120001251/ECK120000987    (256, 336)\n",
       "1  thrA/thrB  ECK120000987/ECK120000988  (2800, 2800)\n",
       "2  thrC/yaaX  ECK120000989/ECK120002701  (5021, 5233)\n",
       "3  yaaX/yaaA  ECK120002701/ECK120000009  (5531, 5682)\n",
       "4  yaaA/yaaJ  ECK120000009/ECK120001508  (6460, 6528)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "intergenic_df = pd.read_pickle(\"./data/intergenic_region_df.pkl\")\n",
    "intergenic_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "a74c90dc-283a-4073-9647-72119d7b3d19",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "d3d0b0b9a1704b8c812e25be8c3827b3",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "VBox(children=(HBox(children=(IntProgress(value=0, description='0.00%', max=1918), Label(value='0 / 1918'))), …"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exp</th>\n",
       "      <th>ale</th>\n",
       "      <th>flask</th>\n",
       "      <th>isolate</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>presence</th>\n",
       "      <th>Position</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>Details</th>\n",
       "      <th>mutation target annotation</th>\n",
       "      <th>Reference Seq</th>\n",
       "      <th>ALEdb source</th>\n",
       "      <th>sample</th>\n",
       "      <th>coding</th>\n",
       "      <th>exp ALE</th>\n",
       "      <th>range</th>\n",
       "      <th>gene RegulonDB ID</th>\n",
       "      <th>genetic features</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>23836</th>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>702352</td>\n",
       "      <td>DEL</td>\n",
       "      <td>Δ21 bp</td>\n",
       "      <td>coding (380‑400/1149 nt)</td>\n",
       "      <td>nagA</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>public</td>\n",
       "      <td>1 124 1 1</td>\n",
       "      <td>True</td>\n",
       "      <td>42C 1</td>\n",
       "      <td>(702352, 702372)</td>\n",
       "      <td>{ECK120000625}</td>\n",
       "      <td>[{'name': 'nagA', 'RegulonDB ID': 'ECK12000062...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>183980</th>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>702352</td>\n",
       "      <td>DEL</td>\n",
       "      <td>Δ21 bp</td>\n",
       "      <td>coding (380‑400/1149 nt)</td>\n",
       "      <td>nagA</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>public</td>\n",
       "      <td>1 124 1 1</td>\n",
       "      <td>True</td>\n",
       "      <td>42C 1</td>\n",
       "      <td>(702352, 702372)</td>\n",
       "      <td>{ECK120000625}</td>\n",
       "      <td>[{'name': 'nagA', 'RegulonDB ID': 'ECK12000062...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23837</th>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1308318</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G→C</td>\n",
       "      <td>R110G (CGT→GGT)</td>\n",
       "      <td>clsA</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>public</td>\n",
       "      <td>1 124 1 1</td>\n",
       "      <td>True</td>\n",
       "      <td>42C 1</td>\n",
       "      <td>(1308318, 1308318)</td>\n",
       "      <td>{ECK120001556}</td>\n",
       "      <td>[{'name': 'clsA', 'RegulonDB ID': 'ECK12000155...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>183981</th>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1308318</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G→C</td>\n",
       "      <td>R110G (CGT→GGT)</td>\n",
       "      <td>clsA</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>public</td>\n",
       "      <td>1 124 1 1</td>\n",
       "      <td>True</td>\n",
       "      <td>42C 1</td>\n",
       "      <td>(1308318, 1308318)</td>\n",
       "      <td>{ECK120001556}</td>\n",
       "      <td>[{'name': 'clsA', 'RegulonDB ID': 'ECK12000155...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>183979</th>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2173364</td>\n",
       "      <td>DEL</td>\n",
       "      <td>Δ1 bp</td>\n",
       "      <td>intergenic (‑2/+1)</td>\n",
       "      <td>gatC, gatC</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>public</td>\n",
       "      <td>1 124 1 1</td>\n",
       "      <td>False</td>\n",
       "      <td>42C 1</td>\n",
       "      <td>(2173364, 2173364)</td>\n",
       "      <td>{ECK120002276}</td>\n",
       "      <td>[{'name': 'gatC', 'RegulonDB ID': 'ECK12000227...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        exp  ale  flask  isolate  tech_rep  presence  Position Mutation Type  \\\n",
       "23836   42C    1    124        1         1       1.0    702352           DEL   \n",
       "183980  42C    1    124        1         1       1.0    702352           DEL   \n",
       "23837   42C    1    124        1         1       1.0   1308318           SNP   \n",
       "183981  42C    1    124        1         1       1.0   1308318           SNP   \n",
       "183979  42C    1    124        1         1       1.0   2173364           DEL   \n",
       "\n",
       "       Sequence Change                   Details mutation target annotation  \\\n",
       "23836           Δ21 bp  coding (380‑400/1149 nt)                       nagA   \n",
       "183980          Δ21 bp  coding (380‑400/1149 nt)                       nagA   \n",
       "23837              G→C           R110G (CGT→GGT)                       clsA   \n",
       "183981             G→C          R110G (CGT→GGT)                        clsA   \n",
       "183979           Δ1 bp        intergenic (‑2/+1)                 gatC, gatC   \n",
       "\n",
       "       Reference Seq ALEdb source     sample  coding exp ALE  \\\n",
       "23836      NC_000913       public  1 124 1 1    True   42C 1   \n",
       "183980     NC_000913       public  1 124 1 1    True   42C 1   \n",
       "23837      NC_000913       public  1 124 1 1    True   42C 1   \n",
       "183981     NC_000913       public  1 124 1 1    True   42C 1   \n",
       "183979     NC_000913       public  1 124 1 1   False   42C 1   \n",
       "\n",
       "                     range gene RegulonDB ID  \\\n",
       "23836     (702352, 702372)    {ECK120000625}   \n",
       "183980    (702352, 702372)    {ECK120000625}   \n",
       "23837   (1308318, 1308318)    {ECK120001556}   \n",
       "183981  (1308318, 1308318)    {ECK120001556}   \n",
       "183979  (2173364, 2173364)    {ECK120002276}   \n",
       "\n",
       "                                         genetic features  \n",
       "23836   [{'name': 'nagA', 'RegulonDB ID': 'ECK12000062...  \n",
       "183980  [{'name': 'nagA', 'RegulonDB ID': 'ECK12000062...  \n",
       "23837   [{'name': 'clsA', 'RegulonDB ID': 'ECK12000155...  \n",
       "183981  [{'name': 'clsA', 'RegulonDB ID': 'ECK12000155...  \n",
       "183979  [{'name': 'gatC', 'RegulonDB ID': 'ECK12000227...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "def get_feature_range(feat_id, feat_type):\n",
    "    r = ()\n",
    "    feat_id_col = 0\n",
    "    range_col = \"range\"\n",
    "    feat_id_col = \"GENE_ID\"\n",
    "    feat_type_df = gene_df\n",
    "\n",
    "    df = feat_type_df[feat_type_df[feat_id_col] == feat_id]\n",
    "    if not df.empty:\n",
    "        r = df.iloc[0][range_col]  # Assuming all rows describe the same range.\n",
    "    l = [int(i) for i in r]\n",
    "    return tuple(l)\n",
    "\n",
    "\n",
    "# This will include all implicit intergenic regions, where get_genomic_features(...) does not.\n",
    "# TODO: this may not handle wrapping around the genome from the last gene to the first gene. Write test for this.\n",
    "def get_ordered_genetic_targets(genetic_targets):\n",
    "    ordered_genetic_targets = sorted(\n",
    "        genetic_targets, key=lambda k: k[\"range\"][0])\n",
    "    return ordered_genetic_targets\n",
    "\n",
    "\n",
    "test_genetic_targets = [\n",
    "    {'RegulonDB ID': \"ECK120000003\",\n",
    "     \"range\": (2531463, 2532224)},\n",
    "    {'RegulonDB ID': \"ECK120000001\",\n",
    "     \"range\": (4265782, 4266861)},\n",
    "    {'RegulonDB ID': \"ECK120000002\",\n",
    "     \"range\": (795862, 796551)}]\n",
    "assert(get_ordered_genetic_targets(test_genetic_targets) ==\n",
    "       [{'RegulonDB ID': 'ECK120000002', 'range': (795862, 796551)},\n",
    "        {'RegulonDB ID': 'ECK120000003', 'range': (2531463, 2532224)},\n",
    "        {'RegulonDB ID': 'ECK120000001', 'range': (4265782, 4266861)}])\n",
    "\n",
    "\n",
    "#Simply returning the first gene name in the second column that has the RegulonDB ID in the first column.\n",
    "def get_gene_name(RegulonDB_ID, debug=False):\n",
    "    if debug:\n",
    "        display(RegulonDB_ID, gene_df[gene_df[\"GENE_ID\"] == RegulonDB_ID])\n",
    "    return gene_df[gene_df[\"GENE_ID\"] == RegulonDB_ID].iloc[0][\"GENE_NAME\"]\n",
    "\n",
    "\n",
    "def get_genetic_feat_JSON(mut_row):\n",
    "    l = []\n",
    "    for gene_id in mut_row[\"gene RegulonDB ID\"]:\n",
    "        l.append({\n",
    "            # IDs are unique; assuming only be one entry per DF.\n",
    "            \"name\": get_gene_name(gene_id),\n",
    "            \"RegulonDB ID\": gene_id,\n",
    "            \"range\": get_feature_range(gene_id, \"gene RegulonDB ID\"),\n",
    "            \"feature type\": \"gene\"\n",
    "        })\n",
    "\n",
    "    intergenic_regions = get_feature_hit_set(mut_row[\"range\"], intergenic_df, \"range\", \"RegulonDB ID\")\n",
    "    for intergenic_region in intergenic_regions:\n",
    "        l.append({\n",
    "            \"RegulonDB ID\": intergenic_region,\n",
    "            \"name\": str(get_gene_name(intergenic_region.split('/')[0])) + '/' + str(get_gene_name(intergenic_region.split('/')[1])),\n",
    "            \"range\": intergenic_df[intergenic_df[\"RegulonDB ID\"] == intergenic_region].iloc[0][\"range\"],\n",
    "            \"feature type\": \"unknown\"\n",
    "        })\n",
    "\n",
    "    return l\n",
    "\n",
    "\n",
    "def _f(r):  # parallel_apply doesn't do lambda functions, therefore making temporary function\n",
    "    if r.exp not in NON_K12_EXP_L:\n",
    "        return get_genetic_feat_JSON(r)\n",
    "    else:\n",
    "        return []\n",
    "\n",
    "\n",
    "all_muts_df[\"genetic features\"] = all_muts_df.parallel_apply(_f, axis=1)\n",
    "display(all_muts_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "f27614b9-f3f8-4a83-a26e-b0660f41c250",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "all_muts_df.to_pickle(\"./out_annot_gen_feats.pkl\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  },
  "vscode": {
   "interpreter": {
    "hash": "7ebccac1095285a8f3a7365f16cbaa2bb02a7e2674096928d95b1b0e9894f90a"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
