{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b5f77ca1-ae8f-4777-a074-80199010193c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "321d86bb-7b46-4a99-8fd5-b0d97642ed3d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2003"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>clonal_or_population</th>\n",
       "      <th>tech_rep_description</th>\n",
       "      <th>media_description</th>\n",
       "      <th>carbon_source</th>\n",
       "      <th>nitrogen_source</th>\n",
       "      <th>phosphorous_source</th>\n",
       "      <th>sulfur_source</th>\n",
       "      <th>calcium_source</th>\n",
       "      <th>supplement</th>\n",
       "      <th>...</th>\n",
       "      <th>strain</th>\n",
       "      <th>strain_details</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>reseq_reference</th>\n",
       "      <th>breseq_version</th>\n",
       "      <th>reseq_date</th>\n",
       "      <th>experiment</th>\n",
       "      <th>project</th>\n",
       "      <th>person</th>\n",
       "      <th>doi</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1 F50 I1 R1</td>\n",
       "      <td>clonal</td>\n",
       "      <td></td>\n",
       "      <td>M9</td>\n",
       "      <td>glucose(10)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3),Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "      <td>hexamethylenediamine()</td>\n",
       "      <td>...</td>\n",
       "      <td>511145.0</td>\n",
       "      <td>BOP27</td>\n",
       "      <td></td>\n",
       "      <td>NC_000913_3.gb</td>\n",
       "      <td>0.29.0</td>\n",
       "      <td>20:49:22 03 Jan 2017</td>\n",
       "      <td>TOL hexamethylenediamine</td>\n",
       "      <td>E. coli chemical tolerance</td>\n",
       "      <td>Markus Herrgard</td>\n",
       "      <td>10.1101/634105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A2 F50 I1 R1</td>\n",
       "      <td>clonal</td>\n",
       "      <td></td>\n",
       "      <td>M9</td>\n",
       "      <td>glucose(10)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3),Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "      <td>hexamethylenediamine()</td>\n",
       "      <td>...</td>\n",
       "      <td>511145.0</td>\n",
       "      <td>BOP27</td>\n",
       "      <td></td>\n",
       "      <td>NC_000913_3.gb</td>\n",
       "      <td>0.29.0</td>\n",
       "      <td>22:24:11 03 Jan 2017</td>\n",
       "      <td>TOL hexamethylenediamine</td>\n",
       "      <td>E. coli chemical tolerance</td>\n",
       "      <td>Markus Herrgard</td>\n",
       "      <td>10.1101/634105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A2 F50 I2 R1</td>\n",
       "      <td>clonal</td>\n",
       "      <td></td>\n",
       "      <td>M9</td>\n",
       "      <td>glucose(10)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3),Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "      <td>hexamethylenediamine()</td>\n",
       "      <td>...</td>\n",
       "      <td>511145.0</td>\n",
       "      <td>BOP27</td>\n",
       "      <td></td>\n",
       "      <td>NC_000913_3.gb</td>\n",
       "      <td>0.29.0</td>\n",
       "      <td>19:02:44 03 Jan 2017</td>\n",
       "      <td>TOL hexamethylenediamine</td>\n",
       "      <td>E. coli chemical tolerance</td>\n",
       "      <td>Markus Herrgard</td>\n",
       "      <td>10.1101/634105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A3 F50 I1 R1</td>\n",
       "      <td>clonal</td>\n",
       "      <td></td>\n",
       "      <td>M9</td>\n",
       "      <td>glucose(10)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3),Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "      <td>hexamethylenediamine()</td>\n",
       "      <td>...</td>\n",
       "      <td>511145.0</td>\n",
       "      <td>BOP27</td>\n",
       "      <td></td>\n",
       "      <td>NC_000913_3.gb</td>\n",
       "      <td>0.29.0</td>\n",
       "      <td>19:40:26 03 Jan 2017</td>\n",
       "      <td>TOL hexamethylenediamine</td>\n",
       "      <td>E. coli chemical tolerance</td>\n",
       "      <td>Markus Herrgard</td>\n",
       "      <td>10.1101/634105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A3 F50 I2 R1</td>\n",
       "      <td>clonal</td>\n",
       "      <td></td>\n",
       "      <td>M9</td>\n",
       "      <td>glucose(10)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3),Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "      <td>hexamethylenediamine()</td>\n",
       "      <td>...</td>\n",
       "      <td>511145.0</td>\n",
       "      <td>BOP27</td>\n",
       "      <td></td>\n",
       "      <td>NC_000913_3.gb</td>\n",
       "      <td>0.29.0</td>\n",
       "      <td>13:39:51 03 Jan 2017</td>\n",
       "      <td>TOL hexamethylenediamine</td>\n",
       "      <td>E. coli chemical tolerance</td>\n",
       "      <td>Markus Herrgard</td>\n",
       "      <td>10.1101/634105</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    sample_name clonal_or_population tech_rep_description media_description  \\\n",
       "0  A1 F50 I1 R1               clonal                                     M9   \n",
       "1  A2 F50 I1 R1               clonal                                     M9   \n",
       "2  A2 F50 I2 R1               clonal                                     M9   \n",
       "3  A3 F50 I1 R1               clonal                                     M9   \n",
       "4  A3 F50 I2 R1               clonal                                     M9   \n",
       "\n",
       "  carbon_source nitrogen_source      phosphorous_source sulfur_source  \\\n",
       "0   glucose(10)        NH4Cl(1)  KH2PO4(3),Na2HPO4(6.8)   MgSO4(0.24)   \n",
       "1   glucose(10)        NH4Cl(1)  KH2PO4(3),Na2HPO4(6.8)   MgSO4(0.24)   \n",
       "2   glucose(10)        NH4Cl(1)  KH2PO4(3),Na2HPO4(6.8)   MgSO4(0.24)   \n",
       "3   glucose(10)        NH4Cl(1)  KH2PO4(3),Na2HPO4(6.8)   MgSO4(0.24)   \n",
       "4   glucose(10)        NH4Cl(1)  KH2PO4(3),Na2HPO4(6.8)   MgSO4(0.24)   \n",
       "\n",
       "  calcium_source              supplement  ...    strain strain_details  \\\n",
       "0     CaCl2(0.1)  hexamethylenediamine()  ...  511145.0          BOP27   \n",
       "1     CaCl2(0.1)  hexamethylenediamine()  ...  511145.0          BOP27   \n",
       "2     CaCl2(0.1)  hexamethylenediamine()  ...  511145.0          BOP27   \n",
       "3     CaCl2(0.1)  hexamethylenediamine()  ...  511145.0          BOP27   \n",
       "4     CaCl2(0.1)  hexamethylenediamine()  ...  511145.0          BOP27   \n",
       "\n",
       "  taxonomy_id reseq_reference breseq_version            reseq_date  \\\n",
       "0              NC_000913_3.gb         0.29.0  20:49:22 03 Jan 2017   \n",
       "1              NC_000913_3.gb         0.29.0  22:24:11 03 Jan 2017   \n",
       "2              NC_000913_3.gb         0.29.0  19:02:44 03 Jan 2017   \n",
       "3              NC_000913_3.gb         0.29.0  19:40:26 03 Jan 2017   \n",
       "4              NC_000913_3.gb         0.29.0  13:39:51 03 Jan 2017   \n",
       "\n",
       "                 experiment                     project           person  \\\n",
       "0  TOL hexamethylenediamine  E. coli chemical tolerance  Markus Herrgard   \n",
       "1  TOL hexamethylenediamine  E. coli chemical tolerance  Markus Herrgard   \n",
       "2  TOL hexamethylenediamine  E. coli chemical tolerance  Markus Herrgard   \n",
       "3  TOL hexamethylenediamine  E. coli chemical tolerance  Markus Herrgard   \n",
       "4  TOL hexamethylenediamine  E. coli chemical tolerance  Markus Herrgard   \n",
       "\n",
       "              doi  \n",
       "0  10.1101/634105  \n",
       "1  10.1101/634105  \n",
       "2  10.1101/634105  \n",
       "3  10.1101/634105  \n",
       "4  10.1101/634105  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import os\n",
    "md_df = pd.DataFrame()\n",
    "md_dirs = [\"./data/aledb/md/pub/\",\n",
    "]\n",
    "for md_dir in md_dirs:\n",
    "    for f in os.listdir(md_dir):\n",
    "        md_df = pd.concat([\n",
    "            md_df,\n",
    "            pd.read_csv(md_dir+f)\n",
    "        ])\n",
    "md_df = md_df.fillna('')\n",
    "display(len(md_df), md_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "e0b0b852-097b-49b8-8b07-7d252226f127",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(7670, 20)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exp</th>\n",
       "      <th>ale</th>\n",
       "      <th>flask</th>\n",
       "      <th>isolate</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>presence</th>\n",
       "      <th>Position</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>Details</th>\n",
       "      <th>mutation target annotation</th>\n",
       "      <th>Reference Seq</th>\n",
       "      <th>ALEdb source</th>\n",
       "      <th>sample</th>\n",
       "      <th>coding</th>\n",
       "      <th>exp ALE</th>\n",
       "      <th>range</th>\n",
       "      <th>gene RegulonDB ID</th>\n",
       "      <th>genetic features</th>\n",
       "      <th>afir</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>23836</th>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>702352</td>\n",
       "      <td>DEL</td>\n",
       "      <td>Î”21 bp</td>\n",
       "      <td>coding (380â€‘400/1149 nt)</td>\n",
       "      <td>nagA</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>public</td>\n",
       "      <td>1 124 1 1</td>\n",
       "      <td>True</td>\n",
       "      <td>42C 1</td>\n",
       "      <td>(702352, 702372)</td>\n",
       "      <td>{ECK120000625}</td>\n",
       "      <td>[{'name': 'nagA', 'RegulonDB ID': 'ECK12000062...</td>\n",
       "      <td>A1 F124 I1 R1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>183980</th>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>702352</td>\n",
       "      <td>DEL</td>\n",
       "      <td>Î”21Â bp</td>\n",
       "      <td>codingÂ (380â€‘400/1149Â nt)</td>\n",
       "      <td>nagA</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>public</td>\n",
       "      <td>1 124 1 1</td>\n",
       "      <td>True</td>\n",
       "      <td>42C 1</td>\n",
       "      <td>(702352, 702372)</td>\n",
       "      <td>{ECK120000625}</td>\n",
       "      <td>[{'name': 'nagA', 'RegulonDB ID': 'ECK12000062...</td>\n",
       "      <td>A1 F124 I1 R1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23837</th>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1308318</td>\n",
       "      <td>SNP</td>\n",
       "      <td>Gâ†’C</td>\n",
       "      <td>R110G (CGTâ†’GGT)</td>\n",
       "      <td>clsA</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>public</td>\n",
       "      <td>1 124 1 1</td>\n",
       "      <td>True</td>\n",
       "      <td>42C 1</td>\n",
       "      <td>(1308318, 1308318)</td>\n",
       "      <td>{ECK120001556}</td>\n",
       "      <td>[{'name': 'clsA', 'RegulonDB ID': 'ECK12000155...</td>\n",
       "      <td>A1 F124 I1 R1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>183981</th>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1308318</td>\n",
       "      <td>SNP</td>\n",
       "      <td>Gâ†’C</td>\n",
       "      <td>R110GÂ (CGTâ†’GGT)</td>\n",
       "      <td>clsA</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>public</td>\n",
       "      <td>1 124 1 1</td>\n",
       "      <td>True</td>\n",
       "      <td>42C 1</td>\n",
       "      <td>(1308318, 1308318)</td>\n",
       "      <td>{ECK120001556}</td>\n",
       "      <td>[{'name': 'clsA', 'RegulonDB ID': 'ECK12000155...</td>\n",
       "      <td>A1 F124 I1 R1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>183979</th>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2173364</td>\n",
       "      <td>DEL</td>\n",
       "      <td>Î”1Â bp</td>\n",
       "      <td>intergenicÂ (â€‘2/+1)</td>\n",
       "      <td>gatC, gatC</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>public</td>\n",
       "      <td>1 124 1 1</td>\n",
       "      <td>False</td>\n",
       "      <td>42C 1</td>\n",
       "      <td>(2173364, 2173364)</td>\n",
       "      <td>{ECK120002276}</td>\n",
       "      <td>[{'name': 'gatC', 'RegulonDB ID': 'ECK12000227...</td>\n",
       "      <td>A1 F124 I1 R1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        exp  ale  flask  isolate  tech_rep  presence  Position Mutation Type  \\\n",
       "23836   42C    1    124        1         1       1.0    702352           DEL   \n",
       "183980  42C    1    124        1         1       1.0    702352           DEL   \n",
       "23837   42C    1    124        1         1       1.0   1308318           SNP   \n",
       "183981  42C    1    124        1         1       1.0   1308318           SNP   \n",
       "183979  42C    1    124        1         1       1.0   2173364           DEL   \n",
       "\n",
       "       Sequence Change                   Details mutation target annotation  \\\n",
       "23836           Î”21 bp  coding (380â€‘400/1149 nt)                       nagA   \n",
       "183980          Î”21Â bp  codingÂ (380â€‘400/1149Â nt)                       nagA   \n",
       "23837              Gâ†’C           R110G (CGTâ†’GGT)                       clsA   \n",
       "183981             Gâ†’C          R110GÂ (CGTâ†’GGT)Â                        clsA   \n",
       "183979           Î”1Â bp        intergenicÂ (â€‘2/+1)                 gatC, gatC   \n",
       "\n",
       "       Reference Seq ALEdb source     sample  coding exp ALE  \\\n",
       "23836      NC_000913       public  1 124 1 1    True   42C 1   \n",
       "183980     NC_000913       public  1 124 1 1    True   42C 1   \n",
       "23837      NC_000913       public  1 124 1 1    True   42C 1   \n",
       "183981     NC_000913       public  1 124 1 1    True   42C 1   \n",
       "183979     NC_000913       public  1 124 1 1   False   42C 1   \n",
       "\n",
       "                     range gene RegulonDB ID  \\\n",
       "23836     (702352, 702372)    {ECK120000625}   \n",
       "183980    (702352, 702372)    {ECK120000625}   \n",
       "23837   (1308318, 1308318)    {ECK120001556}   \n",
       "183981  (1308318, 1308318)    {ECK120001556}   \n",
       "183979  (2173364, 2173364)    {ECK120002276}   \n",
       "\n",
       "                                         genetic features           afir  \n",
       "23836   [{'name': 'nagA', 'RegulonDB ID': 'ECK12000062...  A1 F124 I1 R1  \n",
       "183980  [{'name': 'nagA', 'RegulonDB ID': 'ECK12000062...  A1 F124 I1 R1  \n",
       "23837   [{'name': 'clsA', 'RegulonDB ID': 'ECK12000155...  A1 F124 I1 R1  \n",
       "183981  [{'name': 'clsA', 'RegulonDB ID': 'ECK12000155...  A1 F124 I1 R1  \n",
       "183979  [{'name': 'gatC', 'RegulonDB ID': 'ECK12000227...  A1 F124 I1 R1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "all_muts_df = pd.read_pickle(\"./out_annot_gen_feats.pkl\")\n",
    "all_muts_df['afir'] = all_muts_df.apply(lambda r: 'A'+str(r.ale)+' F'+str(r.flask)+' I'+str(r.isolate)+' R'+str(r.tech_rep),axis=1)  # To match MD sample name\n",
    "display(all_muts_df.shape, all_muts_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "23442986-3eff-4f22-9e61-f2e89e0aff0f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exp</th>\n",
       "      <th>ale</th>\n",
       "      <th>flask</th>\n",
       "      <th>isolate</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>presence</th>\n",
       "      <th>Position</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>Details</th>\n",
       "      <th>mutation target annotation</th>\n",
       "      <th>Reference Seq</th>\n",
       "      <th>ALEdb source</th>\n",
       "      <th>sample</th>\n",
       "      <th>coding</th>\n",
       "      <th>exp ALE</th>\n",
       "      <th>range</th>\n",
       "      <th>gene RegulonDB ID</th>\n",
       "      <th>genetic features</th>\n",
       "      <th>afir</th>\n",
       "      <th>media_description</th>\n",
       "      <th>carbon_source</th>\n",
       "      <th>nitrogen_source</th>\n",
       "      <th>phosphorous_source</th>\n",
       "      <th>sulfur_source</th>\n",
       "      <th>calcium_source</th>\n",
       "      <th>supplement</th>\n",
       "      <th>temperature</th>\n",
       "      <th>strain</th>\n",
       "      <th>strain_details</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>tech_rep_description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>23836</th>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>702352</td>\n",
       "      <td>DEL</td>\n",
       "      <td>Î”21 bp</td>\n",
       "      <td>coding (380â€‘400/1149 nt)</td>\n",
       "      <td>nagA</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>public</td>\n",
       "      <td>1 124 1 1</td>\n",
       "      <td>True</td>\n",
       "      <td>42C 1</td>\n",
       "      <td>(702352, 702372)</td>\n",
       "      <td>{ECK120000625}</td>\n",
       "      <td>[{'name': 'nagA', 'RegulonDB ID': 'ECK12000062...</td>\n",
       "      <td>A1 F124 I1 R1</td>\n",
       "      <td>M9</td>\n",
       "      <td>glucose(4)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3),Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "      <td>NaCl(0.5)</td>\n",
       "      <td>42</td>\n",
       "      <td>511145.0</td>\n",
       "      <td>BOP27</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>183980</th>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>702352</td>\n",
       "      <td>DEL</td>\n",
       "      <td>Î”21Â bp</td>\n",
       "      <td>codingÂ (380â€‘400/1149Â nt)</td>\n",
       "      <td>nagA</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>public</td>\n",
       "      <td>1 124 1 1</td>\n",
       "      <td>True</td>\n",
       "      <td>42C 1</td>\n",
       "      <td>(702352, 702372)</td>\n",
       "      <td>{ECK120000625}</td>\n",
       "      <td>[{'name': 'nagA', 'RegulonDB ID': 'ECK12000062...</td>\n",
       "      <td>A1 F124 I1 R1</td>\n",
       "      <td>M9</td>\n",
       "      <td>glucose(4)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3),Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "      <td>NaCl(0.5)</td>\n",
       "      <td>42</td>\n",
       "      <td>511145.0</td>\n",
       "      <td>BOP27</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23837</th>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1308318</td>\n",
       "      <td>SNP</td>\n",
       "      <td>Gâ†’C</td>\n",
       "      <td>R110G (CGTâ†’GGT)</td>\n",
       "      <td>clsA</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>public</td>\n",
       "      <td>1 124 1 1</td>\n",
       "      <td>True</td>\n",
       "      <td>42C 1</td>\n",
       "      <td>(1308318, 1308318)</td>\n",
       "      <td>{ECK120001556}</td>\n",
       "      <td>[{'name': 'clsA', 'RegulonDB ID': 'ECK12000155...</td>\n",
       "      <td>A1 F124 I1 R1</td>\n",
       "      <td>M9</td>\n",
       "      <td>glucose(4)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3),Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "      <td>NaCl(0.5)</td>\n",
       "      <td>42</td>\n",
       "      <td>511145.0</td>\n",
       "      <td>BOP27</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>183981</th>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1308318</td>\n",
       "      <td>SNP</td>\n",
       "      <td>Gâ†’C</td>\n",
       "      <td>R110GÂ (CGTâ†’GGT)</td>\n",
       "      <td>clsA</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>public</td>\n",
       "      <td>1 124 1 1</td>\n",
       "      <td>True</td>\n",
       "      <td>42C 1</td>\n",
       "      <td>(1308318, 1308318)</td>\n",
       "      <td>{ECK120001556}</td>\n",
       "      <td>[{'name': 'clsA', 'RegulonDB ID': 'ECK12000155...</td>\n",
       "      <td>A1 F124 I1 R1</td>\n",
       "      <td>M9</td>\n",
       "      <td>glucose(4)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3),Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "      <td>NaCl(0.5)</td>\n",
       "      <td>42</td>\n",
       "      <td>511145.0</td>\n",
       "      <td>BOP27</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>183979</th>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2173364</td>\n",
       "      <td>DEL</td>\n",
       "      <td>Î”1Â bp</td>\n",
       "      <td>intergenicÂ (â€‘2/+1)</td>\n",
       "      <td>gatC, gatC</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>public</td>\n",
       "      <td>1 124 1 1</td>\n",
       "      <td>False</td>\n",
       "      <td>42C 1</td>\n",
       "      <td>(2173364, 2173364)</td>\n",
       "      <td>{ECK120002276}</td>\n",
       "      <td>[{'name': 'gatC', 'RegulonDB ID': 'ECK12000227...</td>\n",
       "      <td>A1 F124 I1 R1</td>\n",
       "      <td>M9</td>\n",
       "      <td>glucose(4)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3),Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "      <td>NaCl(0.5)</td>\n",
       "      <td>42</td>\n",
       "      <td>511145.0</td>\n",
       "      <td>BOP27</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121642</th>\n",
       "      <td>tpiVCH</td>\n",
       "      <td>28</td>\n",
       "      <td>14</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1544022</td>\n",
       "      <td>SUB</td>\n",
       "      <td>1 bpâ†’AA</td>\n",
       "      <td>coding (39/1389 nt)</td>\n",
       "      <td>narU</td>\n",
       "      <td>NC_000913_3_vch_tpiA</td>\n",
       "      <td>public</td>\n",
       "      <td>28 14 1 1</td>\n",
       "      <td>True</td>\n",
       "      <td>tpiVCH 28</td>\n",
       "      <td>(1544022, 1544022)</td>\n",
       "      <td>{ECK120002052}</td>\n",
       "      <td>[{'name': 'narU', 'RegulonDB ID': 'ECK12000205...</td>\n",
       "      <td>A28 F14 I1 R1</td>\n",
       "      <td>M9</td>\n",
       "      <td>glucose(4)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3),Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "      <td></td>\n",
       "      <td>37</td>\n",
       "      <td>511145.0</td>\n",
       "      <td>V. cholerae tpi E. coli K-12 MG1655</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121644</th>\n",
       "      <td>tpiVCH</td>\n",
       "      <td>28</td>\n",
       "      <td>14</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2214152</td>\n",
       "      <td>SNP</td>\n",
       "      <td>Aâ†’T</td>\n",
       "      <td>Y165N (TATâ†’AAT)</td>\n",
       "      <td>yehU</td>\n",
       "      <td>NC_000913_3_vch_tpiA</td>\n",
       "      <td>public</td>\n",
       "      <td>28 14 1 1</td>\n",
       "      <td>True</td>\n",
       "      <td>tpiVCH 28</td>\n",
       "      <td>(2214152, 2214152)</td>\n",
       "      <td>{ECK120001926}</td>\n",
       "      <td>[{'name': 'btsS', 'RegulonDB ID': 'ECK12000192...</td>\n",
       "      <td>A28 F14 I1 R1</td>\n",
       "      <td>M9</td>\n",
       "      <td>glucose(4)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3),Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "      <td></td>\n",
       "      <td>37</td>\n",
       "      <td>511145.0</td>\n",
       "      <td>V. cholerae tpi E. coli K-12 MG1655</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121662</th>\n",
       "      <td>tpiVCH</td>\n",
       "      <td>28</td>\n",
       "      <td>25</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3815825</td>\n",
       "      <td>SNP</td>\n",
       "      <td>Câ†’T</td>\n",
       "      <td>intergenic (â€‘57/+9)</td>\n",
       "      <td>pyrE, rph</td>\n",
       "      <td>NC_000913_3_vch_tpiA</td>\n",
       "      <td>public</td>\n",
       "      <td>28 25 1 1</td>\n",
       "      <td>False</td>\n",
       "      <td>tpiVCH 28</td>\n",
       "      <td>(3815825, 3815825)</td>\n",
       "      <td>{}</td>\n",
       "      <td>[{'RegulonDB ID': 'ECK120000799/ECK120000854',...</td>\n",
       "      <td>A28 F25 I1 R1</td>\n",
       "      <td>M9</td>\n",
       "      <td>glucose(4)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3),Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "      <td></td>\n",
       "      <td>37</td>\n",
       "      <td>511145.0</td>\n",
       "      <td>V. cholerae tpi E. coli K-12 MG1655</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121675</th>\n",
       "      <td>tpiVCH</td>\n",
       "      <td>28</td>\n",
       "      <td>66</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4159190</td>\n",
       "      <td>SNP</td>\n",
       "      <td>Tâ†’A</td>\n",
       "      <td>L224Q (CTGâ†’CAG)</td>\n",
       "      <td>oxyR</td>\n",
       "      <td>NC_000913_3_vch_tpiA</td>\n",
       "      <td>public</td>\n",
       "      <td>28 66 1 1</td>\n",
       "      <td>True</td>\n",
       "      <td>tpiVCH 28</td>\n",
       "      <td>(4159190, 4159190)</td>\n",
       "      <td>{ECK120000674}</td>\n",
       "      <td>[{'name': 'oxyR', 'RegulonDB ID': 'ECK12000067...</td>\n",
       "      <td>A28 F66 I1 R1</td>\n",
       "      <td>M9</td>\n",
       "      <td>glucose(4)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3),Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "      <td></td>\n",
       "      <td>37</td>\n",
       "      <td>511145.0</td>\n",
       "      <td>V. cholerae tpi E. coli K-12 MG1655</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121676</th>\n",
       "      <td>tpiVCH</td>\n",
       "      <td>28</td>\n",
       "      <td>66</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4182833</td>\n",
       "      <td>SNP</td>\n",
       "      <td>Câ†’T</td>\n",
       "      <td>P520L (CCGâ†’CTG)</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>NC_000913_3_vch_tpiA</td>\n",
       "      <td>public</td>\n",
       "      <td>28 66 1 1</td>\n",
       "      <td>True</td>\n",
       "      <td>tpiVCH 28</td>\n",
       "      <td>(4182833, 4182833)</td>\n",
       "      <td>{ECK120000885}</td>\n",
       "      <td>[{'name': 'rpoB', 'RegulonDB ID': 'ECK12000088...</td>\n",
       "      <td>A28 F66 I1 R1</td>\n",
       "      <td>M9</td>\n",
       "      <td>glucose(4)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3),Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "      <td></td>\n",
       "      <td>37</td>\n",
       "      <td>511145.0</td>\n",
       "      <td>V. cholerae tpi E. coli K-12 MG1655</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7670 rows Ã— 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           exp  ale  flask  isolate  tech_rep  presence  Position  \\\n",
       "23836      42C    1    124        1         1       1.0    702352   \n",
       "183980     42C    1    124        1         1       1.0    702352   \n",
       "23837      42C    1    124        1         1       1.0   1308318   \n",
       "183981     42C    1    124        1         1       1.0   1308318   \n",
       "183979     42C    1    124        1         1       1.0   2173364   \n",
       "...        ...  ...    ...      ...       ...       ...       ...   \n",
       "121642  tpiVCH   28     14        1         1       1.0   1544022   \n",
       "121644  tpiVCH   28     14        1         1       1.0   2214152   \n",
       "121662  tpiVCH   28     25        1         1       1.0   3815825   \n",
       "121675  tpiVCH   28     66        1         1       1.0   4159190   \n",
       "121676  tpiVCH   28     66        1         1       1.0   4182833   \n",
       "\n",
       "       Mutation Type Sequence Change                   Details  \\\n",
       "23836            DEL          Î”21 bp  coding (380â€‘400/1149 nt)   \n",
       "183980           DEL          Î”21Â bp  codingÂ (380â€‘400/1149Â nt)   \n",
       "23837            SNP             Gâ†’C           R110G (CGTâ†’GGT)   \n",
       "183981           SNP             Gâ†’C          R110GÂ (CGTâ†’GGT)Â    \n",
       "183979           DEL           Î”1Â bp        intergenicÂ (â€‘2/+1)   \n",
       "...              ...             ...                       ...   \n",
       "121642           SUB         1 bpâ†’AA       coding (39/1389 nt)   \n",
       "121644           SNP             Aâ†’T           Y165N (TATâ†’AAT)   \n",
       "121662           SNP             Câ†’T       intergenic (â€‘57/+9)   \n",
       "121675           SNP             Tâ†’A           L224Q (CTGâ†’CAG)   \n",
       "121676           SNP             Câ†’T           P520L (CCGâ†’CTG)   \n",
       "\n",
       "       mutation target annotation         Reference Seq ALEdb source  \\\n",
       "23836                        nagA             NC_000913       public   \n",
       "183980                       nagA             NC_000913       public   \n",
       "23837                        clsA             NC_000913       public   \n",
       "183981                       clsA             NC_000913       public   \n",
       "183979                 gatC, gatC             NC_000913       public   \n",
       "...                           ...                   ...          ...   \n",
       "121642                       narU  NC_000913_3_vch_tpiA       public   \n",
       "121644                       yehU  NC_000913_3_vch_tpiA       public   \n",
       "121662                  pyrE, rph  NC_000913_3_vch_tpiA       public   \n",
       "121675                       oxyR  NC_000913_3_vch_tpiA       public   \n",
       "121676                       rpoB  NC_000913_3_vch_tpiA       public   \n",
       "\n",
       "           sample  coding    exp ALE               range gene RegulonDB ID  \\\n",
       "23836   1 124 1 1    True      42C 1    (702352, 702372)    {ECK120000625}   \n",
       "183980  1 124 1 1    True      42C 1    (702352, 702372)    {ECK120000625}   \n",
       "23837   1 124 1 1    True      42C 1  (1308318, 1308318)    {ECK120001556}   \n",
       "183981  1 124 1 1    True      42C 1  (1308318, 1308318)    {ECK120001556}   \n",
       "183979  1 124 1 1   False      42C 1  (2173364, 2173364)    {ECK120002276}   \n",
       "...           ...     ...        ...                 ...               ...   \n",
       "121642  28 14 1 1    True  tpiVCH 28  (1544022, 1544022)    {ECK120002052}   \n",
       "121644  28 14 1 1    True  tpiVCH 28  (2214152, 2214152)    {ECK120001926}   \n",
       "121662  28 25 1 1   False  tpiVCH 28  (3815825, 3815825)                {}   \n",
       "121675  28 66 1 1    True  tpiVCH 28  (4159190, 4159190)    {ECK120000674}   \n",
       "121676  28 66 1 1    True  tpiVCH 28  (4182833, 4182833)    {ECK120000885}   \n",
       "\n",
       "                                         genetic features           afir  \\\n",
       "23836   [{'name': 'nagA', 'RegulonDB ID': 'ECK12000062...  A1 F124 I1 R1   \n",
       "183980  [{'name': 'nagA', 'RegulonDB ID': 'ECK12000062...  A1 F124 I1 R1   \n",
       "23837   [{'name': 'clsA', 'RegulonDB ID': 'ECK12000155...  A1 F124 I1 R1   \n",
       "183981  [{'name': 'clsA', 'RegulonDB ID': 'ECK12000155...  A1 F124 I1 R1   \n",
       "183979  [{'name': 'gatC', 'RegulonDB ID': 'ECK12000227...  A1 F124 I1 R1   \n",
       "...                                                   ...            ...   \n",
       "121642  [{'name': 'narU', 'RegulonDB ID': 'ECK12000205...  A28 F14 I1 R1   \n",
       "121644  [{'name': 'btsS', 'RegulonDB ID': 'ECK12000192...  A28 F14 I1 R1   \n",
       "121662  [{'RegulonDB ID': 'ECK120000799/ECK120000854',...  A28 F25 I1 R1   \n",
       "121675  [{'name': 'oxyR', 'RegulonDB ID': 'ECK12000067...  A28 F66 I1 R1   \n",
       "121676  [{'name': 'rpoB', 'RegulonDB ID': 'ECK12000088...  A28 F66 I1 R1   \n",
       "\n",
       "       media_description carbon_source nitrogen_source  \\\n",
       "23836                 M9    glucose(4)        NH4Cl(1)   \n",
       "183980                M9    glucose(4)        NH4Cl(1)   \n",
       "23837                 M9    glucose(4)        NH4Cl(1)   \n",
       "183981                M9    glucose(4)        NH4Cl(1)   \n",
       "183979                M9    glucose(4)        NH4Cl(1)   \n",
       "...                  ...           ...             ...   \n",
       "121642                M9    glucose(4)        NH4Cl(1)   \n",
       "121644                M9    glucose(4)        NH4Cl(1)   \n",
       "121662                M9    glucose(4)        NH4Cl(1)   \n",
       "121675                M9    glucose(4)        NH4Cl(1)   \n",
       "121676                M9    glucose(4)        NH4Cl(1)   \n",
       "\n",
       "            phosphorous_source sulfur_source calcium_source supplement  \\\n",
       "23836   KH2PO4(3),Na2HPO4(6.8)   MgSO4(0.24)     CaCl2(0.1)  NaCl(0.5)   \n",
       "183980  KH2PO4(3),Na2HPO4(6.8)   MgSO4(0.24)     CaCl2(0.1)  NaCl(0.5)   \n",
       "23837   KH2PO4(3),Na2HPO4(6.8)   MgSO4(0.24)     CaCl2(0.1)  NaCl(0.5)   \n",
       "183981  KH2PO4(3),Na2HPO4(6.8)   MgSO4(0.24)     CaCl2(0.1)  NaCl(0.5)   \n",
       "183979  KH2PO4(3),Na2HPO4(6.8)   MgSO4(0.24)     CaCl2(0.1)  NaCl(0.5)   \n",
       "...                        ...           ...            ...        ...   \n",
       "121642  KH2PO4(3),Na2HPO4(6.8)   MgSO4(0.24)     CaCl2(0.1)              \n",
       "121644  KH2PO4(3),Na2HPO4(6.8)   MgSO4(0.24)     CaCl2(0.1)              \n",
       "121662  KH2PO4(3),Na2HPO4(6.8)   MgSO4(0.24)     CaCl2(0.1)              \n",
       "121675  KH2PO4(3),Na2HPO4(6.8)   MgSO4(0.24)     CaCl2(0.1)              \n",
       "121676  KH2PO4(3),Na2HPO4(6.8)   MgSO4(0.24)     CaCl2(0.1)              \n",
       "\n",
       "       temperature    strain                       strain_details taxonomy_id  \\\n",
       "23836           42  511145.0                                BOP27               \n",
       "183980          42  511145.0                                BOP27               \n",
       "23837           42  511145.0                                BOP27               \n",
       "183981          42  511145.0                                BOP27               \n",
       "183979          42  511145.0                                BOP27               \n",
       "...            ...       ...                                  ...         ...   \n",
       "121642          37  511145.0  V. cholerae tpi E. coli K-12 MG1655               \n",
       "121644          37  511145.0  V. cholerae tpi E. coli K-12 MG1655               \n",
       "121662          37  511145.0  V. cholerae tpi E. coli K-12 MG1655               \n",
       "121675          37  511145.0  V. cholerae tpi E. coli K-12 MG1655               \n",
       "121676          37  511145.0  V. cholerae tpi E. coli K-12 MG1655               \n",
       "\n",
       "       tech_rep_description  \n",
       "23836                        \n",
       "183980                       \n",
       "23837                        \n",
       "183981                       \n",
       "183979                       \n",
       "...                     ...  \n",
       "121642                       \n",
       "121644                       \n",
       "121662                       \n",
       "121675                       \n",
       "121676                       \n",
       "\n",
       "[7670 rows x 32 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os, sys\n",
    "module_path = os.path.abspath(os.path.join('..'))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "from aledbmutil.md import METADATA_CATEGORIES_FOR_ASSOCIATIONS\n",
    "\n",
    "\n",
    "pd.set_option('display.max_columns', None)\n",
    "\n",
    "\n",
    "all_muts_df[METADATA_CATEGORIES_FOR_ASSOCIATIONS] = ''  # Initialization purposes\n",
    "\n",
    "\n",
    "def _get_sample_md(mut):\n",
    "    srs = pd.Series({k:'' for k in METADATA_CATEGORIES_FOR_ASSOCIATIONS})\n",
    "    df = md_df[(md_df.sample_name==mut['afir']) & (md_df.experiment==mut['exp'])]\n",
    "    if len(df) > 0:\n",
    "        srs = df[METADATA_CATEGORIES_FOR_ASSOCIATIONS].iloc[0]\n",
    "    return srs\n",
    "\n",
    "\n",
    "all_muts_df[METADATA_CATEGORIES_FOR_ASSOCIATIONS] = all_muts_df.apply(_get_sample_md, axis=1)\n",
    "all_muts_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "1011c14f-2d29-43b9-8bc4-5a83ecff1b37",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(7670, 32)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exp</th>\n",
       "      <th>ale</th>\n",
       "      <th>flask</th>\n",
       "      <th>isolate</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>presence</th>\n",
       "      <th>Position</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>Details</th>\n",
       "      <th>mutation target annotation</th>\n",
       "      <th>Reference Seq</th>\n",
       "      <th>ALEdb source</th>\n",
       "      <th>sample</th>\n",
       "      <th>coding</th>\n",
       "      <th>exp ALE</th>\n",
       "      <th>range</th>\n",
       "      <th>gene RegulonDB ID</th>\n",
       "      <th>genetic features</th>\n",
       "      <th>afir</th>\n",
       "      <th>media_description</th>\n",
       "      <th>carbon_source</th>\n",
       "      <th>nitrogen_source</th>\n",
       "      <th>phosphorous_source</th>\n",
       "      <th>sulfur_source</th>\n",
       "      <th>calcium_source</th>\n",
       "      <th>supplement</th>\n",
       "      <th>temperature</th>\n",
       "      <th>strain</th>\n",
       "      <th>strain_details</th>\n",
       "      <th>taxonomy_id</th>\n",
       "      <th>tech_rep_description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>702352</td>\n",
       "      <td>DEL</td>\n",
       "      <td>Î”21 bp</td>\n",
       "      <td>coding (380â€‘400/1149 nt)</td>\n",
       "      <td>nagA</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>public</td>\n",
       "      <td>1 124 1 1</td>\n",
       "      <td>True</td>\n",
       "      <td>42C 1</td>\n",
       "      <td>(702352, 702372)</td>\n",
       "      <td>{ECK120000625}</td>\n",
       "      <td>[{'name': 'nagA', 'RegulonDB ID': 'ECK12000062...</td>\n",
       "      <td>A1 F124 I1 R1</td>\n",
       "      <td>M9</td>\n",
       "      <td>glucose(4)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3),Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "      <td>NaCl(0.5)</td>\n",
       "      <td>42</td>\n",
       "      <td>511145.0</td>\n",
       "      <td>BOP27</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>702352</td>\n",
       "      <td>DEL</td>\n",
       "      <td>Î”21Â bp</td>\n",
       "      <td>codingÂ (380â€‘400/1149Â nt)</td>\n",
       "      <td>nagA</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>public</td>\n",
       "      <td>1 124 1 1</td>\n",
       "      <td>True</td>\n",
       "      <td>42C 1</td>\n",
       "      <td>(702352, 702372)</td>\n",
       "      <td>{ECK120000625}</td>\n",
       "      <td>[{'name': 'nagA', 'RegulonDB ID': 'ECK12000062...</td>\n",
       "      <td>A1 F124 I1 R1</td>\n",
       "      <td>M9</td>\n",
       "      <td>glucose(4)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3),Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "      <td>NaCl(0.5)</td>\n",
       "      <td>42</td>\n",
       "      <td>511145.0</td>\n",
       "      <td>BOP27</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1308318</td>\n",
       "      <td>SNP</td>\n",
       "      <td>Gâ†’C</td>\n",
       "      <td>R110G (CGTâ†’GGT)</td>\n",
       "      <td>clsA</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>public</td>\n",
       "      <td>1 124 1 1</td>\n",
       "      <td>True</td>\n",
       "      <td>42C 1</td>\n",
       "      <td>(1308318, 1308318)</td>\n",
       "      <td>{ECK120001556}</td>\n",
       "      <td>[{'name': 'clsA', 'RegulonDB ID': 'ECK12000155...</td>\n",
       "      <td>A1 F124 I1 R1</td>\n",
       "      <td>M9</td>\n",
       "      <td>glucose(4)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3),Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "      <td>NaCl(0.5)</td>\n",
       "      <td>42</td>\n",
       "      <td>511145.0</td>\n",
       "      <td>BOP27</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1308318</td>\n",
       "      <td>SNP</td>\n",
       "      <td>Gâ†’C</td>\n",
       "      <td>R110GÂ (CGTâ†’GGT)</td>\n",
       "      <td>clsA</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>public</td>\n",
       "      <td>1 124 1 1</td>\n",
       "      <td>True</td>\n",
       "      <td>42C 1</td>\n",
       "      <td>(1308318, 1308318)</td>\n",
       "      <td>{ECK120001556}</td>\n",
       "      <td>[{'name': 'clsA', 'RegulonDB ID': 'ECK12000155...</td>\n",
       "      <td>A1 F124 I1 R1</td>\n",
       "      <td>M9</td>\n",
       "      <td>glucose(4)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3),Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "      <td>NaCl(0.5)</td>\n",
       "      <td>42</td>\n",
       "      <td>511145.0</td>\n",
       "      <td>BOP27</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>42C</td>\n",
       "      <td>1</td>\n",
       "      <td>124</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2173364</td>\n",
       "      <td>DEL</td>\n",
       "      <td>Î”1Â bp</td>\n",
       "      <td>intergenicÂ (â€‘2/+1)</td>\n",
       "      <td>gatC, gatC</td>\n",
       "      <td>NC_000913</td>\n",
       "      <td>public</td>\n",
       "      <td>1 124 1 1</td>\n",
       "      <td>False</td>\n",
       "      <td>42C 1</td>\n",
       "      <td>(2173364, 2173364)</td>\n",
       "      <td>{ECK120002276}</td>\n",
       "      <td>[{'name': 'gatC', 'RegulonDB ID': 'ECK12000227...</td>\n",
       "      <td>A1 F124 I1 R1</td>\n",
       "      <td>M9</td>\n",
       "      <td>glucose(4)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>KH2PO4(3),Na2HPO4(6.8)</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "      <td>CaCl2(0.1)</td>\n",
       "      <td>NaCl(0.5)</td>\n",
       "      <td>42</td>\n",
       "      <td>511145.0</td>\n",
       "      <td>BOP27</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   exp  ale  flask  isolate  tech_rep  presence  Position Mutation Type  \\\n",
       "0  42C    1    124        1         1       1.0    702352           DEL   \n",
       "1  42C    1    124        1         1       1.0    702352           DEL   \n",
       "2  42C    1    124        1         1       1.0   1308318           SNP   \n",
       "3  42C    1    124        1         1       1.0   1308318           SNP   \n",
       "4  42C    1    124        1         1       1.0   2173364           DEL   \n",
       "\n",
       "  Sequence Change                   Details mutation target annotation  \\\n",
       "0          Î”21 bp  coding (380â€‘400/1149 nt)                       nagA   \n",
       "1          Î”21Â bp  codingÂ (380â€‘400/1149Â nt)                       nagA   \n",
       "2             Gâ†’C           R110G (CGTâ†’GGT)                       clsA   \n",
       "3             Gâ†’C          R110GÂ (CGTâ†’GGT)Â                        clsA   \n",
       "4           Î”1Â bp        intergenicÂ (â€‘2/+1)                 gatC, gatC   \n",
       "\n",
       "  Reference Seq ALEdb source     sample  coding exp ALE               range  \\\n",
       "0     NC_000913       public  1 124 1 1    True   42C 1    (702352, 702372)   \n",
       "1     NC_000913       public  1 124 1 1    True   42C 1    (702352, 702372)   \n",
       "2     NC_000913       public  1 124 1 1    True   42C 1  (1308318, 1308318)   \n",
       "3     NC_000913       public  1 124 1 1    True   42C 1  (1308318, 1308318)   \n",
       "4     NC_000913       public  1 124 1 1   False   42C 1  (2173364, 2173364)   \n",
       "\n",
       "  gene RegulonDB ID                                   genetic features  \\\n",
       "0    {ECK120000625}  [{'name': 'nagA', 'RegulonDB ID': 'ECK12000062...   \n",
       "1    {ECK120000625}  [{'name': 'nagA', 'RegulonDB ID': 'ECK12000062...   \n",
       "2    {ECK120001556}  [{'name': 'clsA', 'RegulonDB ID': 'ECK12000155...   \n",
       "3    {ECK120001556}  [{'name': 'clsA', 'RegulonDB ID': 'ECK12000155...   \n",
       "4    {ECK120002276}  [{'name': 'gatC', 'RegulonDB ID': 'ECK12000227...   \n",
       "\n",
       "            afir media_description carbon_source nitrogen_source  \\\n",
       "0  A1 F124 I1 R1                M9    glucose(4)        NH4Cl(1)   \n",
       "1  A1 F124 I1 R1                M9    glucose(4)        NH4Cl(1)   \n",
       "2  A1 F124 I1 R1                M9    glucose(4)        NH4Cl(1)   \n",
       "3  A1 F124 I1 R1                M9    glucose(4)        NH4Cl(1)   \n",
       "4  A1 F124 I1 R1                M9    glucose(4)        NH4Cl(1)   \n",
       "\n",
       "       phosphorous_source sulfur_source calcium_source supplement temperature  \\\n",
       "0  KH2PO4(3),Na2HPO4(6.8)   MgSO4(0.24)     CaCl2(0.1)  NaCl(0.5)          42   \n",
       "1  KH2PO4(3),Na2HPO4(6.8)   MgSO4(0.24)     CaCl2(0.1)  NaCl(0.5)          42   \n",
       "2  KH2PO4(3),Na2HPO4(6.8)   MgSO4(0.24)     CaCl2(0.1)  NaCl(0.5)          42   \n",
       "3  KH2PO4(3),Na2HPO4(6.8)   MgSO4(0.24)     CaCl2(0.1)  NaCl(0.5)          42   \n",
       "4  KH2PO4(3),Na2HPO4(6.8)   MgSO4(0.24)     CaCl2(0.1)  NaCl(0.5)          42   \n",
       "\n",
       "     strain strain_details taxonomy_id tech_rep_description  \n",
       "0  511145.0          BOP27                                   \n",
       "1  511145.0          BOP27                                   \n",
       "2  511145.0          BOP27                                   \n",
       "3  511145.0          BOP27                                   \n",
       "4  511145.0          BOP27                                   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "all_muts_df = all_muts_df.reset_index(drop=True)\n",
    "\n",
    "# The execution of get_ALE_max_freq_mut_df(get_endpoints_only=True) earlier in the pipeline should have already removed all duplicate mutations\n",
    "# Commenting the old logic (above) out. To use only when using more than just endpoints. Annotated in NB1 for the need to uncomment this region is not using endpoints.\n",
    "# This code is inappropriate anyways because you can have two different and unique mutations to the same gene though with different position, or same position though with different sequence changes.\n",
    "# Important to note that associations are made on the resolution of genomic feature and not position, though two mutations with different detail though the same feature will be represented, as currently expected.\n",
    "# The logic uncommented above is therefore incomplete for removing duplicates for sample sets that include midpoints and endpoints.\n",
    "\n",
    "# The below is the necessary logic without this filtering.\n",
    "exp_target_cond_df = all_muts_df.copy()\n",
    "display(exp_target_cond_df.shape, exp_target_cond_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "d3ebed22-3685-4b1b-867d-667d7f3bfc59",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'',\n",
       " '(NH4)2SO4(2)',\n",
       " '(NH4)2SO4(2),MgSO4(.12)',\n",
       " '20uM FeSO4 and 10mM sodium citrate',\n",
       " 37,\n",
       " 38,\n",
       " 42,\n",
       " 511145.0,\n",
       " 'B. melitensis pgi E. coli K-12 MG1655',\n",
       " 'BOP1000',\n",
       " 'BOP27',\n",
       " 'BOP27 E. coli K-12 MG1655',\n",
       " 'C13H10N2O(.168184)',\n",
       " 'C6H5FeO7(0.00049),casamino(5)',\n",
       " 'CGSC6300',\n",
       " 'CaCl2(0.1)',\n",
       " 'CaCl2(0.1mM)',\n",
       " 'D-2-deoxyribose()',\n",
       " 'D-arabinose()',\n",
       " 'D-lyxose()',\n",
       " 'E. coli K-12 MG1655',\n",
       " 'E. coli K-12 MG1655 with all UAG stop codons replaced with UAA.',\n",
       " 'E. coli K12 MG1655 Î”pgi',\n",
       " 'E. coli K12 MG1655 Î”sdaA Î”sdaB Î”tdcG Î”glyA',\n",
       " 'Escherichia coli MS56',\n",
       " 'GLU ALE4 EP',\n",
       " 'Glucose(4)',\n",
       " 'Glucose(5)',\n",
       " 'H. sapien pgi E. coli K-12 MG1655',\n",
       " 'H. sapien tpi E. coli K-12 MG1655',\n",
       " 'Illumina/ 150',\n",
       " 'Illumina/ 50',\n",
       " 'Illumina/ 600',\n",
       " 'JCL260',\n",
       " 'KH2PO4(13.6)',\n",
       " 'KH2PO4(3),Na2HPO4(6.8)',\n",
       " 'KH2PO4(3),Na2PO4(6.8)',\n",
       " 'KH2PO4(3g/L),Na2HPO4(6.78g/L)',\n",
       " 'KH2PO4(6),Na2HPO4(13.6)',\n",
       " 'Kapa/ ',\n",
       " 'LB',\n",
       " 'M63',\n",
       " 'M9',\n",
       " 'MG1655',\n",
       " 'MgSO4(0.24)',\n",
       " 'MgSO4(1mM)',\n",
       " 'MgSO4(24)',\n",
       " 'NC_000913',\n",
       " 'NH4Cl(1)',\n",
       " 'NH4Cl(1g/L)',\n",
       " 'NH4Cl(2)',\n",
       " 'NaCl(0.5)',\n",
       " 'NaCl(1g/L),trace elements(), NaCl(0.5)',\n",
       " 'None',\n",
       " 'P. aerophilum pgi E. coli K-12 MG1655',\n",
       " 'P. aerophilum tpi E. coli K-12 MG1655',\n",
       " 'Qiagen/ 600',\n",
       " 'U-13C glucose(2)',\n",
       " 'V. cholerae pgi E. coli K-12 MG1655',\n",
       " 'V. cholerae tpi E. coli K-12 MG1655',\n",
       " 'WT P. aeruginosa was first grown in the medium, and an increasing % of this spent\" medium was mixed with fresh medium to grow the E. coli\"',\n",
       " 'glucose(2)',\n",
       " 'glucose(2g/L)',\n",
       " 'glucose(3)',\n",
       " 'glucose(4)',\n",
       " 'glucose(4g/L)',\n",
       " 'glycerol(2.5)',\n",
       " 'glycine(2mM),L-serine(), NaCl(0.5)',\n",
       " 'initially supplemented with 0.1% LB medium reduced in a stepwise manner to eventually achieve supplement-free growth. ',\n",
       " 'isobutanol(8g/L)',\n",
       " 'm-tartrate()',\n",
       " 'monomethyl_succinate()',\n",
       " 'paraquat',\n",
       " 'pvdJ KO P. aeruginosa was first grown in the medium, and an increasing % of this spent\" medium was mixed with fresh medium to grow the E. coli\"',\n",
       " 'Î”gnd',\n",
       " 'Î”pgi',\n",
       " 'Î”pgi E. coli K-12 MG1655',\n",
       " 'Î”pts',\n",
       " 'Î”sdh',\n",
       " 'Î”tpi',\n",
       " 'Î”tpi E. coli K-12 MG1655'}"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_condition_set = set()\n",
    "for cond_cat in METADATA_CATEGORIES_FOR_ASSOCIATIONS:\n",
    "    unique_condition_set |= set(exp_target_cond_df[cond_cat].unique())\n",
    "unique_condition_set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "63b50661-02e3-4506-926d-c002a3134715",
   "metadata": {},
   "outputs": [],
   "source": [
    "FEAT_ANNOT = \"genetic features\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "84e66b13-6f5d-4f4d-aac2-a3c9048383be",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 7670/7670 [02:33<00:00, 49.85it/s] \n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "123432"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>feature</th>\n",
       "      <th>condition</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>nagA</td>\n",
       "      <td>M9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>nagA</td>\n",
       "      <td>glucose(4)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>nagA</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>nagA</td>\n",
       "      <td>KH2PO4(3),Na2HPO4(6.8)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>nagA</td>\n",
       "      <td>MgSO4(0.24)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  feature               condition\n",
       "0    nagA                      M9\n",
       "0    nagA              glucose(4)\n",
       "0    nagA                NH4Cl(1)\n",
       "0    nagA  KH2PO4(3),Na2HPO4(6.8)\n",
       "0    nagA             MgSO4(0.24)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from tqdm import tqdm\n",
    "\n",
    "\n",
    "feat_cond_df = pd.DataFrame()\n",
    "for _, r in tqdm(exp_target_cond_df.iterrows(), total=exp_target_cond_df.shape[0]):                    \n",
    "    for gen_feat_d in r[FEAT_ANNOT]:\n",
    "        for cond_col in METADATA_CATEGORIES_FOR_ASSOCIATIONS:\n",
    "            df = pd.DataFrame.from_dict([{\"feature\": gen_feat_d[\"name\"], \"condition\": r[cond_col]}])\n",
    "            feat_cond_df = pd.concat([feat_cond_df, df])\n",
    "            \n",
    "\n",
    "display(len(feat_cond_df), feat_cond_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ca9a3962-89cc-48e5-ab13-9cb0a56c02a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "from aledbmutil.params import ASSOC_ALPHA, MULTI_HYP_CORR_METHOD\n",
    "\n",
    "\n",
    "from statsmodels.stats import multitest\n",
    "import scipy.stats\n",
    "\n",
    "\n",
    "def get_contingency_table(count_mat, row_name, col_name):\n",
    "    row_col_count = count_mat.loc[row_name, col_name]\n",
    "    row_not_col_sum_count = count_mat.loc[row_name].sum() - row_col_count\n",
    "    not_row_col_sum_count = count_mat.T.loc[col_name].sum() - row_col_count\n",
    "    all_mat_sum_count = count_mat.sum(axis=1).sum()\n",
    "    not_row_not_col_sum_count = all_mat_sum_count - \\\n",
    "        row_col_count - row_not_col_sum_count - not_row_col_sum_count\n",
    "    contingency_table = [\n",
    "        [row_col_count, row_not_col_sum_count],\n",
    "        [not_row_col_sum_count, not_row_not_col_sum_count]\n",
    "    ]\n",
    "    return contingency_table\n",
    "\n",
    "\n",
    "test_df = pd.DataFrame([[1, 2, 3],\n",
    "                        [4, 5, 6]],\n",
    "                       index=[\"y1\", \"y2\"],\n",
    "                       columns=[\"x1\", \"x2\", \"x3\"])\n",
    "contingency_table = get_contingency_table(test_df, \"y2\", \"x3\")\n",
    "expected_contingency_table = [[6, 9], [3, 3]]\n",
    "assert(contingency_table == expected_contingency_table)\n",
    "\n",
    "\n",
    "def get_multiple_hypothesis_correction(pval_df):\n",
    "    # Dataframe seemed to be designed to iterate by column more easily than row (see df.iteritems())\n",
    "    # Have to combine all p-value columns because multitest.multipletests only takes 1D array inputs.\n",
    "    pval_l = []\n",
    "    for col in list(pval_df.columns.values):\n",
    "        pval_l += list(pval_df[col])\n",
    "    corrected_pval_result = multitest.multipletests(\n",
    "        pvals=pval_l,\n",
    "        alpha=ASSOC_ALPHA,\n",
    "        method=MULTI_HYP_CORR_METHOD)\n",
    "\n",
    "    # The following splits the multitest.multipletests 1D array output into the same shape, though transposed.\n",
    "    corrected_pval_l = corrected_pval_result[1]\n",
    "    df_col_len = pval_df.shape[0]\n",
    "    corrected_pval_mat = [corrected_pval_l[i:i+df_col_len]\n",
    "                          for i in range(0, len(corrected_pval_l), df_col_len)]\n",
    "\n",
    "    # building df with new p-values\n",
    "    corrected_pval_df = pval_df.copy()\n",
    "    for outer_l_idx in range(0, len(corrected_pval_df.columns)):\n",
    "        for inner_l_idx in range(0, len(corrected_pval_df)):\n",
    "            corrected_pval = corrected_pval_mat[outer_l_idx][inner_l_idx]\n",
    "            corrected_pval_df.iloc[inner_l_idx, outer_l_idx] = corrected_pval\n",
    "    return corrected_pval_df\n",
    "\n",
    "\n",
    "# def get_enrich_genetic_target_df(cond_col, mut_df):\n",
    "def get_enrich_genetic_target_df(feat_cond_df):\n",
    "    # Creates tables of counts between each mutation and unique condition.\n",
    "    cross_counts_df = pd.crosstab(\n",
    "        feat_cond_df[\"feature\"], feat_cond_df[\"condition\"]\n",
    "    )\n",
    "\n",
    "    # To reuse the same DF indeces without having to remake it.\n",
    "    enrich_odds_ratio_df = cross_counts_df.copy()\n",
    "    enrich_pvals_df = cross_counts_df.copy()\n",
    "\n",
    "    for cond in tqdm(cross_counts_df.columns.values):\n",
    "        for mut_target in cross_counts_df.index:\n",
    "            contingency_table = get_contingency_table(cross_counts_df, mut_target, cond)\n",
    "            odds_ratio, p_val = scipy.stats.fisher_exact(contingency_table, alternative=\"greater\")\n",
    "            enrich_odds_ratio_df.loc[mut_target, cond] = odds_ratio\n",
    "            enrich_pvals_df.loc[mut_target, cond] = p_val\n",
    "\n",
    "    enriched_pvals_df = get_multiple_hypothesis_correction(enrich_pvals_df)\n",
    "\n",
    "    return enrich_odds_ratio_df, enrich_pvals_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "d0812525-5fef-44c8-8740-96380f9fd224",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 81/81 [03:46<00:00,  2.79s/it]\n"
     ]
    }
   ],
   "source": [
    "enrich_odds_ratio_df, enrich_pvals_df = get_enrich_genetic_target_df(feat_cond_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "108b31cc-3a47-4146-971c-0433213496fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "signif_genomic_feat_cond_json = []\n",
    "mut_feat_cond_assoc_json = []\n",
    "for (mut_serial, row) in enrich_pvals_df.iterrows():\n",
    "    for cond in row.index:\n",
    "        p_val = enrich_pvals_df.at[mut_serial, cond]\n",
    "        odds_ratio = enrich_odds_ratio_df.at[mut_serial, cond]\n",
    "        mut_feat_cond_assoc_json.append({\"mutated features\": mut_serial, \"condition\": cond, \"odd ratio\": odds_ratio, \"p value\":p_val})\n",
    "        if odds_ratio > 1 and p_val < ASSOC_ALPHA:\n",
    "            signif_genomic_feat_cond_json.append({\"mutated features\": mut_serial, \"condition\": cond, \"odd ratio\": odds_ratio, \"p value\":p_val})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "da5166e8-77b7-4a1a-a3ec-9cb799311916",
   "metadata": {},
   "outputs": [],
   "source": [
    "assoc_df = pd.DataFrame(mut_feat_cond_assoc_json)\n",
    "assoc_df = assoc_df.set_index(\"mutated features\")\n",
    "assoc_df.to_csv(\"./data/mut_gen_cond_assocs.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "45164f36-63e5-4bbf-8aff-99ec8452bf0f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "f = open(\"./data/signif_genetic_feat_cond_json.pkl\", 'wb')\n",
    "pickle.dump(signif_genomic_feat_cond_json, f)\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "b51c5faf-8d73-48e3-bfcd-bfe0be2e4ba1",
   "metadata": {},
   "outputs": [],
   "source": [
    "signf_assoc_feat_cond_d = {d[\"mutated features\"]:set() for d in signif_genomic_feat_cond_json}\n",
    "for d in signif_genomic_feat_cond_json:\n",
    "    target = d[\"mutated features\"]\n",
    "    condition = d[\"condition\"]\n",
    "    if target in signf_assoc_feat_cond_d.keys():\n",
    "        signf_assoc_feat_cond_d[target].add(condition)\n",
    "    else:\n",
    "        signf_assoc_feat_cond_d[target] = {condition}\n",
    "# signf_assoc_feat_cond_d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "a597b63a-93ad-43e0-8798-55747cf22662",
   "metadata": {},
   "outputs": [],
   "source": [
    "def _get_feat_signf_assoc_cond_set(feat_d, mut_conds):\n",
    "    feat_signf_assoc_cond_set = set()\n",
    "    if feat_d[\"name\"] in signf_assoc_feat_cond_d.keys():\n",
    "        \n",
    "        # Only includes conditions from exp mut comes from, therefore may be subset of signf assoc cond.\n",
    "        for c in signf_assoc_feat_cond_d[feat_d[\"name\"]]:\n",
    "            if c in mut_conds:\n",
    "                feat_signf_assoc_cond_set.add(c)\n",
    "    \n",
    "    return feat_signf_assoc_cond_set\n",
    "\n",
    "\n",
    "def get_feat_signf_assoc_cond_set_json(mut_df_row, feat_annot_name):\n",
    "    feat_json = []\n",
    "    for d in mut_df_row[feat_annot_name]:\n",
    "        feat_d = d.copy()\n",
    "        mut_conds = list(mut_df_row[METADATA_CATEGORIES_FOR_ASSOCIATIONS])\n",
    "        feat_d[\"significantly associated conditions\"] = _get_feat_signf_assoc_cond_set(d, mut_conds)\n",
    "        feat_json.append(feat_d)\n",
    "    return feat_json\n",
    "\n",
    "\n",
    "exp_target_cond_df[FEAT_ANNOT] = exp_target_cond_df.apply(lambda r: get_feat_signf_assoc_cond_set_json(r, FEAT_ANNOT), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "c8a316f1-4dda-431b-b83d-29461a053909",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(7670, 32)"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exp_target_cond_df.to_pickle(\"./out_assoc_gen_feats.pkl\")\n",
    "exp_target_cond_df.shape"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  },
  "vscode": {
   "interpreter": {
    "hash": "7ebccac1095285a8f3a7365f16cbaa2bb02a7e2674096928d95b1b0e9894f90a"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
