{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "39"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exp</th>\n",
       "      <th>sample</th>\n",
       "      <th>Position</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>Details</th>\n",
       "      <th>Gene</th>\n",
       "      <th>ALEdb source</th>\n",
       "      <th>genetic</th>\n",
       "      <th>clean genes</th>\n",
       "      <th>bnums</th>\n",
       "      <th>genes of interest</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>229085</th>\n",
       "      <td>DDB_MS56_strain_verification</td>\n",
       "      <td>37 1 1 1</td>\n",
       "      <td>710439</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G→T</td>\n",
       "      <td>R70S (CGC→AGC)</td>\n",
       "      <td>fur</td>\n",
       "      <td>private</td>\n",
       "      <td>True</td>\n",
       "      <td>{fur}</td>\n",
       "      <td>{b0683}</td>\n",
       "      <td>{b0683}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>468632</th>\n",
       "      <td>GLU_16321-M9-NC_000913_3gb-stationary-37-Gluco...</td>\n",
       "      <td>13 51 0 1</td>\n",
       "      <td>710595</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G→T</td>\n",
       "      <td>P18T (CCT→ACT)</td>\n",
       "      <td>fur</td>\n",
       "      <td>private</td>\n",
       "      <td>True</td>\n",
       "      <td>{fur}</td>\n",
       "      <td>{b0683}</td>\n",
       "      <td>{b0683}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>468721</th>\n",
       "      <td>GLU_16321-M9-NC_000913_3gb-stationary-37-Gluco...</td>\n",
       "      <td>17 61 0 1</td>\n",
       "      <td>710436</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G→A</td>\n",
       "      <td>H71Y (CAC→TAC)</td>\n",
       "      <td>fur</td>\n",
       "      <td>private</td>\n",
       "      <td>True</td>\n",
       "      <td>{fur}</td>\n",
       "      <td>{b0683}</td>\n",
       "      <td>{b0683}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>468785</th>\n",
       "      <td>GLU_16321-M9-NC_000913_3gb-stationary-37-Gluco...</td>\n",
       "      <td>19 50 0 1</td>\n",
       "      <td>710595</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G→T</td>\n",
       "      <td>P18T (CCT→ACT)</td>\n",
       "      <td>fur</td>\n",
       "      <td>private</td>\n",
       "      <td>True</td>\n",
       "      <td>{fur}</td>\n",
       "      <td>{b0683}</td>\n",
       "      <td>{b0683}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>445643</th>\n",
       "      <td>GLU_4721-M9-NC_000913_3gb-stationary-37-Glucos...</td>\n",
       "      <td>3 44 0 1</td>\n",
       "      <td>710489</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G→C</td>\n",
       "      <td>A53G (GCT→GGT)</td>\n",
       "      <td>fur</td>\n",
       "      <td>private</td>\n",
       "      <td>True</td>\n",
       "      <td>{fur}</td>\n",
       "      <td>{b0683}</td>\n",
       "      <td>{b0683}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                      exp     sample  \\\n",
       "229085                       DDB_MS56_strain_verification   37 1 1 1   \n",
       "468632  GLU_16321-M9-NC_000913_3gb-stationary-37-Gluco...  13 51 0 1   \n",
       "468721  GLU_16321-M9-NC_000913_3gb-stationary-37-Gluco...  17 61 0 1   \n",
       "468785  GLU_16321-M9-NC_000913_3gb-stationary-37-Gluco...  19 50 0 1   \n",
       "445643  GLU_4721-M9-NC_000913_3gb-stationary-37-Glucos...   3 44 0 1   \n",
       "\n",
       "        Position Mutation Type Sequence Change         Details Gene  \\\n",
       "229085    710439           SNP             G→T  R70S (CGC→AGC)  fur   \n",
       "468632    710595           SNP             G→T  P18T (CCT→ACT)  fur   \n",
       "468721    710436           SNP             G→A  H71Y (CAC→TAC)  fur   \n",
       "468785    710595           SNP             G→T  P18T (CCT→ACT)  fur   \n",
       "445643    710489           SNP             G→C  A53G (GCT→GGT)  fur   \n",
       "\n",
       "       ALEdb source  genetic clean genes    bnums genes of interest  \n",
       "229085      private     True       {fur}  {b0683}           {b0683}  \n",
       "468632      private     True       {fur}  {b0683}           {b0683}  \n",
       "468721      private     True       {fur}  {b0683}           {b0683}  \n",
       "468785      private     True       {fur}  {b0683}           {b0683}  \n",
       "445643      private     True       {fur}  {b0683}           {b0683}  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "meta_muts = pd.read_pickle('./targ_mut_df.pkl')\n",
    "display(len(meta_muts), meta_muts.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "meta_muts = meta_muts[meta_muts['Sequence Change'] != \"Δ411 bp\"]  # ignoring large deletion for now"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# !!! Temporary; need to reprocess all ALE experiment data for this project with the now removed ERS project experiments which were being duplicated by the newer ROS_eAMP ALE project/experiment.\n",
    "\n",
    "meta_muts = meta_muts[meta_muts.exp != 'GLU_16321-M9-NC_000913_3gb-stationary-37-Glucose2-NH4Cl1-KH2PO43-MgSO4024-O2']\n",
    "meta_muts = meta_muts[meta_muts.exp != 'GLU_18361-M9-NC_000913_3gb-stationary-37-Glucose2-NH4Cl1-KH2PO43-MgSO4024-O2']\n",
    "meta_muts = meta_muts[meta_muts.exp != 'GLU_4721-M9-NC_000913_3gb-stationary-37-Glucose2-NH4Cl1-KH2PO43-MgSO4024-O2']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, sys\n",
    "module_path = os.path.abspath(os.path.join('..'))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "from aledbmutil.mut import get_mut_size, get_mob_size, is_genetic_mut, is_coding_mut, predict_mutation_effect_on_feature, get_SNP_aa_pos, get_DEL_INS_MOB_aa_start_pos, get_DEL_AA_range\n",
    "\n",
    "\n",
    "meta_muts[\"genetic\"] = meta_muts[\"Details\"].apply(is_genetic_mut)\n",
    "meta_muts[\"coding\"] = meta_muts[\"Details\"].apply(is_coding_mut)\n",
    "meta_muts[\"mutation size\"] = meta_muts.apply(get_mut_size, axis=1)  # get_mut_size currently doesn't return MOB mut sizes. Isn't something currently necessary."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exp</th>\n",
       "      <th>sample</th>\n",
       "      <th>Position</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>Details</th>\n",
       "      <th>Gene</th>\n",
       "      <th>ALEdb source</th>\n",
       "      <th>genetic</th>\n",
       "      <th>clean genes</th>\n",
       "      <th>bnums</th>\n",
       "      <th>genes of interest</th>\n",
       "      <th>coding</th>\n",
       "      <th>mutation size</th>\n",
       "      <th>AA range</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>229085</th>\n",
       "      <td>DDB_MS56_strain_verification</td>\n",
       "      <td>37 1 1 1</td>\n",
       "      <td>710439</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G→T</td>\n",
       "      <td>R70S (CGC→AGC)</td>\n",
       "      <td>fur</td>\n",
       "      <td>private</td>\n",
       "      <td>True</td>\n",
       "      <td>{fur}</td>\n",
       "      <td>{b0683}</td>\n",
       "      <td>{b0683}</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>(70, 70)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159670</th>\n",
       "      <td>GLU_ALE4-M9-NC_000913_3gb-stationary-37-Glucos...</td>\n",
       "      <td>16 28 1 1</td>\n",
       "      <td>710522</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C→T</td>\n",
       "      <td>R42H (CGT→CAT)</td>\n",
       "      <td>fur</td>\n",
       "      <td>private</td>\n",
       "      <td>True</td>\n",
       "      <td>{fur}</td>\n",
       "      <td>{b0683}</td>\n",
       "      <td>{b0683}</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>(42, 42)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>189532</th>\n",
       "      <td>MS56</td>\n",
       "      <td>7 48 0 1</td>\n",
       "      <td>710439</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G→T</td>\n",
       "      <td>R70S (CGC→AGC)</td>\n",
       "      <td>fur</td>\n",
       "      <td>private</td>\n",
       "      <td>True</td>\n",
       "      <td>{fur}</td>\n",
       "      <td>{b0683}</td>\n",
       "      <td>{b0683}</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>(70, 70)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>189761</th>\n",
       "      <td>MS56</td>\n",
       "      <td>9 44 0 1</td>\n",
       "      <td>710439</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G→T</td>\n",
       "      <td>R70S (CGC→AGC)</td>\n",
       "      <td>fur</td>\n",
       "      <td>private</td>\n",
       "      <td>True</td>\n",
       "      <td>{fur}</td>\n",
       "      <td>{b0683}</td>\n",
       "      <td>{b0683}</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>(70, 70)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>188967</th>\n",
       "      <td>MS56</td>\n",
       "      <td>10 42 0 1</td>\n",
       "      <td>710439</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G→T</td>\n",
       "      <td>R70S (CGC→AGC)</td>\n",
       "      <td>fur</td>\n",
       "      <td>private</td>\n",
       "      <td>True</td>\n",
       "      <td>{fur}</td>\n",
       "      <td>{b0683}</td>\n",
       "      <td>{b0683}</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>(70, 70)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                      exp     sample  \\\n",
       "229085                       DDB_MS56_strain_verification   37 1 1 1   \n",
       "159670  GLU_ALE4-M9-NC_000913_3gb-stationary-37-Glucos...  16 28 1 1   \n",
       "189532                                               MS56   7 48 0 1   \n",
       "189761                                               MS56   9 44 0 1   \n",
       "188967                                               MS56  10 42 0 1   \n",
       "\n",
       "        Position Mutation Type Sequence Change         Details Gene  \\\n",
       "229085    710439           SNP             G→T  R70S (CGC→AGC)  fur   \n",
       "159670    710522           SNP             C→T  R42H (CGT→CAT)  fur   \n",
       "189532    710439           SNP             G→T  R70S (CGC→AGC)  fur   \n",
       "189761    710439           SNP             G→T  R70S (CGC→AGC)  fur   \n",
       "188967    710439           SNP             G→T  R70S (CGC→AGC)  fur   \n",
       "\n",
       "       ALEdb source  genetic clean genes    bnums genes of interest  coding  \\\n",
       "229085      private     True       {fur}  {b0683}           {b0683}    True   \n",
       "159670      private     True       {fur}  {b0683}           {b0683}    True   \n",
       "189532      private     True       {fur}  {b0683}           {b0683}    True   \n",
       "189761      private     True       {fur}  {b0683}           {b0683}    True   \n",
       "188967      private     True       {fur}  {b0683}           {b0683}    True   \n",
       "\n",
       "        mutation size  AA range  \n",
       "229085              1  (70, 70)  \n",
       "159670              1  (42, 42)  \n",
       "189532              1  (70, 70)  \n",
       "189761              1  (70, 70)  \n",
       "188967              1  (70, 70)  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from aledbmutil.mut import get_SNP_aa_pos, get_DEL_AA_range\n",
    "\n",
    "def _get_SNP_AA_range(mut_details_str):\n",
    "    SNP_AA_range = ()\n",
    "    AA_pos = get_SNP_aa_pos(mut_details_str)\n",
    "    SNP_AA_range = (AA_pos, AA_pos)\n",
    "    return SNP_AA_range\n",
    "\n",
    "meta_muts[\"AA range\"] = None\n",
    "# meta_muts[\"AA range\"] = meta_muts.apply(\n",
    "#     lambda r: _get_SNP_AA_range(r[\"Details\"]) if ((r[\"Mutation Type\"] == \"SNP\") & (r[\"coding\"]==1)) else r[\"AA range\"],\n",
    "#     axis=1)\n",
    "meta_muts[\"AA range\"] = meta_muts.apply(lambda r:\n",
    "                                        _get_SNP_AA_range(r[\"Details\"]) if ((r[\"Mutation Type\"] == \"SNP\") & (r[\"coding\"]==1)) \\\n",
    "                                        else (get_DEL_AA_range(r[\"Details\"]) if ((r[\"Mutation Type\"] == \"DEL\") & (r[\"coding\"]==1)) \\\n",
    "                                        else r[\"AA range\"]),\n",
    "                                        axis=1)\n",
    "meta_muts.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exp</th>\n",
       "      <th>sample</th>\n",
       "      <th>Position</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>Details</th>\n",
       "      <th>Gene</th>\n",
       "      <th>ALEdb source</th>\n",
       "      <th>genetic</th>\n",
       "      <th>clean genes</th>\n",
       "      <th>bnums</th>\n",
       "      <th>genes of interest</th>\n",
       "      <th>coding</th>\n",
       "      <th>mutation size</th>\n",
       "      <th>AA range</th>\n",
       "      <th>AA ref seq</th>\n",
       "      <th>AA seq change</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>229085</th>\n",
       "      <td>DDB_MS56_strain_verification</td>\n",
       "      <td>37 1 1 1</td>\n",
       "      <td>710439</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G→T</td>\n",
       "      <td>R70S (CGC→AGC)</td>\n",
       "      <td>fur</td>\n",
       "      <td>private</td>\n",
       "      <td>True</td>\n",
       "      <td>{fur}</td>\n",
       "      <td>{b0683}</td>\n",
       "      <td>{b0683}</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>(70, 70)</td>\n",
       "      <td>R</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159670</th>\n",
       "      <td>GLU_ALE4-M9-NC_000913_3gb-stationary-37-Glucos...</td>\n",
       "      <td>16 28 1 1</td>\n",
       "      <td>710522</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C→T</td>\n",
       "      <td>R42H (CGT→CAT)</td>\n",
       "      <td>fur</td>\n",
       "      <td>private</td>\n",
       "      <td>True</td>\n",
       "      <td>{fur}</td>\n",
       "      <td>{b0683}</td>\n",
       "      <td>{b0683}</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>(42, 42)</td>\n",
       "      <td>R</td>\n",
       "      <td>H</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>189532</th>\n",
       "      <td>MS56</td>\n",
       "      <td>7 48 0 1</td>\n",
       "      <td>710439</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G→T</td>\n",
       "      <td>R70S (CGC→AGC)</td>\n",
       "      <td>fur</td>\n",
       "      <td>private</td>\n",
       "      <td>True</td>\n",
       "      <td>{fur}</td>\n",
       "      <td>{b0683}</td>\n",
       "      <td>{b0683}</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>(70, 70)</td>\n",
       "      <td>R</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>189761</th>\n",
       "      <td>MS56</td>\n",
       "      <td>9 44 0 1</td>\n",
       "      <td>710439</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G→T</td>\n",
       "      <td>R70S (CGC→AGC)</td>\n",
       "      <td>fur</td>\n",
       "      <td>private</td>\n",
       "      <td>True</td>\n",
       "      <td>{fur}</td>\n",
       "      <td>{b0683}</td>\n",
       "      <td>{b0683}</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>(70, 70)</td>\n",
       "      <td>R</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>188967</th>\n",
       "      <td>MS56</td>\n",
       "      <td>10 42 0 1</td>\n",
       "      <td>710439</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G→T</td>\n",
       "      <td>R70S (CGC→AGC)</td>\n",
       "      <td>fur</td>\n",
       "      <td>private</td>\n",
       "      <td>True</td>\n",
       "      <td>{fur}</td>\n",
       "      <td>{b0683}</td>\n",
       "      <td>{b0683}</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>(70, 70)</td>\n",
       "      <td>R</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                      exp     sample  \\\n",
       "229085                       DDB_MS56_strain_verification   37 1 1 1   \n",
       "159670  GLU_ALE4-M9-NC_000913_3gb-stationary-37-Glucos...  16 28 1 1   \n",
       "189532                                               MS56   7 48 0 1   \n",
       "189761                                               MS56   9 44 0 1   \n",
       "188967                                               MS56  10 42 0 1   \n",
       "\n",
       "        Position Mutation Type Sequence Change         Details Gene  \\\n",
       "229085    710439           SNP             G→T  R70S (CGC→AGC)  fur   \n",
       "159670    710522           SNP             C→T  R42H (CGT→CAT)  fur   \n",
       "189532    710439           SNP             G→T  R70S (CGC→AGC)  fur   \n",
       "189761    710439           SNP             G→T  R70S (CGC→AGC)  fur   \n",
       "188967    710439           SNP             G→T  R70S (CGC→AGC)  fur   \n",
       "\n",
       "       ALEdb source  genetic clean genes    bnums genes of interest  coding  \\\n",
       "229085      private     True       {fur}  {b0683}           {b0683}    True   \n",
       "159670      private     True       {fur}  {b0683}           {b0683}    True   \n",
       "189532      private     True       {fur}  {b0683}           {b0683}    True   \n",
       "189761      private     True       {fur}  {b0683}           {b0683}    True   \n",
       "188967      private     True       {fur}  {b0683}           {b0683}    True   \n",
       "\n",
       "        mutation size  AA range AA ref seq AA seq change  \n",
       "229085              1  (70, 70)          R             S  \n",
       "159670              1  (42, 42)          R             H  \n",
       "189532              1  (70, 70)          R             S  \n",
       "189761              1  (70, 70)          R             S  \n",
       "188967              1  (70, 70)          R             S  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# currently only tracking seq changes with SNPs for ALEdb mutations for later processing with AA property change trends\n",
    "meta_muts['AA ref seq'] = meta_muts.apply(lambda m: m['Details'][0] if m[\"Mutation Type\"]=='SNP' else '', axis=1)\n",
    "meta_muts['AA seq change'] = meta_muts.apply(lambda m: m['Details'].split(' (')[0][-1] if m[\"Mutation Type\"]=='SNP' else '', axis=1)\n",
    "meta_muts.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "exp_sel_pres_d = {\n",
    "    'ROS_eAMP': \"paraquat\",  # starting strains were also pre-evolved on glucose\n",
    " }\n",
    "\n",
    "meta_muts['selection pressure'] = meta_muts.apply(lambda r: 'unpublished' if r['ALEdb source'] == 'private' else '', axis=1)\n",
    "meta_muts['selection pressure'] = meta_muts.exp.apply(lambda e: exp_sel_pres_d[e] if e in exp_sel_pres_d.keys() else 'unpublished')\n",
    "assert('' not in set(meta_muts['selection pressure']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "33"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AA pos</th>\n",
       "      <th>mutation type</th>\n",
       "      <th>mutation effect</th>\n",
       "      <th>mutation source</th>\n",
       "      <th>AA ref seq</th>\n",
       "      <th>AA seq change</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>70</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>R</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>42</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>R</td>\n",
       "      <td>H</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>70</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>R</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>70</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>R</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>70</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>R</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   AA pos mutation type             mutation effect mutation source  \\\n",
       "0      70           SUB  nonsynonymous substitution     unpublished   \n",
       "1      42           SUB  nonsynonymous substitution     unpublished   \n",
       "2      70           SUB  nonsynonymous substitution     unpublished   \n",
       "3      70           SUB  nonsynonymous substitution     unpublished   \n",
       "4      70           SUB  nonsynonymous substitution     unpublished   \n",
       "\n",
       "  AA ref seq AA seq change  \n",
       "0          R             S  \n",
       "1          R             H  \n",
       "2          R             S  \n",
       "3          R             S  \n",
       "4          R             S  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# TODO: fix warnings with append\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "\n",
    "\n",
    "aa_var_df = pd.DataFrame()\n",
    "\n",
    "from aledbmutil.mut import get_mut_size  #, get_DEL_INS_MOB_aa_start_pos\n",
    "\n",
    "\n",
    "for mt, mut_type_mut_df in meta_muts.groupby(\"Mutation Type\"):\n",
    "    assert mt != \"MOB\", 'Currently not handilng MOB. The below INS logic might be able to handle this.'\n",
    "    if mt == \"SNP\":\n",
    "        for _, m in mut_type_mut_df.iterrows():\n",
    "            mut_eff = \"\"\n",
    "            details_str = m[\"Details\"]\n",
    "            details_str = details_str[0:details_str.find(\" (\")]\n",
    "            \n",
    "            if details_str[-1] == '*':  # Could have them show the effected downstream positions like with ACS SynBio paper. \n",
    "                mut_eff = \"truncation\"\n",
    "            elif details_str[0] == details_str[-1]:\n",
    "                mut_eff = \"synonymous substitution\"\n",
    "            elif details_str[0] != details_str[-1]:\n",
    "                mut_eff = \"nonsynonymous substitution\"\n",
    "            else:\n",
    "                mut_eff = \"unknown\"\n",
    "            \n",
    "            aa_pos = int(details_str[1:-1])\n",
    "            aa_var_df= aa_var_df.append({\n",
    "                \"AA pos\": aa_pos,\n",
    "                \"mutation type\": \"SUB\",\n",
    "                \"mutation effect\": mut_eff,\n",
    "                \"mutation source\": m['selection pressure'],\n",
    "                'AA ref seq': m['AA ref seq'],\n",
    "                'AA seq change': m['AA seq change'],\n",
    "            }, ignore_index=True)\n",
    "    if (mt == 'DEL'):  # Unravel since downstream analysis just looks at individual mutated AAs no matter if they come from the same allele.\n",
    "        for _, m in mut_type_mut_df.iterrows():\n",
    "            mut_eff = \"deletion\"  # currently assuming only insertions remain\n",
    "            mut_size = m['mutation size']\n",
    "            if mut_size % 3 != 0:\n",
    "                assert mut_size % 3 != 0, 'not dealing with frameshifts yet' # Could have them show the effected downstream positions like with ACS SynBio paper.\n",
    "            else:\n",
    "                for aa_pos in range(m['AA range'][0], m['AA range'][1]+1):\n",
    "                    aa_var_df= aa_var_df.append({\n",
    "                        \"AA pos\": aa_pos,\n",
    "                        \"mutation type\": mt,\n",
    "                        \"mutation effect\": mut_eff,\n",
    "                        \"mutation source\": m['selection pressure'],\n",
    "                        'AA ref seq': '',\n",
    "                        'AA seq change': '',\n",
    "                    },ignore_index=True)\n",
    "    # if (mt == 'INS'):\n",
    "    #     for _, m in mut_type_mut_df.iterrows():\n",
    "    #         mut_eff = \"insertion\"  # currently assuming only insertions remain\n",
    "    #         mut_size = get_mut_size(m)\n",
    "    #         if mut_size % 3 != 0:\n",
    "    #             mut_eff = \"frameshift\"\n",
    "    #         aa_var_df= aa_var_df.append({\n",
    "    #             \"AA pos\": get_DEL_INS_MOB_aa_start_pos(m[\"Details\"]),\n",
    "    #             \"mutation type\": mt,\n",
    "    #             \"mutation effect\": mut_eff,\n",
    "    #             \"mutation source\": m['selection pressure'],\n",
    "    #             'AA ref seq': '',\n",
    "    #             'AA seq change': '',\n",
    "    #         },ignore_index=True)\n",
    "    if (mt == 'SUB'):  # Unravel since downstream analysis just looks at individual mutated AAs no matter if they come from the same allele.\n",
    "        for _, m in mut_type_mut_df.iterrows():\n",
    "            for aa_pos in range(m['AA range'][0], m['AA range'][1]+1):\n",
    "                aa_var_df= aa_var_df.append({\n",
    "                    \"AA pos\": aa_pos,\n",
    "                    \"mutation type\": mt,\n",
    "                    \"mutation effect\": \"nonsynonymous substitution\",\n",
    "                    \"mutation source\": m['selection pressure'],\n",
    "                    'AA ref seq': '',\n",
    "                    'AA seq change': '',\n",
    "                },ignore_index=True)\n",
    "\n",
    "\n",
    "display(len(aa_var_df), aa_var_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AA pos</th>\n",
       "      <th>mutation type</th>\n",
       "      <th>mutation effect</th>\n",
       "      <th>mutation source</th>\n",
       "      <th>AA ref seq</th>\n",
       "      <th>AA seq change</th>\n",
       "      <th>mutation count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>7</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>A</td>\n",
       "      <td>P</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>14</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>K</td>\n",
       "      <td>N</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>14</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>K</td>\n",
       "      <td>Q</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>18</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>paraquat</td>\n",
       "      <td>P</td>\n",
       "      <td>T</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>18</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>P</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   AA pos mutation type             mutation effect mutation source  \\\n",
       "0       7           SUB  nonsynonymous substitution     unpublished   \n",
       "1      14           SUB  nonsynonymous substitution     unpublished   \n",
       "2      14           SUB  nonsynonymous substitution     unpublished   \n",
       "3      18           SUB  nonsynonymous substitution        paraquat   \n",
       "4      18           SUB  nonsynonymous substitution     unpublished   \n",
       "\n",
       "  AA ref seq AA seq change  mutation count  \n",
       "0          A             P               1  \n",
       "1          K             N               1  \n",
       "2          K             Q               1  \n",
       "3          P             T               3  \n",
       "4          P             T               1  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aa_var_cnt_df = pd.DataFrame()\n",
    "gcols = ['AA pos', 'mutation type', 'mutation effect', 'mutation source', 'AA ref seq', 'AA seq change']\n",
    "for g, gdf in aa_var_df.groupby(gcols):  # Essentially just finding duplicates\n",
    "    d = dict(zip(gcols, g))\n",
    "    d['mutation count'] = len(gdf)\n",
    "    aa_var_cnt_df = aa_var_cnt_df.append(d, ignore_index=True)\n",
    "aa_var_cnt_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AA pos</th>\n",
       "      <th>mutation type</th>\n",
       "      <th>mutation effect</th>\n",
       "      <th>mutation source</th>\n",
       "      <th>AA ref seq</th>\n",
       "      <th>AA seq change</th>\n",
       "      <th>mutation count</th>\n",
       "      <th>mutation effect color</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>7</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>A</td>\n",
       "      <td>P</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>14</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>K</td>\n",
       "      <td>N</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>14</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>K</td>\n",
       "      <td>Q</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>18</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>paraquat</td>\n",
       "      <td>P</td>\n",
       "      <td>T</td>\n",
       "      <td>3</td>\n",
       "      <td>#DD8452</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>18</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>P</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   AA pos mutation type             mutation effect mutation source  \\\n",
       "0       7           SUB  nonsynonymous substitution     unpublished   \n",
       "1      14           SUB  nonsynonymous substitution     unpublished   \n",
       "2      14           SUB  nonsynonymous substitution     unpublished   \n",
       "3      18           SUB  nonsynonymous substitution        paraquat   \n",
       "4      18           SUB  nonsynonymous substitution     unpublished   \n",
       "\n",
       "  AA ref seq AA seq change  mutation count mutation effect color  \n",
       "0          A             P               1               #DD8452  \n",
       "1          K             N               1               #DD8452  \n",
       "2          K             Q               1               #DD8452  \n",
       "3          P             T               3               #DD8452  \n",
       "4          P             T               1               #DD8452  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os, sys\n",
    "module_path = os.path.abspath(os.path.join('..'))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "import const\n",
    "\n",
    "\n",
    "aa_var_cnt_df['mutation effect color'] = aa_var_cnt_df['mutation effect'].apply(lambda e: const.MUTATION_TYPE_COLOR_PALETTE[e])\n",
    "aa_var_cnt_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AA pos</th>\n",
       "      <th>mutation type</th>\n",
       "      <th>mutation effect</th>\n",
       "      <th>mutation source</th>\n",
       "      <th>AA ref seq</th>\n",
       "      <th>AA seq change</th>\n",
       "      <th>mutation count</th>\n",
       "      <th>mutation effect color</th>\n",
       "      <th>selection pressure color</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>7</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>A</td>\n",
       "      <td>P</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#AD8BC9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>14</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>K</td>\n",
       "      <td>N</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#AD8BC9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>14</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>K</td>\n",
       "      <td>Q</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#AD8BC9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>18</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>paraquat</td>\n",
       "      <td>P</td>\n",
       "      <td>T</td>\n",
       "      <td>3</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#6DCCDA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>18</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>P</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#AD8BC9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>23</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>L</td>\n",
       "      <td>Q</td>\n",
       "      <td>2</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#AD8BC9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>42</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>paraquat</td>\n",
       "      <td>R</td>\n",
       "      <td>H</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#6DCCDA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>42</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>R</td>\n",
       "      <td>H</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#AD8BC9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>53</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>paraquat</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#6DCCDA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>61</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>Q</td>\n",
       "      <td>E</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#AD8BC9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>70</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>R</td>\n",
       "      <td>S</td>\n",
       "      <td>13</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#AD8BC9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>71</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>paraquat</td>\n",
       "      <td>H</td>\n",
       "      <td>Y</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#6DCCDA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>73</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>F</td>\n",
       "      <td>I</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#AD8BC9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>92</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>I</td>\n",
       "      <td>N</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#AD8BC9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>102</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>F</td>\n",
       "      <td>L</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#AD8BC9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>104</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>D</td>\n",
       "      <td>V</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#AD8BC9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>128</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>Y</td>\n",
       "      <td>F</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#AD8BC9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>131</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#AD8BC9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    AA pos mutation type             mutation effect mutation source  \\\n",
       "0        7           SUB  nonsynonymous substitution     unpublished   \n",
       "1       14           SUB  nonsynonymous substitution     unpublished   \n",
       "2       14           SUB  nonsynonymous substitution     unpublished   \n",
       "3       18           SUB  nonsynonymous substitution        paraquat   \n",
       "4       18           SUB  nonsynonymous substitution     unpublished   \n",
       "5       23           SUB  nonsynonymous substitution     unpublished   \n",
       "6       42           SUB  nonsynonymous substitution        paraquat   \n",
       "7       42           SUB  nonsynonymous substitution     unpublished   \n",
       "8       53           SUB  nonsynonymous substitution        paraquat   \n",
       "9       61           SUB  nonsynonymous substitution     unpublished   \n",
       "10      70           SUB  nonsynonymous substitution     unpublished   \n",
       "11      71           SUB  nonsynonymous substitution        paraquat   \n",
       "12      73           SUB  nonsynonymous substitution     unpublished   \n",
       "13      92           SUB  nonsynonymous substitution     unpublished   \n",
       "14     102           SUB  nonsynonymous substitution     unpublished   \n",
       "15     104           SUB  nonsynonymous substitution     unpublished   \n",
       "16     128           SUB  nonsynonymous substitution     unpublished   \n",
       "17     131           SUB  nonsynonymous substitution     unpublished   \n",
       "\n",
       "   AA ref seq AA seq change  mutation count mutation effect color  \\\n",
       "0           A             P               1               #DD8452   \n",
       "1           K             N               1               #DD8452   \n",
       "2           K             Q               1               #DD8452   \n",
       "3           P             T               3               #DD8452   \n",
       "4           P             T               1               #DD8452   \n",
       "5           L             Q               2               #DD8452   \n",
       "6           R             H               1               #DD8452   \n",
       "7           R             H               1               #DD8452   \n",
       "8           A             G               1               #DD8452   \n",
       "9           Q             E               1               #DD8452   \n",
       "10          R             S              13               #DD8452   \n",
       "11          H             Y               1               #DD8452   \n",
       "12          F             I               1               #DD8452   \n",
       "13          I             N               1               #DD8452   \n",
       "14          F             L               1               #DD8452   \n",
       "15          D             V               1               #DD8452   \n",
       "16          Y             F               1               #DD8452   \n",
       "17          G             A               1               #DD8452   \n",
       "\n",
       "   selection pressure color  \n",
       "0                   #AD8BC9  \n",
       "1                   #AD8BC9  \n",
       "2                   #AD8BC9  \n",
       "3                   #6DCCDA  \n",
       "4                   #AD8BC9  \n",
       "5                   #AD8BC9  \n",
       "6                   #6DCCDA  \n",
       "7                   #AD8BC9  \n",
       "8                   #6DCCDA  \n",
       "9                   #AD8BC9  \n",
       "10                  #AD8BC9  \n",
       "11                  #6DCCDA  \n",
       "12                  #AD8BC9  \n",
       "13                  #AD8BC9  \n",
       "14                  #AD8BC9  \n",
       "15                  #AD8BC9  \n",
       "16                  #AD8BC9  \n",
       "17                  #AD8BC9  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sel_pres_clr_d = {\n",
    "    # 'FeSO4': '#729ECE',\n",
    "    # 'Homo sapiens tpi': '#FF9E4A',\n",
    "    # 'Pyrobaculum aerophilum pgi': '#67BF5C',\n",
    "    # 'Pyrobaculum aerophilum tpi': '#ED665D',\n",
    "    # 'Vibrio cholerae tpi': '#AD8BC9',\n",
    "    # 'isobutanol': '#ED97CA',\n",
    "    'paraquat': '#6DCCDA',\n",
    "    'unpublished': '#AD8BC9'\n",
    " }\n",
    "\n",
    "aa_var_cnt_df['selection pressure color'] = aa_var_cnt_df['mutation source'].apply(lambda e: sel_pres_clr_d[e])\n",
    "aa_var_cnt_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AA pos</th>\n",
       "      <th>mutation type</th>\n",
       "      <th>mutation effect</th>\n",
       "      <th>mutation source</th>\n",
       "      <th>AA ref seq</th>\n",
       "      <th>AA seq change</th>\n",
       "      <th>mutation count</th>\n",
       "      <th>mutation effect color</th>\n",
       "      <th>selection pressure color</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>7</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>A</td>\n",
       "      <td>P</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#AD8BC9</td>\n",
       "      <td>SUB A7P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>14</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>K</td>\n",
       "      <td>N</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#AD8BC9</td>\n",
       "      <td>SUB K14N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>14</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>K</td>\n",
       "      <td>Q</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#AD8BC9</td>\n",
       "      <td>SUB K14Q</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>18</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>paraquat</td>\n",
       "      <td>P</td>\n",
       "      <td>T</td>\n",
       "      <td>3</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#6DCCDA</td>\n",
       "      <td>SUB P18T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>18</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>P</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#AD8BC9</td>\n",
       "      <td>SUB P18T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>23</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>L</td>\n",
       "      <td>Q</td>\n",
       "      <td>2</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#AD8BC9</td>\n",
       "      <td>SUB L23Q</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>42</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>paraquat</td>\n",
       "      <td>R</td>\n",
       "      <td>H</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#6DCCDA</td>\n",
       "      <td>SUB R42H</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>42</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>R</td>\n",
       "      <td>H</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#AD8BC9</td>\n",
       "      <td>SUB R42H</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>53</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>paraquat</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#6DCCDA</td>\n",
       "      <td>SUB A53G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>61</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>Q</td>\n",
       "      <td>E</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#AD8BC9</td>\n",
       "      <td>SUB Q61E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>70</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>R</td>\n",
       "      <td>S</td>\n",
       "      <td>13</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#AD8BC9</td>\n",
       "      <td>SUB R70S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>71</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>paraquat</td>\n",
       "      <td>H</td>\n",
       "      <td>Y</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#6DCCDA</td>\n",
       "      <td>SUB H71Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>73</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>F</td>\n",
       "      <td>I</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#AD8BC9</td>\n",
       "      <td>SUB F73I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>92</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>I</td>\n",
       "      <td>N</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#AD8BC9</td>\n",
       "      <td>SUB I92N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>102</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>F</td>\n",
       "      <td>L</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#AD8BC9</td>\n",
       "      <td>SUB F102L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>104</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>D</td>\n",
       "      <td>V</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#AD8BC9</td>\n",
       "      <td>SUB D104V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>128</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>Y</td>\n",
       "      <td>F</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#AD8BC9</td>\n",
       "      <td>SUB Y128F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>131</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#AD8BC9</td>\n",
       "      <td>SUB G131A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    AA pos mutation type             mutation effect mutation source  \\\n",
       "0        7           SUB  nonsynonymous substitution     unpublished   \n",
       "1       14           SUB  nonsynonymous substitution     unpublished   \n",
       "2       14           SUB  nonsynonymous substitution     unpublished   \n",
       "3       18           SUB  nonsynonymous substitution        paraquat   \n",
       "4       18           SUB  nonsynonymous substitution     unpublished   \n",
       "5       23           SUB  nonsynonymous substitution     unpublished   \n",
       "6       42           SUB  nonsynonymous substitution        paraquat   \n",
       "7       42           SUB  nonsynonymous substitution     unpublished   \n",
       "8       53           SUB  nonsynonymous substitution        paraquat   \n",
       "9       61           SUB  nonsynonymous substitution     unpublished   \n",
       "10      70           SUB  nonsynonymous substitution     unpublished   \n",
       "11      71           SUB  nonsynonymous substitution        paraquat   \n",
       "12      73           SUB  nonsynonymous substitution     unpublished   \n",
       "13      92           SUB  nonsynonymous substitution     unpublished   \n",
       "14     102           SUB  nonsynonymous substitution     unpublished   \n",
       "15     104           SUB  nonsynonymous substitution     unpublished   \n",
       "16     128           SUB  nonsynonymous substitution     unpublished   \n",
       "17     131           SUB  nonsynonymous substitution     unpublished   \n",
       "\n",
       "   AA ref seq AA seq change  mutation count mutation effect color  \\\n",
       "0           A             P               1               #DD8452   \n",
       "1           K             N               1               #DD8452   \n",
       "2           K             Q               1               #DD8452   \n",
       "3           P             T               3               #DD8452   \n",
       "4           P             T               1               #DD8452   \n",
       "5           L             Q               2               #DD8452   \n",
       "6           R             H               1               #DD8452   \n",
       "7           R             H               1               #DD8452   \n",
       "8           A             G               1               #DD8452   \n",
       "9           Q             E               1               #DD8452   \n",
       "10          R             S              13               #DD8452   \n",
       "11          H             Y               1               #DD8452   \n",
       "12          F             I               1               #DD8452   \n",
       "13          I             N               1               #DD8452   \n",
       "14          F             L               1               #DD8452   \n",
       "15          D             V               1               #DD8452   \n",
       "16          Y             F               1               #DD8452   \n",
       "17          G             A               1               #DD8452   \n",
       "\n",
       "   selection pressure color       name  \n",
       "0                   #AD8BC9    SUB A7P  \n",
       "1                   #AD8BC9   SUB K14N  \n",
       "2                   #AD8BC9   SUB K14Q  \n",
       "3                   #6DCCDA   SUB P18T  \n",
       "4                   #AD8BC9   SUB P18T  \n",
       "5                   #AD8BC9   SUB L23Q  \n",
       "6                   #6DCCDA   SUB R42H  \n",
       "7                   #AD8BC9   SUB R42H  \n",
       "8                   #6DCCDA   SUB A53G  \n",
       "9                   #AD8BC9   SUB Q61E  \n",
       "10                  #AD8BC9   SUB R70S  \n",
       "11                  #6DCCDA   SUB H71Y  \n",
       "12                  #AD8BC9   SUB F73I  \n",
       "13                  #AD8BC9   SUB I92N  \n",
       "14                  #AD8BC9  SUB F102L  \n",
       "15                  #AD8BC9  SUB D104V  \n",
       "16                  #AD8BC9  SUB Y128F  \n",
       "17                  #AD8BC9  SUB G131A  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def _get_variant_name(variant):\n",
    "    var_name = ''\n",
    "    if variant['mutation type'] == 'SUB':\n",
    "        var_name = variant['mutation type'] + \" \" + variant['AA ref seq'] + str(variant['AA pos']) + variant['AA seq change']\n",
    "    return var_name\n",
    "\n",
    "\n",
    "aa_var_cnt_df['name'] = aa_var_cnt_df.apply(lambda r: _get_variant_name(r), axis=1)\n",
    "aa_var_cnt_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "aa_var_cnt_df.to_csv('fur_ale_muts.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>fill</th>\n",
       "      <th>edge</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>unpublished</td>\n",
       "      <td>#AD8BC9</td>\n",
       "      <td>white</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>paraquat</td>\n",
       "      <td>#6DCCDA</td>\n",
       "      <td>white</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>white</td>\n",
       "      <td>#DD8452</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         name     fill     edge\n",
       "0                 unpublished  #AD8BC9    white\n",
       "1                    paraquat  #6DCCDA    white\n",
       "2  nonsynonymous substitution    white  #DD8452"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "mut_needle_plot_legend_df = pd.DataFrame()\n",
    "mut_needle_plot_legend_df = aa_var_cnt_df.copy()\n",
    "mut_needle_plot_legend_df = mut_needle_plot_legend_df[['mutation source', 'selection pressure color', ]]\n",
    "mut_needle_plot_legend_df.rename(columns={'mutation source': 'name', 'selection pressure color': 'fill'}, inplace=True)\n",
    "mut_needle_plot_legend_df['edge'] = 'white'\n",
    "mut_needle_plot_legend_df.drop_duplicates(inplace=True)\n",
    "\n",
    "df = aa_var_cnt_df.copy()\n",
    "df = df[['mutation effect', 'mutation effect color']]\n",
    "df.rename(columns={'mutation effect': 'name', 'mutation effect color': 'edge'}, inplace=True)\n",
    "df.drop_duplicates(inplace=True)\n",
    "df['fill'] = 'white'\n",
    "mut_needle_plot_legend_df = mut_needle_plot_legend_df.append(df, ignore_index=True)\n",
    "display(mut_needle_plot_legend_df)\n",
    "\n",
    "mut_needle_plot_legend_df.to_csv('fur_ale_muts_legend.csv', index=False)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Creating functional feature annotation output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>range</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Dimerization region</td>\n",
       "      <td>85</td>\n",
       "      <td>148</td>\n",
       "      <td>(85, 64)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chain</td>\n",
       "      <td>2</td>\n",
       "      <td>148</td>\n",
       "      <td>(2, 147)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Fe binding site</td>\n",
       "      <td>125</td>\n",
       "      <td>125</td>\n",
       "      <td>(125, 1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Fe binding site</td>\n",
       "      <td>108</td>\n",
       "      <td>108</td>\n",
       "      <td>(108, 1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Zn binding site</td>\n",
       "      <td>101</td>\n",
       "      <td>101</td>\n",
       "      <td>(101, 1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Zn binding site</td>\n",
       "      <td>96</td>\n",
       "      <td>96</td>\n",
       "      <td>(96, 1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Zn binding site</td>\n",
       "      <td>93</td>\n",
       "      <td>93</td>\n",
       "      <td>(93, 1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Zn binding site</td>\n",
       "      <td>90</td>\n",
       "      <td>90</td>\n",
       "      <td>(90, 1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Fe binding site</td>\n",
       "      <td>89</td>\n",
       "      <td>89</td>\n",
       "      <td>(89, 1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Fe binding site</td>\n",
       "      <td>87</td>\n",
       "      <td>87</td>\n",
       "      <td>(87, 1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>DNA binding region</td>\n",
       "      <td>2</td>\n",
       "      <td>84</td>\n",
       "      <td>(2, 83)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Zn binding site</td>\n",
       "      <td>81</td>\n",
       "      <td>81</td>\n",
       "      <td>(81, 1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>subunit interface</td>\n",
       "      <td>78</td>\n",
       "      <td>78</td>\n",
       "      <td>(78, 1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>subunit interface</td>\n",
       "      <td>72</td>\n",
       "      <td>76</td>\n",
       "      <td>(72, 5)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>subunit interface</td>\n",
       "      <td>70</td>\n",
       "      <td>70</td>\n",
       "      <td>(70, 1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>subunit interface</td>\n",
       "      <td>63</td>\n",
       "      <td>64</td>\n",
       "      <td>(63, 2)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>subunit interface</td>\n",
       "      <td>60</td>\n",
       "      <td>60</td>\n",
       "      <td>(60, 1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>subunit interface</td>\n",
       "      <td>56</td>\n",
       "      <td>57</td>\n",
       "      <td>(56, 2)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>subunit interface</td>\n",
       "      <td>52</td>\n",
       "      <td>53</td>\n",
       "      <td>(52, 2)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Zn binding site</td>\n",
       "      <td>33</td>\n",
       "      <td>33</td>\n",
       "      <td>(33, 1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>subunit interface</td>\n",
       "      <td>92</td>\n",
       "      <td>94</td>\n",
       "      <td>(92, 3)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>subunit interface</td>\n",
       "      <td>110</td>\n",
       "      <td>111</td>\n",
       "      <td>(110, 2)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>subunit interface</td>\n",
       "      <td>120</td>\n",
       "      <td>122</td>\n",
       "      <td>(120, 3)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>subunit interface</td>\n",
       "      <td>124</td>\n",
       "      <td>132</td>\n",
       "      <td>(124, 9)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>putative DNA binding helix</td>\n",
       "      <td>51</td>\n",
       "      <td>65</td>\n",
       "      <td>(51, 15)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Zn binding site</td>\n",
       "      <td>133</td>\n",
       "      <td>133</td>\n",
       "      <td>(133, 1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Cd binding site</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>(3, 1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Cd binding site</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "      <td>(30, 1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Cd binding site</td>\n",
       "      <td>33</td>\n",
       "      <td>33</td>\n",
       "      <td>(33, 1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Cd binding site</td>\n",
       "      <td>49</td>\n",
       "      <td>49</td>\n",
       "      <td>(49, 1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>Cd binding site</td>\n",
       "      <td>63</td>\n",
       "      <td>64</td>\n",
       "      <td>(63, 2)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>Cd binding site</td>\n",
       "      <td>71</td>\n",
       "      <td>71</td>\n",
       "      <td>(71, 1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>Cd binding site</td>\n",
       "      <td>74</td>\n",
       "      <td>74</td>\n",
       "      <td>(74, 1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>Cd binding site</td>\n",
       "      <td>81</td>\n",
       "      <td>81</td>\n",
       "      <td>(81, 1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>Cl binding site</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>(10, 1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>Cl binding site</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "      <td>(30, 1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>Cl binding site</td>\n",
       "      <td>32</td>\n",
       "      <td>33</td>\n",
       "      <td>(32, 2)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>Cl binding site</td>\n",
       "      <td>51</td>\n",
       "      <td>54</td>\n",
       "      <td>(51, 4)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>Cl binding site</td>\n",
       "      <td>71</td>\n",
       "      <td>72</td>\n",
       "      <td>(71, 2)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>Cl binding site</td>\n",
       "      <td>74</td>\n",
       "      <td>74</td>\n",
       "      <td>(74, 1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>Cl binding site</td>\n",
       "      <td>81</td>\n",
       "      <td>81</td>\n",
       "      <td>(81, 1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>glycerol binding site</td>\n",
       "      <td>37</td>\n",
       "      <td>38</td>\n",
       "      <td>(37, 2)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>glycerol binding site</td>\n",
       "      <td>41</td>\n",
       "      <td>41</td>\n",
       "      <td>(41, 1)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       feature start  end     range\n",
       "0          Dimerization region    85  148  (85, 64)\n",
       "1                        Chain     2  148  (2, 147)\n",
       "2              Fe binding site   125  125  (125, 1)\n",
       "3              Fe binding site   108  108  (108, 1)\n",
       "4              Zn binding site   101  101  (101, 1)\n",
       "5              Zn binding site    96   96   (96, 1)\n",
       "6              Zn binding site    93   93   (93, 1)\n",
       "7              Zn binding site    90   90   (90, 1)\n",
       "8              Fe binding site    89   89   (89, 1)\n",
       "9              Fe binding site    87   87   (87, 1)\n",
       "10          DNA binding region     2   84   (2, 83)\n",
       "11             Zn binding site    81   81   (81, 1)\n",
       "12           subunit interface    78   78   (78, 1)\n",
       "13           subunit interface    72   76   (72, 5)\n",
       "14           subunit interface    70   70   (70, 1)\n",
       "15           subunit interface    63   64   (63, 2)\n",
       "16           subunit interface    60   60   (60, 1)\n",
       "17           subunit interface    56   57   (56, 2)\n",
       "18           subunit interface    52   53   (52, 2)\n",
       "19             Zn binding site    33   33   (33, 1)\n",
       "20           subunit interface    92   94   (92, 3)\n",
       "21           subunit interface   110  111  (110, 2)\n",
       "22           subunit interface   120  122  (120, 3)\n",
       "23           subunit interface   124  132  (124, 9)\n",
       "24  putative DNA binding helix    51   65  (51, 15)\n",
       "25             Zn binding site   133  133  (133, 1)\n",
       "26             Cd binding site     3    3    (3, 1)\n",
       "27             Cd binding site    30   30   (30, 1)\n",
       "28             Cd binding site    33   33   (33, 1)\n",
       "29             Cd binding site    49   49   (49, 1)\n",
       "30             Cd binding site    63   64   (63, 2)\n",
       "31             Cd binding site    71   71   (71, 1)\n",
       "32             Cd binding site    74   74   (74, 1)\n",
       "33             Cd binding site    81   81   (81, 1)\n",
       "34             Cl binding site    10   10   (10, 1)\n",
       "35             Cl binding site    30   30   (30, 1)\n",
       "36             Cl binding site    32   33   (32, 2)\n",
       "37             Cl binding site    51   54   (51, 4)\n",
       "38             Cl binding site    71   72   (71, 2)\n",
       "39             Cl binding site    74   74   (74, 1)\n",
       "40             Cl binding site    81   81   (81, 1)\n",
       "41       glycerol binding site    37   38   (37, 2)\n",
       "42       glycerol binding site    41   41   (41, 1)"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "feats_df = pd.read_pickle('./annot_df.pkl')\n",
    "\n",
    "# filtering specific to fur's features\n",
    "feats_df[\"feature\"] = feats_df[\"feature\"].apply(lambda s: s.replace(\"Fur subunit\", \"subunit\"))\n",
    "feats_df[\"feature\"] = feats_df[\"feature\"].apply(lambda s: s.replace(\"dimer\", \"subunit\"))  # \"dimer interface\" and \"Fur subunit interface\" are the same thing, though coming from different sources\"\n",
    "\n",
    "# filtering out features that aren't useful\n",
    "feats_to_remove = [\n",
    "    'Initiator methionine',\n",
    "    'FERRIC UPTAKE REGULATION PROTEIN',\n",
    "    'ZINC UPTAKE REGULATION PROTEIN',\n",
    "    'Ferric-uptake regulator',\n",
    "    'Ferric-uptake regulator family',\n",
    "    'Winged helix-like DNA-binding domain superfamily/Winged helix DNA-binding domain',\n",
    "    'Winged helix-like DNA-binding domain superfamily',\n",
    "    'Winged helix DNA-binding domain superfamily',\n",
    "    'G3DSA:3.30.1490.190',\n",
    "    'Ferric uptake regulator family',\n",
    "    'Ferric uptake regulator(Fur) and related metalloregulatory proteins; typically iron-dependent, DNA-binding repressors and activators',\n",
    "    'Winged helix DNA-binding domain',\n",
    "    'Ferric-uptake regulator, C-terminal domain',\n",
    "    'Metal binding',\n",
    "    # 'Chain',\n",
    "]\n",
    "feats_df = feats_df[~(feats_df['feature'].isin(feats_to_remove))]\n",
    "feats_df.reset_index(inplace=True, drop=True)\n",
    "feats_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>range</th>\n",
       "      <th>layer</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Dimerization region</td>\n",
       "      <td>85</td>\n",
       "      <td>148</td>\n",
       "      <td>(85, 64)</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chain</td>\n",
       "      <td>2</td>\n",
       "      <td>148</td>\n",
       "      <td>(2, 147)</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Fe binding site</td>\n",
       "      <td>125</td>\n",
       "      <td>125</td>\n",
       "      <td>(125, 1)</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Fe binding site</td>\n",
       "      <td>108</td>\n",
       "      <td>108</td>\n",
       "      <td>(108, 1)</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Zn binding site</td>\n",
       "      <td>101</td>\n",
       "      <td>101</td>\n",
       "      <td>(101, 1)</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               feature start  end     range  layer\n",
       "0  Dimerization region    85  148  (85, 64)      0\n",
       "1                Chain     2  148  (2, 147)      1\n",
       "2      Fe binding site   125  125  (125, 1)      2\n",
       "3      Fe binding site   108  108  (108, 1)      2\n",
       "4      Zn binding site   101  101  (101, 1)      3"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "layer = 0\n",
    "layer_d = dict()\n",
    "for f in feats_df.feature.unique():\n",
    "    layer_d[f] = layer\n",
    "    layer += 1\n",
    "feats_df['layer'] = feats_df['feature'].apply(lambda f: layer_d[f])\n",
    "feats_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>layer</th>\n",
       "      <th>color</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Dimerization region</td>\n",
       "      <td>85</td>\n",
       "      <td>148</td>\n",
       "      <td>0</td>\n",
       "      <td>#4878D0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chain</td>\n",
       "      <td>2</td>\n",
       "      <td>148</td>\n",
       "      <td>1</td>\n",
       "      <td>#EE854A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Fe binding site</td>\n",
       "      <td>125</td>\n",
       "      <td>125</td>\n",
       "      <td>2</td>\n",
       "      <td>#6ACC64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Fe binding site</td>\n",
       "      <td>108</td>\n",
       "      <td>108</td>\n",
       "      <td>2</td>\n",
       "      <td>#6ACC64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Zn binding site</td>\n",
       "      <td>101</td>\n",
       "      <td>101</td>\n",
       "      <td>3</td>\n",
       "      <td>#D65F5F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Zn binding site</td>\n",
       "      <td>96</td>\n",
       "      <td>96</td>\n",
       "      <td>3</td>\n",
       "      <td>#D65F5F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Zn binding site</td>\n",
       "      <td>93</td>\n",
       "      <td>93</td>\n",
       "      <td>3</td>\n",
       "      <td>#D65F5F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Zn binding site</td>\n",
       "      <td>90</td>\n",
       "      <td>90</td>\n",
       "      <td>3</td>\n",
       "      <td>#D65F5F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Fe binding site</td>\n",
       "      <td>89</td>\n",
       "      <td>89</td>\n",
       "      <td>2</td>\n",
       "      <td>#6ACC64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Fe binding site</td>\n",
       "      <td>87</td>\n",
       "      <td>87</td>\n",
       "      <td>2</td>\n",
       "      <td>#6ACC64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>DNA binding region</td>\n",
       "      <td>2</td>\n",
       "      <td>84</td>\n",
       "      <td>4</td>\n",
       "      <td>#956CB4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Zn binding site</td>\n",
       "      <td>81</td>\n",
       "      <td>81</td>\n",
       "      <td>3</td>\n",
       "      <td>#D65F5F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>subunit interface</td>\n",
       "      <td>78</td>\n",
       "      <td>78</td>\n",
       "      <td>5</td>\n",
       "      <td>#8C613C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>subunit interface</td>\n",
       "      <td>72</td>\n",
       "      <td>76</td>\n",
       "      <td>5</td>\n",
       "      <td>#8C613C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>subunit interface</td>\n",
       "      <td>70</td>\n",
       "      <td>70</td>\n",
       "      <td>5</td>\n",
       "      <td>#8C613C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>subunit interface</td>\n",
       "      <td>63</td>\n",
       "      <td>64</td>\n",
       "      <td>5</td>\n",
       "      <td>#8C613C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>subunit interface</td>\n",
       "      <td>60</td>\n",
       "      <td>60</td>\n",
       "      <td>5</td>\n",
       "      <td>#8C613C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>subunit interface</td>\n",
       "      <td>56</td>\n",
       "      <td>57</td>\n",
       "      <td>5</td>\n",
       "      <td>#8C613C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>subunit interface</td>\n",
       "      <td>52</td>\n",
       "      <td>53</td>\n",
       "      <td>5</td>\n",
       "      <td>#8C613C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Zn binding site</td>\n",
       "      <td>33</td>\n",
       "      <td>33</td>\n",
       "      <td>3</td>\n",
       "      <td>#D65F5F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>subunit interface</td>\n",
       "      <td>92</td>\n",
       "      <td>94</td>\n",
       "      <td>5</td>\n",
       "      <td>#8C613C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>subunit interface</td>\n",
       "      <td>110</td>\n",
       "      <td>111</td>\n",
       "      <td>5</td>\n",
       "      <td>#8C613C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>subunit interface</td>\n",
       "      <td>120</td>\n",
       "      <td>122</td>\n",
       "      <td>5</td>\n",
       "      <td>#8C613C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>subunit interface</td>\n",
       "      <td>124</td>\n",
       "      <td>132</td>\n",
       "      <td>5</td>\n",
       "      <td>#8C613C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>putative DNA binding helix</td>\n",
       "      <td>51</td>\n",
       "      <td>65</td>\n",
       "      <td>6</td>\n",
       "      <td>#DC7EC0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Zn binding site</td>\n",
       "      <td>133</td>\n",
       "      <td>133</td>\n",
       "      <td>3</td>\n",
       "      <td>#D65F5F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Cd binding site</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>7</td>\n",
       "      <td>#797979</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Cd binding site</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "      <td>7</td>\n",
       "      <td>#797979</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Cd binding site</td>\n",
       "      <td>33</td>\n",
       "      <td>33</td>\n",
       "      <td>7</td>\n",
       "      <td>#797979</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Cd binding site</td>\n",
       "      <td>49</td>\n",
       "      <td>49</td>\n",
       "      <td>7</td>\n",
       "      <td>#797979</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>Cd binding site</td>\n",
       "      <td>63</td>\n",
       "      <td>64</td>\n",
       "      <td>7</td>\n",
       "      <td>#797979</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>Cd binding site</td>\n",
       "      <td>71</td>\n",
       "      <td>71</td>\n",
       "      <td>7</td>\n",
       "      <td>#797979</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>Cd binding site</td>\n",
       "      <td>74</td>\n",
       "      <td>74</td>\n",
       "      <td>7</td>\n",
       "      <td>#797979</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>Cd binding site</td>\n",
       "      <td>81</td>\n",
       "      <td>81</td>\n",
       "      <td>7</td>\n",
       "      <td>#797979</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>Cl binding site</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>8</td>\n",
       "      <td>#D5BB67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>Cl binding site</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "      <td>8</td>\n",
       "      <td>#D5BB67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>Cl binding site</td>\n",
       "      <td>32</td>\n",
       "      <td>33</td>\n",
       "      <td>8</td>\n",
       "      <td>#D5BB67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>Cl binding site</td>\n",
       "      <td>51</td>\n",
       "      <td>54</td>\n",
       "      <td>8</td>\n",
       "      <td>#D5BB67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>Cl binding site</td>\n",
       "      <td>71</td>\n",
       "      <td>72</td>\n",
       "      <td>8</td>\n",
       "      <td>#D5BB67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>Cl binding site</td>\n",
       "      <td>74</td>\n",
       "      <td>74</td>\n",
       "      <td>8</td>\n",
       "      <td>#D5BB67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>Cl binding site</td>\n",
       "      <td>81</td>\n",
       "      <td>81</td>\n",
       "      <td>8</td>\n",
       "      <td>#D5BB67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>glycerol binding site</td>\n",
       "      <td>37</td>\n",
       "      <td>38</td>\n",
       "      <td>9</td>\n",
       "      <td>#82C6E2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>glycerol binding site</td>\n",
       "      <td>41</td>\n",
       "      <td>41</td>\n",
       "      <td>9</td>\n",
       "      <td>#82C6E2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       feature start  end  layer    color\n",
       "0          Dimerization region    85  148      0  #4878D0\n",
       "1                        Chain     2  148      1  #EE854A\n",
       "2              Fe binding site   125  125      2  #6ACC64\n",
       "3              Fe binding site   108  108      2  #6ACC64\n",
       "4              Zn binding site   101  101      3  #D65F5F\n",
       "5              Zn binding site    96   96      3  #D65F5F\n",
       "6              Zn binding site    93   93      3  #D65F5F\n",
       "7              Zn binding site    90   90      3  #D65F5F\n",
       "8              Fe binding site    89   89      2  #6ACC64\n",
       "9              Fe binding site    87   87      2  #6ACC64\n",
       "10          DNA binding region     2   84      4  #956CB4\n",
       "11             Zn binding site    81   81      3  #D65F5F\n",
       "12           subunit interface    78   78      5  #8C613C\n",
       "13           subunit interface    72   76      5  #8C613C\n",
       "14           subunit interface    70   70      5  #8C613C\n",
       "15           subunit interface    63   64      5  #8C613C\n",
       "16           subunit interface    60   60      5  #8C613C\n",
       "17           subunit interface    56   57      5  #8C613C\n",
       "18           subunit interface    52   53      5  #8C613C\n",
       "19             Zn binding site    33   33      3  #D65F5F\n",
       "20           subunit interface    92   94      5  #8C613C\n",
       "21           subunit interface   110  111      5  #8C613C\n",
       "22           subunit interface   120  122      5  #8C613C\n",
       "23           subunit interface   124  132      5  #8C613C\n",
       "24  putative DNA binding helix    51   65      6  #DC7EC0\n",
       "25             Zn binding site   133  133      3  #D65F5F\n",
       "26             Cd binding site     3    3      7  #797979\n",
       "27             Cd binding site    30   30      7  #797979\n",
       "28             Cd binding site    33   33      7  #797979\n",
       "29             Cd binding site    49   49      7  #797979\n",
       "30             Cd binding site    63   64      7  #797979\n",
       "31             Cd binding site    71   71      7  #797979\n",
       "32             Cd binding site    74   74      7  #797979\n",
       "33             Cd binding site    81   81      7  #797979\n",
       "34             Cl binding site    10   10      8  #D5BB67\n",
       "35             Cl binding site    30   30      8  #D5BB67\n",
       "36             Cl binding site    32   33      8  #D5BB67\n",
       "37             Cl binding site    51   54      8  #D5BB67\n",
       "38             Cl binding site    71   72      8  #D5BB67\n",
       "39             Cl binding site    74   74      8  #D5BB67\n",
       "40             Cl binding site    81   81      8  #D5BB67\n",
       "41       glycerol binding site    37   38      9  #82C6E2\n",
       "42       glycerol binding site    41   41      9  #82C6E2"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "{'Dimerization region': '#4878D0',\n",
       " 'Chain': '#EE854A',\n",
       " 'Fe binding site': '#6ACC64',\n",
       " 'Zn binding site': '#D65F5F',\n",
       " 'DNA binding region': '#956CB4',\n",
       " 'subunit interface': '#8C613C',\n",
       " 'putative DNA binding helix': '#DC7EC0',\n",
       " 'Cd binding site': '#797979',\n",
       " 'Cl binding site': '#D5BB67',\n",
       " 'glycerol binding site': '#82C6E2'}"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "feat_color_d = dict()\n",
    "muted=[\"#4878D0\", \"#EE854A\", \"#6ACC64\", \"#D65F5F\", \"#956CB4\", \"#8C613C\", \"#DC7EC0\", \"#797979\", \"#D5BB67\", \"#82C6E2\"]  # from seaborn\n",
    "feats = list(feats_df.feature.unique())\n",
    "for i in range(0, len(feats_df.feature.unique())):\n",
    "    feat_color_d[feats[i]] = muted[i]\n",
    "feat_color_d\n",
    "\n",
    "feats_df['color'] = feats_df['feature'].apply(lambda f: feat_color_d[f])\n",
    "\n",
    "feats_df.drop(columns=[\"range\"], inplace=True)\n",
    "feats_df.drop_duplicates(inplace=True)\n",
    "\n",
    "display(feats_df, feat_color_d)\n",
    "feats_df.to_csv('./fur_feats.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filtering out some features that are not useful for the NGL visual 3D analysis\n",
    "ngl_feat_df = feats_df.copy()\n",
    "ngl_feat_df = ngl_feat_df[ngl_feat_df['feature'].isin(['subunit interface', 'DNA binding region', 'Dimerization region'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>color</th>\n",
       "      <th>size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DNA binding region</td>\n",
       "      <td>2</td>\n",
       "      <td>84</td>\n",
       "      <td>#956CB4</td>\n",
       "      <td>83</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Dimerization region</td>\n",
       "      <td>85</td>\n",
       "      <td>148</td>\n",
       "      <td>#4878D0</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>subunit interface</td>\n",
       "      <td>124</td>\n",
       "      <td>132</td>\n",
       "      <td>#8C613C</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>subunit interface</td>\n",
       "      <td>72</td>\n",
       "      <td>76</td>\n",
       "      <td>#8C613C</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>subunit interface</td>\n",
       "      <td>92</td>\n",
       "      <td>94</td>\n",
       "      <td>#8C613C</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>subunit interface</td>\n",
       "      <td>120</td>\n",
       "      <td>122</td>\n",
       "      <td>#8C613C</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>subunit interface</td>\n",
       "      <td>63</td>\n",
       "      <td>64</td>\n",
       "      <td>#8C613C</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>subunit interface</td>\n",
       "      <td>56</td>\n",
       "      <td>57</td>\n",
       "      <td>#8C613C</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>subunit interface</td>\n",
       "      <td>52</td>\n",
       "      <td>53</td>\n",
       "      <td>#8C613C</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>subunit interface</td>\n",
       "      <td>110</td>\n",
       "      <td>111</td>\n",
       "      <td>#8C613C</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>subunit interface</td>\n",
       "      <td>78</td>\n",
       "      <td>78</td>\n",
       "      <td>#8C613C</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>subunit interface</td>\n",
       "      <td>70</td>\n",
       "      <td>70</td>\n",
       "      <td>#8C613C</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>subunit interface</td>\n",
       "      <td>60</td>\n",
       "      <td>60</td>\n",
       "      <td>#8C613C</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                feature start  end    color  size\n",
       "0    DNA binding region     2   84  #956CB4    83\n",
       "1   Dimerization region    85  148  #4878D0    64\n",
       "2     subunit interface   124  132  #8C613C     9\n",
       "3     subunit interface    72   76  #8C613C     5\n",
       "4     subunit interface    92   94  #8C613C     3\n",
       "5     subunit interface   120  122  #8C613C     3\n",
       "6     subunit interface    63   64  #8C613C     2\n",
       "7     subunit interface    56   57  #8C613C     2\n",
       "8     subunit interface    52   53  #8C613C     2\n",
       "9     subunit interface   110  111  #8C613C     2\n",
       "10    subunit interface    78   78  #8C613C     1\n",
       "11    subunit interface    70   70  #8C613C     1\n",
       "12    subunit interface    60   60  #8C613C     1"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# for NGL\n",
    "ngl_feat_df[\"size\"] = ngl_feat_df.apply(lambda r: r[\"end\"] - r[\"start\"] + 1, axis=1)\n",
    "ngl_feat_df = ngl_feat_df[[\"feature\", \"start\", \"end\", \"color\", \"size\"]]\n",
    "ngl_feat_df = ngl_feat_df.sort_values(\"size\", ascending=False)\n",
    "ngl_feat_df = ngl_feat_df.reset_index(drop=True)  # json explicitly uses the indexes\n",
    "\n",
    "with open('feats.json', 'w') as outfile:   # for NGL\n",
    "    outfile.write(ngl_feat_df.to_json())\n",
    "ngl_feat_df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "7ebccac1095285a8f3a7365f16cbaa2bb02a7e2674096928d95b1b0e9894f90a"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
