{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "46"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exp</th>\n",
       "      <th>sample</th>\n",
       "      <th>Position</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>Details</th>\n",
       "      <th>Gene</th>\n",
       "      <th>ALEdb source</th>\n",
       "      <th>genetic</th>\n",
       "      <th>clean genes</th>\n",
       "      <th>bnums</th>\n",
       "      <th>genes of interest</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>468518</th>\n",
       "      <td>GLU_16321-M9-NC_000913_3gb-stationary-37-Gluco...</td>\n",
       "      <td>11 20 0 1</td>\n",
       "      <td>3041632</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T→A</td>\n",
       "      <td>V107E (GTG→GAG)</td>\n",
       "      <td>ygfZ</td>\n",
       "      <td>private</td>\n",
       "      <td>True</td>\n",
       "      <td>{ygfZ}</td>\n",
       "      <td>{b2898}</td>\n",
       "      <td>{b2898}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>468601</th>\n",
       "      <td>GLU_16321-M9-NC_000913_3gb-stationary-37-Gluco...</td>\n",
       "      <td>13 25 0 1</td>\n",
       "      <td>3041632</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T→A</td>\n",
       "      <td>V107E (GTG→GAG)</td>\n",
       "      <td>ygfZ</td>\n",
       "      <td>private</td>\n",
       "      <td>True</td>\n",
       "      <td>{ygfZ}</td>\n",
       "      <td>{b2898}</td>\n",
       "      <td>{b2898}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>468720</th>\n",
       "      <td>GLU_16321-M9-NC_000913_3gb-stationary-37-Gluco...</td>\n",
       "      <td>17 23 0 1</td>\n",
       "      <td>3041632</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T→A</td>\n",
       "      <td>V107E (GTG→GAG)</td>\n",
       "      <td>ygfZ</td>\n",
       "      <td>private</td>\n",
       "      <td>True</td>\n",
       "      <td>{ygfZ}</td>\n",
       "      <td>{b2898}</td>\n",
       "      <td>{b2898}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>468758</th>\n",
       "      <td>GLU_16321-M9-NC_000913_3gb-stationary-37-Gluco...</td>\n",
       "      <td>19 21 0 1</td>\n",
       "      <td>3041632</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T→A</td>\n",
       "      <td>V107E (GTG→GAG)</td>\n",
       "      <td>ygfZ</td>\n",
       "      <td>private</td>\n",
       "      <td>True</td>\n",
       "      <td>{ygfZ}</td>\n",
       "      <td>{b2898}</td>\n",
       "      <td>{b2898}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>158708</th>\n",
       "      <td>GLU_18361-M9-NC_000913_3gb-stationary-37-Gluco...</td>\n",
       "      <td>21 21 0 1</td>\n",
       "      <td>3041634</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A→C</td>\n",
       "      <td>T108P (ACC→CCC)</td>\n",
       "      <td>ygfZ</td>\n",
       "      <td>private</td>\n",
       "      <td>True</td>\n",
       "      <td>{ygfZ}</td>\n",
       "      <td>{b2898}</td>\n",
       "      <td>{b2898}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                      exp     sample  \\\n",
       "468518  GLU_16321-M9-NC_000913_3gb-stationary-37-Gluco...  11 20 0 1   \n",
       "468601  GLU_16321-M9-NC_000913_3gb-stationary-37-Gluco...  13 25 0 1   \n",
       "468720  GLU_16321-M9-NC_000913_3gb-stationary-37-Gluco...  17 23 0 1   \n",
       "468758  GLU_16321-M9-NC_000913_3gb-stationary-37-Gluco...  19 21 0 1   \n",
       "158708  GLU_18361-M9-NC_000913_3gb-stationary-37-Gluco...  21 21 0 1   \n",
       "\n",
       "        Position Mutation Type Sequence Change          Details  Gene  \\\n",
       "468518   3041632           SNP             T→A  V107E (GTG→GAG)  ygfZ   \n",
       "468601   3041632           SNP             T→A  V107E (GTG→GAG)  ygfZ   \n",
       "468720   3041632           SNP             T→A  V107E (GTG→GAG)  ygfZ   \n",
       "468758   3041632           SNP             T→A  V107E (GTG→GAG)  ygfZ   \n",
       "158708   3041634           SNP             A→C  T108P (ACC→CCC)  ygfZ   \n",
       "\n",
       "       ALEdb source  genetic clean genes    bnums genes of interest  \n",
       "468518      private     True      {ygfZ}  {b2898}           {b2898}  \n",
       "468601      private     True      {ygfZ}  {b2898}           {b2898}  \n",
       "468720      private     True      {ygfZ}  {b2898}           {b2898}  \n",
       "468758      private     True      {ygfZ}  {b2898}           {b2898}  \n",
       "158708      private     True      {ygfZ}  {b2898}           {b2898}  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "meta_muts = pd.read_pickle('./targ_mut_df.pkl')\n",
    "display(len(meta_muts), meta_muts.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# !!! Temporary; need to reprocess all ALE experiment data for this project with the now removed ERS project experiments which were being duplicated by the newer ROS_eAMP ALE project/experiment.\n",
    "\n",
    "meta_muts = meta_muts[meta_muts.exp != 'GLU_16321-M9-NC_000913_3gb-stationary-37-Glucose2-NH4Cl1-KH2PO43-MgSO4024-O2']\n",
    "meta_muts = meta_muts[meta_muts.exp != 'GLU_18361-M9-NC_000913_3gb-stationary-37-Glucose2-NH4Cl1-KH2PO43-MgSO4024-O2']\n",
    "meta_muts = meta_muts[meta_muts.exp != 'GLU_4721-M9-NC_000913_3gb-stationary-37-Glucose2-NH4Cl1-KH2PO43-MgSO4024-O2']\n",
    "meta_muts = meta_muts[meta_muts.exp != 'OxidizeME_PQ3']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, sys\n",
    "module_path = os.path.abspath(os.path.join('..'))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "from aledbmutil.mut import get_mut_size, get_mob_size, is_genetic_mut, is_coding_mut, predict_mutation_effect_on_feature, get_SNP_aa_pos, get_DEL_INS_MOB_aa_start_pos, get_DEL_AA_range\n",
    "\n",
    "\n",
    "meta_muts[\"genetic\"] = meta_muts[\"Details\"].apply(is_genetic_mut)\n",
    "meta_muts[\"coding\"] = meta_muts[\"Details\"].apply(is_coding_mut)\n",
    "meta_muts[\"mutation size\"] = meta_muts.apply(get_mut_size, axis=1)  # get_mut_size currently doesn't return MOB mut sizes. Isn't something currently necessary."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exp</th>\n",
       "      <th>sample</th>\n",
       "      <th>Position</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>Details</th>\n",
       "      <th>Gene</th>\n",
       "      <th>ALEdb source</th>\n",
       "      <th>genetic</th>\n",
       "      <th>clean genes</th>\n",
       "      <th>bnums</th>\n",
       "      <th>genes of interest</th>\n",
       "      <th>coding</th>\n",
       "      <th>mutation size</th>\n",
       "      <th>AA range</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>159621</th>\n",
       "      <td>GLU_ALE4-M9-NC_000913_3gb-stationary-37-Glucos...</td>\n",
       "      <td>14 21 0 1</td>\n",
       "      <td>3041398</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T→G</td>\n",
       "      <td>L29R (CTT→CGT)</td>\n",
       "      <td>ygfZ</td>\n",
       "      <td>private</td>\n",
       "      <td>True</td>\n",
       "      <td>{ygfZ}</td>\n",
       "      <td>{b2898}</td>\n",
       "      <td>{b2898}</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>(29, 29)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159668</th>\n",
       "      <td>GLU_ALE4-M9-NC_000913_3gb-stationary-37-Glucos...</td>\n",
       "      <td>16 28 1 1</td>\n",
       "      <td>3041632</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T→A</td>\n",
       "      <td>V107E (GTG→GAG)</td>\n",
       "      <td>ygfZ</td>\n",
       "      <td>private</td>\n",
       "      <td>True</td>\n",
       "      <td>{ygfZ}</td>\n",
       "      <td>{b2898}</td>\n",
       "      <td>{b2898}</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>(107, 107)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159706</th>\n",
       "      <td>GLU_ALE4-M9-NC_000913_3gb-stationary-37-Glucos...</td>\n",
       "      <td>18 24 1 1</td>\n",
       "      <td>3041393</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G→T</td>\n",
       "      <td>W27C (TGG→TGT)</td>\n",
       "      <td>ygfZ</td>\n",
       "      <td>private</td>\n",
       "      <td>True</td>\n",
       "      <td>{ygfZ}</td>\n",
       "      <td>{b2898}</td>\n",
       "      <td>{b2898}</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>(27, 27)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159725</th>\n",
       "      <td>GLU_ALE4-M9-NC_000913_3gb-stationary-37-Glucos...</td>\n",
       "      <td>18 36 1 1</td>\n",
       "      <td>3041634</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A→C</td>\n",
       "      <td>T108P (ACC→CCC)</td>\n",
       "      <td>ygfZ</td>\n",
       "      <td>private</td>\n",
       "      <td>True</td>\n",
       "      <td>{ygfZ}</td>\n",
       "      <td>{b2898}</td>\n",
       "      <td>{b2898}</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>(108, 108)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159693</th>\n",
       "      <td>GLU_ALE4-M9-NC_000913_3gb-stationary-37-Glucos...</td>\n",
       "      <td>18 11 1 1</td>\n",
       "      <td>3041973</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G→C</td>\n",
       "      <td>G221R (GGT→CGT)</td>\n",
       "      <td>ygfZ</td>\n",
       "      <td>private</td>\n",
       "      <td>True</td>\n",
       "      <td>{ygfZ}</td>\n",
       "      <td>{b2898}</td>\n",
       "      <td>{b2898}</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>(221, 221)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                      exp     sample  \\\n",
       "159621  GLU_ALE4-M9-NC_000913_3gb-stationary-37-Glucos...  14 21 0 1   \n",
       "159668  GLU_ALE4-M9-NC_000913_3gb-stationary-37-Glucos...  16 28 1 1   \n",
       "159706  GLU_ALE4-M9-NC_000913_3gb-stationary-37-Glucos...  18 24 1 1   \n",
       "159725  GLU_ALE4-M9-NC_000913_3gb-stationary-37-Glucos...  18 36 1 1   \n",
       "159693  GLU_ALE4-M9-NC_000913_3gb-stationary-37-Glucos...  18 11 1 1   \n",
       "\n",
       "        Position Mutation Type Sequence Change          Details  Gene  \\\n",
       "159621   3041398           SNP             T→G   L29R (CTT→CGT)  ygfZ   \n",
       "159668   3041632           SNP             T→A  V107E (GTG→GAG)  ygfZ   \n",
       "159706   3041393           SNP             G→T   W27C (TGG→TGT)  ygfZ   \n",
       "159725   3041634           SNP             A→C  T108P (ACC→CCC)  ygfZ   \n",
       "159693   3041973           SNP             G→C  G221R (GGT→CGT)  ygfZ   \n",
       "\n",
       "       ALEdb source  genetic clean genes    bnums genes of interest  coding  \\\n",
       "159621      private     True      {ygfZ}  {b2898}           {b2898}    True   \n",
       "159668      private     True      {ygfZ}  {b2898}           {b2898}    True   \n",
       "159706      private     True      {ygfZ}  {b2898}           {b2898}    True   \n",
       "159725      private     True      {ygfZ}  {b2898}           {b2898}    True   \n",
       "159693      private     True      {ygfZ}  {b2898}           {b2898}    True   \n",
       "\n",
       "        mutation size    AA range  \n",
       "159621              1    (29, 29)  \n",
       "159668              1  (107, 107)  \n",
       "159706              1    (27, 27)  \n",
       "159725              1  (108, 108)  \n",
       "159693              1  (221, 221)  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from aledbmutil.mut import get_SNP_aa_pos, get_DEL_AA_range\n",
    "\n",
    "def _get_SNP_AA_range(mut_details_str):\n",
    "    SNP_AA_range = ()\n",
    "    AA_pos = get_SNP_aa_pos(mut_details_str)\n",
    "    SNP_AA_range = (AA_pos, AA_pos)\n",
    "    return SNP_AA_range\n",
    "\n",
    "meta_muts[\"AA range\"] = None\n",
    "# meta_muts[\"AA range\"] = meta_muts.apply(\n",
    "#     lambda r: _get_SNP_AA_range(r[\"Details\"]) if ((r[\"Mutation Type\"] == \"SNP\") & (r[\"coding\"]==1)) else r[\"AA range\"],\n",
    "#     axis=1)\n",
    "meta_muts[\"AA range\"] = meta_muts.apply(lambda r:\n",
    "                                        _get_SNP_AA_range(r[\"Details\"]) if ((r[\"Mutation Type\"] == \"SNP\") & (r[\"coding\"]==1)) \\\n",
    "                                        else (get_DEL_AA_range(r[\"Details\"]) if ((r[\"Mutation Type\"] == \"DEL\") & (r[\"coding\"]==1)) \\\n",
    "                                        else r[\"AA range\"]),\n",
    "                                        axis=1)\n",
    "meta_muts.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exp</th>\n",
       "      <th>sample</th>\n",
       "      <th>Position</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>Details</th>\n",
       "      <th>Gene</th>\n",
       "      <th>ALEdb source</th>\n",
       "      <th>genetic</th>\n",
       "      <th>clean genes</th>\n",
       "      <th>bnums</th>\n",
       "      <th>genes of interest</th>\n",
       "      <th>coding</th>\n",
       "      <th>mutation size</th>\n",
       "      <th>AA range</th>\n",
       "      <th>AA ref seq</th>\n",
       "      <th>AA seq change</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>159621</th>\n",
       "      <td>GLU_ALE4-M9-NC_000913_3gb-stationary-37-Glucos...</td>\n",
       "      <td>14 21 0 1</td>\n",
       "      <td>3041398</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T→G</td>\n",
       "      <td>L29R (CTT→CGT)</td>\n",
       "      <td>ygfZ</td>\n",
       "      <td>private</td>\n",
       "      <td>True</td>\n",
       "      <td>{ygfZ}</td>\n",
       "      <td>{b2898}</td>\n",
       "      <td>{b2898}</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>(29, 29)</td>\n",
       "      <td>L</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159668</th>\n",
       "      <td>GLU_ALE4-M9-NC_000913_3gb-stationary-37-Glucos...</td>\n",
       "      <td>16 28 1 1</td>\n",
       "      <td>3041632</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T→A</td>\n",
       "      <td>V107E (GTG→GAG)</td>\n",
       "      <td>ygfZ</td>\n",
       "      <td>private</td>\n",
       "      <td>True</td>\n",
       "      <td>{ygfZ}</td>\n",
       "      <td>{b2898}</td>\n",
       "      <td>{b2898}</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>(107, 107)</td>\n",
       "      <td>V</td>\n",
       "      <td>E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159706</th>\n",
       "      <td>GLU_ALE4-M9-NC_000913_3gb-stationary-37-Glucos...</td>\n",
       "      <td>18 24 1 1</td>\n",
       "      <td>3041393</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G→T</td>\n",
       "      <td>W27C (TGG→TGT)</td>\n",
       "      <td>ygfZ</td>\n",
       "      <td>private</td>\n",
       "      <td>True</td>\n",
       "      <td>{ygfZ}</td>\n",
       "      <td>{b2898}</td>\n",
       "      <td>{b2898}</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>(27, 27)</td>\n",
       "      <td>W</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159725</th>\n",
       "      <td>GLU_ALE4-M9-NC_000913_3gb-stationary-37-Glucos...</td>\n",
       "      <td>18 36 1 1</td>\n",
       "      <td>3041634</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A→C</td>\n",
       "      <td>T108P (ACC→CCC)</td>\n",
       "      <td>ygfZ</td>\n",
       "      <td>private</td>\n",
       "      <td>True</td>\n",
       "      <td>{ygfZ}</td>\n",
       "      <td>{b2898}</td>\n",
       "      <td>{b2898}</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>(108, 108)</td>\n",
       "      <td>T</td>\n",
       "      <td>P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159693</th>\n",
       "      <td>GLU_ALE4-M9-NC_000913_3gb-stationary-37-Glucos...</td>\n",
       "      <td>18 11 1 1</td>\n",
       "      <td>3041973</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G→C</td>\n",
       "      <td>G221R (GGT→CGT)</td>\n",
       "      <td>ygfZ</td>\n",
       "      <td>private</td>\n",
       "      <td>True</td>\n",
       "      <td>{ygfZ}</td>\n",
       "      <td>{b2898}</td>\n",
       "      <td>{b2898}</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>(221, 221)</td>\n",
       "      <td>G</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                      exp     sample  \\\n",
       "159621  GLU_ALE4-M9-NC_000913_3gb-stationary-37-Glucos...  14 21 0 1   \n",
       "159668  GLU_ALE4-M9-NC_000913_3gb-stationary-37-Glucos...  16 28 1 1   \n",
       "159706  GLU_ALE4-M9-NC_000913_3gb-stationary-37-Glucos...  18 24 1 1   \n",
       "159725  GLU_ALE4-M9-NC_000913_3gb-stationary-37-Glucos...  18 36 1 1   \n",
       "159693  GLU_ALE4-M9-NC_000913_3gb-stationary-37-Glucos...  18 11 1 1   \n",
       "\n",
       "        Position Mutation Type Sequence Change          Details  Gene  \\\n",
       "159621   3041398           SNP             T→G   L29R (CTT→CGT)  ygfZ   \n",
       "159668   3041632           SNP             T→A  V107E (GTG→GAG)  ygfZ   \n",
       "159706   3041393           SNP             G→T   W27C (TGG→TGT)  ygfZ   \n",
       "159725   3041634           SNP             A→C  T108P (ACC→CCC)  ygfZ   \n",
       "159693   3041973           SNP             G→C  G221R (GGT→CGT)  ygfZ   \n",
       "\n",
       "       ALEdb source  genetic clean genes    bnums genes of interest  coding  \\\n",
       "159621      private     True      {ygfZ}  {b2898}           {b2898}    True   \n",
       "159668      private     True      {ygfZ}  {b2898}           {b2898}    True   \n",
       "159706      private     True      {ygfZ}  {b2898}           {b2898}    True   \n",
       "159725      private     True      {ygfZ}  {b2898}           {b2898}    True   \n",
       "159693      private     True      {ygfZ}  {b2898}           {b2898}    True   \n",
       "\n",
       "        mutation size    AA range AA ref seq AA seq change  \n",
       "159621              1    (29, 29)          L             R  \n",
       "159668              1  (107, 107)          V             E  \n",
       "159706              1    (27, 27)          W             C  \n",
       "159725              1  (108, 108)          T             P  \n",
       "159693              1  (221, 221)          G             R  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# currently only tracking seq changes with SNPs for ALEdb mutations for later processing with AA property change trends\n",
    "meta_muts['AA ref seq'] = meta_muts.apply(lambda m: m['Details'][0] if m[\"Mutation Type\"]=='SNP' else '', axis=1)\n",
    "meta_muts['AA seq change'] = meta_muts.apply(lambda m: m['Details'].split(' (')[0][-1] if m[\"Mutation Type\"]=='SNP' else '', axis=1)\n",
    "meta_muts.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "exp_sel_pres_d = {\n",
    "    'ROS_eAMP': \"paraquat\",  # starting strains were also pre-evolved on glucose\n",
    " }\n",
    "\n",
    "meta_muts['selection pressure'] = meta_muts.apply(lambda r: 'unpublished' if r['ALEdb source'] == 'private' else '', axis=1)\n",
    "meta_muts['selection pressure'] = meta_muts.exp.apply(lambda e: exp_sel_pres_d[e] if e in exp_sel_pres_d.keys() else 'unpublished')\n",
    "assert('' not in set(meta_muts['selection pressure']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "31"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AA pos</th>\n",
       "      <th>mutation type</th>\n",
       "      <th>mutation effect</th>\n",
       "      <th>mutation source</th>\n",
       "      <th>AA ref seq</th>\n",
       "      <th>AA seq change</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>29</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>L</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>107</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>V</td>\n",
       "      <td>E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>27</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>W</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>108</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>T</td>\n",
       "      <td>P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>221</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>G</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   AA pos mutation type             mutation effect mutation source  \\\n",
       "0      29           SUB  nonsynonymous substitution     unpublished   \n",
       "1     107           SUB  nonsynonymous substitution     unpublished   \n",
       "2      27           SUB  nonsynonymous substitution     unpublished   \n",
       "3     108           SUB  nonsynonymous substitution     unpublished   \n",
       "4     221           SUB  nonsynonymous substitution     unpublished   \n",
       "\n",
       "  AA ref seq AA seq change  \n",
       "0          L             R  \n",
       "1          V             E  \n",
       "2          W             C  \n",
       "3          T             P  \n",
       "4          G             R  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# TODO: fix warnings with append\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "\n",
    "\n",
    "aa_var_df = pd.DataFrame()\n",
    "\n",
    "from aledbmutil.mut import get_mut_size  #, get_DEL_INS_MOB_aa_start_pos\n",
    "\n",
    "\n",
    "for mt, mut_type_mut_df in meta_muts.groupby(\"Mutation Type\"):\n",
    "    assert mt != \"MOB\", 'Currently not handilng MOB. The below INS logic might be able to handle this.'\n",
    "    if mt == \"SNP\":\n",
    "        for _, m in mut_type_mut_df.iterrows():\n",
    "            mut_eff = \"\"\n",
    "            details_str = m[\"Details\"]\n",
    "            details_str = details_str[0:details_str.find(\" (\")]\n",
    "            \n",
    "            if details_str[-1] == '*':  # Could have them show the effected downstream positions like with ACS SynBio paper. \n",
    "                mut_eff = \"truncation\"\n",
    "            elif details_str[0] == details_str[-1]:\n",
    "                mut_eff = \"synonymous substitution\"\n",
    "            elif details_str[0] != details_str[-1]:\n",
    "                mut_eff = \"nonsynonymous substitution\"\n",
    "            else:\n",
    "                mut_eff = \"unknown\"\n",
    "            \n",
    "            aa_pos = int(details_str[1:-1])\n",
    "            aa_var_df= aa_var_df.append({\n",
    "                \"AA pos\": aa_pos,\n",
    "                \"mutation type\": \"SUB\",\n",
    "                \"mutation effect\": mut_eff,\n",
    "                \"mutation source\": m['selection pressure'],\n",
    "                'AA ref seq': m['AA ref seq'],\n",
    "                'AA seq change': m['AA seq change'],\n",
    "            }, ignore_index=True)\n",
    "    if (mt == 'DEL'):  # Unravel since downstream analysis just looks at individual mutated AAs no matter if they come from the same allele.\n",
    "        for _, m in mut_type_mut_df.iterrows():\n",
    "            mut_eff = \"deletion\"  # currently assuming only insertions remain\n",
    "            mut_size = m['mutation size']\n",
    "            if mut_size % 3 != 0:\n",
    "                assert mut_size % 3 != 0, 'not dealing with frameshifts yet' # Could have them show the effected downstream positions like with ACS SynBio paper.\n",
    "            else:\n",
    "                for aa_pos in range(m['AA range'][0], m['AA range'][1]+1):\n",
    "                    aa_var_df= aa_var_df.append({\n",
    "                        \"AA pos\": aa_pos,\n",
    "                        \"mutation type\": mt,\n",
    "                        \"mutation effect\": mut_eff,\n",
    "                        \"mutation source\": m['selection pressure'],\n",
    "                        'AA ref seq': '',\n",
    "                        'AA seq change': '',\n",
    "                    },ignore_index=True)\n",
    "    # if (mt == 'INS'):\n",
    "    #     for _, m in mut_type_mut_df.iterrows():\n",
    "    #         mut_eff = \"insertion\"  # currently assuming only insertions remain\n",
    "    #         mut_size = get_mut_size(m)\n",
    "    #         if mut_size % 3 != 0:\n",
    "    #             mut_eff = \"frameshift\"\n",
    "    #         aa_var_df= aa_var_df.append({\n",
    "    #             \"AA pos\": get_DEL_INS_MOB_aa_start_pos(m[\"Details\"]),\n",
    "    #             \"mutation type\": mt,\n",
    "    #             \"mutation effect\": mut_eff,\n",
    "    #             \"mutation source\": m['selection pressure'],\n",
    "    #             'AA ref seq': '',\n",
    "    #             'AA seq change': '',\n",
    "    #         },ignore_index=True)\n",
    "    if (mt == 'SUB'):  # Unravel since downstream analysis just looks at individual mutated AAs no matter if they come from the same allele.\n",
    "        for _, m in mut_type_mut_df.iterrows():\n",
    "            for aa_pos in range(m['AA range'][0], m['AA range'][1]+1):\n",
    "                aa_var_df= aa_var_df.append({\n",
    "                    \"AA pos\": aa_pos,\n",
    "                    \"mutation type\": mt,\n",
    "                    \"mutation effect\": \"nonsynonymous substitution\",\n",
    "                    \"mutation source\": m['selection pressure'],\n",
    "                    'AA ref seq': '',\n",
    "                    'AA seq change': '',\n",
    "                },ignore_index=True)\n",
    "\n",
    "\n",
    "display(len(aa_var_df), aa_var_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AA pos</th>\n",
       "      <th>mutation type</th>\n",
       "      <th>mutation effect</th>\n",
       "      <th>mutation source</th>\n",
       "      <th>AA ref seq</th>\n",
       "      <th>AA seq change</th>\n",
       "      <th>mutation count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>27</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>paraquat</td>\n",
       "      <td>W</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>27</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>W</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>29</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>paraquat</td>\n",
       "      <td>L</td>\n",
       "      <td>R</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>29</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>L</td>\n",
       "      <td>R</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>46</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>T</td>\n",
       "      <td>I</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   AA pos mutation type             mutation effect mutation source  \\\n",
       "0      27           SUB  nonsynonymous substitution        paraquat   \n",
       "1      27           SUB  nonsynonymous substitution     unpublished   \n",
       "2      29           SUB  nonsynonymous substitution        paraquat   \n",
       "3      29           SUB  nonsynonymous substitution     unpublished   \n",
       "4      46           SUB  nonsynonymous substitution     unpublished   \n",
       "\n",
       "  AA ref seq AA seq change  mutation count  \n",
       "0          W             C               1  \n",
       "1          W             C               1  \n",
       "2          L             R               6  \n",
       "3          L             R               1  \n",
       "4          T             I               1  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aa_var_cnt_df = pd.DataFrame()\n",
    "gcols = ['AA pos', 'mutation type', 'mutation effect', 'mutation source', 'AA ref seq', 'AA seq change']\n",
    "for g, gdf in aa_var_df.groupby(gcols):  # Essentially just finding duplicates\n",
    "    d = dict(zip(gcols, g))\n",
    "    d['mutation count'] = len(gdf)\n",
    "    aa_var_cnt_df = aa_var_cnt_df.append(d, ignore_index=True)\n",
    "aa_var_cnt_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AA pos</th>\n",
       "      <th>mutation type</th>\n",
       "      <th>mutation effect</th>\n",
       "      <th>mutation source</th>\n",
       "      <th>AA ref seq</th>\n",
       "      <th>AA seq change</th>\n",
       "      <th>mutation count</th>\n",
       "      <th>mutation effect color</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>27</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>paraquat</td>\n",
       "      <td>W</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>27</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>W</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>29</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>paraquat</td>\n",
       "      <td>L</td>\n",
       "      <td>R</td>\n",
       "      <td>6</td>\n",
       "      <td>#DD8452</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>29</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>L</td>\n",
       "      <td>R</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>46</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>T</td>\n",
       "      <td>I</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   AA pos mutation type             mutation effect mutation source  \\\n",
       "0      27           SUB  nonsynonymous substitution        paraquat   \n",
       "1      27           SUB  nonsynonymous substitution     unpublished   \n",
       "2      29           SUB  nonsynonymous substitution        paraquat   \n",
       "3      29           SUB  nonsynonymous substitution     unpublished   \n",
       "4      46           SUB  nonsynonymous substitution     unpublished   \n",
       "\n",
       "  AA ref seq AA seq change  mutation count mutation effect color  \n",
       "0          W             C               1               #DD8452  \n",
       "1          W             C               1               #DD8452  \n",
       "2          L             R               6               #DD8452  \n",
       "3          L             R               1               #DD8452  \n",
       "4          T             I               1               #DD8452  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os, sys\n",
    "module_path = os.path.abspath(os.path.join('..'))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "import const\n",
    "\n",
    "\n",
    "aa_var_cnt_df['mutation effect color'] = aa_var_cnt_df['mutation effect'].apply(lambda e: const.MUTATION_TYPE_COLOR_PALETTE[e])\n",
    "aa_var_cnt_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'paraquat', 'unpublished'}"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(aa_var_cnt_df['mutation source'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AA pos</th>\n",
       "      <th>mutation type</th>\n",
       "      <th>mutation effect</th>\n",
       "      <th>mutation source</th>\n",
       "      <th>AA ref seq</th>\n",
       "      <th>AA seq change</th>\n",
       "      <th>mutation count</th>\n",
       "      <th>mutation effect color</th>\n",
       "      <th>selection pressure color</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>27</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>paraquat</td>\n",
       "      <td>W</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#6DCCDA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>27</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>W</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#AD8BC9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>29</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>paraquat</td>\n",
       "      <td>L</td>\n",
       "      <td>R</td>\n",
       "      <td>6</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#6DCCDA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>29</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>L</td>\n",
       "      <td>R</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#AD8BC9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>46</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>T</td>\n",
       "      <td>I</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#AD8BC9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>61</td>\n",
       "      <td>SUB</td>\n",
       "      <td>synonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>#55A868</td>\n",
       "      <td>#AD8BC9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>107</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>paraquat</td>\n",
       "      <td>V</td>\n",
       "      <td>E</td>\n",
       "      <td>5</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#6DCCDA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>107</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>V</td>\n",
       "      <td>E</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#AD8BC9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>108</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>paraquat</td>\n",
       "      <td>T</td>\n",
       "      <td>P</td>\n",
       "      <td>6</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#6DCCDA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>108</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>T</td>\n",
       "      <td>P</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#AD8BC9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>119</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#AD8BC9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>197</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>G</td>\n",
       "      <td>D</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#AD8BC9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>221</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>paraquat</td>\n",
       "      <td>G</td>\n",
       "      <td>R</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#6DCCDA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>221</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>G</td>\n",
       "      <td>R</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#AD8BC9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>221</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>G</td>\n",
       "      <td>V</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#AD8BC9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>229</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>Y</td>\n",
       "      <td>H</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#AD8BC9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>251</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>L</td>\n",
       "      <td>Q</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#AD8BC9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    AA pos mutation type             mutation effect mutation source  \\\n",
       "0       27           SUB  nonsynonymous substitution        paraquat   \n",
       "1       27           SUB  nonsynonymous substitution     unpublished   \n",
       "2       29           SUB  nonsynonymous substitution        paraquat   \n",
       "3       29           SUB  nonsynonymous substitution     unpublished   \n",
       "4       46           SUB  nonsynonymous substitution     unpublished   \n",
       "5       61           SUB     synonymous substitution     unpublished   \n",
       "6      107           SUB  nonsynonymous substitution        paraquat   \n",
       "7      107           SUB  nonsynonymous substitution     unpublished   \n",
       "8      108           SUB  nonsynonymous substitution        paraquat   \n",
       "9      108           SUB  nonsynonymous substitution     unpublished   \n",
       "10     119           SUB  nonsynonymous substitution     unpublished   \n",
       "11     197           SUB  nonsynonymous substitution     unpublished   \n",
       "12     221           SUB  nonsynonymous substitution        paraquat   \n",
       "13     221           SUB  nonsynonymous substitution     unpublished   \n",
       "14     221           SUB  nonsynonymous substitution     unpublished   \n",
       "15     229           SUB  nonsynonymous substitution     unpublished   \n",
       "16     251           SUB  nonsynonymous substitution     unpublished   \n",
       "\n",
       "   AA ref seq AA seq change  mutation count mutation effect color  \\\n",
       "0           W             C               1               #DD8452   \n",
       "1           W             C               1               #DD8452   \n",
       "2           L             R               6               #DD8452   \n",
       "3           L             R               1               #DD8452   \n",
       "4           T             I               1               #DD8452   \n",
       "5           A             A               1               #55A868   \n",
       "6           V             E               5               #DD8452   \n",
       "7           V             E               1               #DD8452   \n",
       "8           T             P               6               #DD8452   \n",
       "9           T             P               1               #DD8452   \n",
       "10          G             C               1               #DD8452   \n",
       "11          G             D               1               #DD8452   \n",
       "12          G             R               1               #DD8452   \n",
       "13          G             R               1               #DD8452   \n",
       "14          G             V               1               #DD8452   \n",
       "15          Y             H               1               #DD8452   \n",
       "16          L             Q               1               #DD8452   \n",
       "\n",
       "   selection pressure color  \n",
       "0                   #6DCCDA  \n",
       "1                   #AD8BC9  \n",
       "2                   #6DCCDA  \n",
       "3                   #AD8BC9  \n",
       "4                   #AD8BC9  \n",
       "5                   #AD8BC9  \n",
       "6                   #6DCCDA  \n",
       "7                   #AD8BC9  \n",
       "8                   #6DCCDA  \n",
       "9                   #AD8BC9  \n",
       "10                  #AD8BC9  \n",
       "11                  #AD8BC9  \n",
       "12                  #6DCCDA  \n",
       "13                  #AD8BC9  \n",
       "14                  #AD8BC9  \n",
       "15                  #AD8BC9  \n",
       "16                  #AD8BC9  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sel_pres_clr_d = {\n",
    "    # 'FeSO4': '#729ECE',\n",
    "    # 'Homo sapiens tpi': '#FF9E4A',\n",
    "    # 'Pyrobaculum aerophilum pgi': '#67BF5C',\n",
    "    # 'Pyrobaculum aerophilum tpi': '#ED665D',\n",
    "    # 'Vibrio cholerae tpi': '#AD8BC9',\n",
    "    # 'isobutanol': '#ED97CA',\n",
    "    'paraquat': '#6DCCDA',\n",
    "    'unpublished': '#AD8BC9'\n",
    " }\n",
    "\n",
    "aa_var_cnt_df['selection pressure color'] = aa_var_cnt_df['mutation source'].apply(lambda e: sel_pres_clr_d[e])\n",
    "aa_var_cnt_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AA pos</th>\n",
       "      <th>mutation type</th>\n",
       "      <th>mutation effect</th>\n",
       "      <th>mutation source</th>\n",
       "      <th>AA ref seq</th>\n",
       "      <th>AA seq change</th>\n",
       "      <th>mutation count</th>\n",
       "      <th>mutation effect color</th>\n",
       "      <th>selection pressure color</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>27</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>paraquat</td>\n",
       "      <td>W</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#6DCCDA</td>\n",
       "      <td>SUB W27C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>27</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>W</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#AD8BC9</td>\n",
       "      <td>SUB W27C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>29</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>paraquat</td>\n",
       "      <td>L</td>\n",
       "      <td>R</td>\n",
       "      <td>6</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#6DCCDA</td>\n",
       "      <td>SUB L29R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>29</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>L</td>\n",
       "      <td>R</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#AD8BC9</td>\n",
       "      <td>SUB L29R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>46</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>T</td>\n",
       "      <td>I</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#AD8BC9</td>\n",
       "      <td>SUB T46I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>61</td>\n",
       "      <td>SUB</td>\n",
       "      <td>synonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>#55A868</td>\n",
       "      <td>#AD8BC9</td>\n",
       "      <td>SUB A61A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>107</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>paraquat</td>\n",
       "      <td>V</td>\n",
       "      <td>E</td>\n",
       "      <td>5</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#6DCCDA</td>\n",
       "      <td>SUB V107E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>107</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>V</td>\n",
       "      <td>E</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#AD8BC9</td>\n",
       "      <td>SUB V107E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>108</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>paraquat</td>\n",
       "      <td>T</td>\n",
       "      <td>P</td>\n",
       "      <td>6</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#6DCCDA</td>\n",
       "      <td>SUB T108P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>108</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>T</td>\n",
       "      <td>P</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#AD8BC9</td>\n",
       "      <td>SUB T108P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>119</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#AD8BC9</td>\n",
       "      <td>SUB G119C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>197</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>G</td>\n",
       "      <td>D</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#AD8BC9</td>\n",
       "      <td>SUB G197D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>221</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>paraquat</td>\n",
       "      <td>G</td>\n",
       "      <td>R</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#6DCCDA</td>\n",
       "      <td>SUB G221R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>221</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>G</td>\n",
       "      <td>R</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#AD8BC9</td>\n",
       "      <td>SUB G221R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>221</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>G</td>\n",
       "      <td>V</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#AD8BC9</td>\n",
       "      <td>SUB G221V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>229</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>Y</td>\n",
       "      <td>H</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#AD8BC9</td>\n",
       "      <td>SUB Y229H</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>251</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>L</td>\n",
       "      <td>Q</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#AD8BC9</td>\n",
       "      <td>SUB L251Q</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    AA pos mutation type             mutation effect mutation source  \\\n",
       "0       27           SUB  nonsynonymous substitution        paraquat   \n",
       "1       27           SUB  nonsynonymous substitution     unpublished   \n",
       "2       29           SUB  nonsynonymous substitution        paraquat   \n",
       "3       29           SUB  nonsynonymous substitution     unpublished   \n",
       "4       46           SUB  nonsynonymous substitution     unpublished   \n",
       "5       61           SUB     synonymous substitution     unpublished   \n",
       "6      107           SUB  nonsynonymous substitution        paraquat   \n",
       "7      107           SUB  nonsynonymous substitution     unpublished   \n",
       "8      108           SUB  nonsynonymous substitution        paraquat   \n",
       "9      108           SUB  nonsynonymous substitution     unpublished   \n",
       "10     119           SUB  nonsynonymous substitution     unpublished   \n",
       "11     197           SUB  nonsynonymous substitution     unpublished   \n",
       "12     221           SUB  nonsynonymous substitution        paraquat   \n",
       "13     221           SUB  nonsynonymous substitution     unpublished   \n",
       "14     221           SUB  nonsynonymous substitution     unpublished   \n",
       "15     229           SUB  nonsynonymous substitution     unpublished   \n",
       "16     251           SUB  nonsynonymous substitution     unpublished   \n",
       "\n",
       "   AA ref seq AA seq change  mutation count mutation effect color  \\\n",
       "0           W             C               1               #DD8452   \n",
       "1           W             C               1               #DD8452   \n",
       "2           L             R               6               #DD8452   \n",
       "3           L             R               1               #DD8452   \n",
       "4           T             I               1               #DD8452   \n",
       "5           A             A               1               #55A868   \n",
       "6           V             E               5               #DD8452   \n",
       "7           V             E               1               #DD8452   \n",
       "8           T             P               6               #DD8452   \n",
       "9           T             P               1               #DD8452   \n",
       "10          G             C               1               #DD8452   \n",
       "11          G             D               1               #DD8452   \n",
       "12          G             R               1               #DD8452   \n",
       "13          G             R               1               #DD8452   \n",
       "14          G             V               1               #DD8452   \n",
       "15          Y             H               1               #DD8452   \n",
       "16          L             Q               1               #DD8452   \n",
       "\n",
       "   selection pressure color       name  \n",
       "0                   #6DCCDA   SUB W27C  \n",
       "1                   #AD8BC9   SUB W27C  \n",
       "2                   #6DCCDA   SUB L29R  \n",
       "3                   #AD8BC9   SUB L29R  \n",
       "4                   #AD8BC9   SUB T46I  \n",
       "5                   #AD8BC9   SUB A61A  \n",
       "6                   #6DCCDA  SUB V107E  \n",
       "7                   #AD8BC9  SUB V107E  \n",
       "8                   #6DCCDA  SUB T108P  \n",
       "9                   #AD8BC9  SUB T108P  \n",
       "10                  #AD8BC9  SUB G119C  \n",
       "11                  #AD8BC9  SUB G197D  \n",
       "12                  #6DCCDA  SUB G221R  \n",
       "13                  #AD8BC9  SUB G221R  \n",
       "14                  #AD8BC9  SUB G221V  \n",
       "15                  #AD8BC9  SUB Y229H  \n",
       "16                  #AD8BC9  SUB L251Q  "
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def _get_variant_name(variant):\n",
    "    var_name = ''\n",
    "    if variant['mutation type'] == 'SUB':\n",
    "        var_name = variant['mutation type'] + \" \" + variant['AA ref seq'] + str(variant['AA pos']) + variant['AA seq change']\n",
    "    return var_name\n",
    "\n",
    "\n",
    "aa_var_cnt_df['name'] = aa_var_cnt_df.apply(lambda r: _get_variant_name(r), axis=1)\n",
    "aa_var_cnt_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "aa_var_cnt_df.to_csv('ygfZ_ale_muts.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>fill</th>\n",
       "      <th>edge</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>paraquat</td>\n",
       "      <td>#6DCCDA</td>\n",
       "      <td>white</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>unpublished</td>\n",
       "      <td>#AD8BC9</td>\n",
       "      <td>white</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>white</td>\n",
       "      <td>#DD8452</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>synonymous substitution</td>\n",
       "      <td>white</td>\n",
       "      <td>#55A868</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         name     fill     edge\n",
       "0                    paraquat  #6DCCDA    white\n",
       "1                 unpublished  #AD8BC9    white\n",
       "2  nonsynonymous substitution    white  #DD8452\n",
       "3     synonymous substitution    white  #55A868"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "mut_needle_plot_legend_df = pd.DataFrame()\n",
    "mut_needle_plot_legend_df = aa_var_cnt_df.copy()\n",
    "mut_needle_plot_legend_df = mut_needle_plot_legend_df[['mutation source', 'selection pressure color', ]]\n",
    "mut_needle_plot_legend_df.rename(columns={'mutation source': 'name', 'selection pressure color': 'fill'}, inplace=True)\n",
    "mut_needle_plot_legend_df['edge'] = 'white'\n",
    "mut_needle_plot_legend_df.drop_duplicates(inplace=True)\n",
    "\n",
    "df = aa_var_cnt_df.copy()\n",
    "df = df[['mutation effect', 'mutation effect color']]\n",
    "df.rename(columns={'mutation effect': 'name', 'mutation effect color': 'edge'}, inplace=True)\n",
    "df.drop_duplicates(inplace=True)\n",
    "df['fill'] = 'white'\n",
    "mut_needle_plot_legend_df = mut_needle_plot_legend_df.append(df, ignore_index=True)\n",
    "display(mut_needle_plot_legend_df)\n",
    "\n",
    "mut_needle_plot_legend_df.to_csv('ygfZ_ale_muts_legend.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>range</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chain</td>\n",
       "      <td>2</td>\n",
       "      <td>326</td>\n",
       "      <td>(2, 325)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Glycine cleavage T-protein/YgfZ, C-terminal</td>\n",
       "      <td>244</td>\n",
       "      <td>325</td>\n",
       "      <td>(244, 82)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Folate-binding domain</td>\n",
       "      <td>6</td>\n",
       "      <td>242</td>\n",
       "      <td>(6, 237)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Folate binding site</td>\n",
       "      <td>189</td>\n",
       "      <td>189</td>\n",
       "      <td>(189, 1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Aminomethyltransferase folate-binding domain</td>\n",
       "      <td>31</td>\n",
       "      <td>135</td>\n",
       "      <td>(31, 105)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Folate binding site</td>\n",
       "      <td>27</td>\n",
       "      <td>27</td>\n",
       "      <td>(27, 1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>GcvT family signature motif</td>\n",
       "      <td>187</td>\n",
       "      <td>251</td>\n",
       "      <td>(187, 65)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Acetate binding site</td>\n",
       "      <td>27</td>\n",
       "      <td>27</td>\n",
       "      <td>(27, 1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Acetate binding site</td>\n",
       "      <td>74</td>\n",
       "      <td>74</td>\n",
       "      <td>(74, 1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Acetate binding site</td>\n",
       "      <td>162</td>\n",
       "      <td>162</td>\n",
       "      <td>(162, 1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Acetate binding site</td>\n",
       "      <td>189</td>\n",
       "      <td>189</td>\n",
       "      <td>(189, 1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Acetate binding site</td>\n",
       "      <td>193</td>\n",
       "      <td>193</td>\n",
       "      <td>(193, 1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Ca binding site</td>\n",
       "      <td>286</td>\n",
       "      <td>286</td>\n",
       "      <td>(286, 1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Cl binding site</td>\n",
       "      <td>68</td>\n",
       "      <td>68</td>\n",
       "      <td>(68, 1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Cl binding site</td>\n",
       "      <td>209</td>\n",
       "      <td>210</td>\n",
       "      <td>(209, 2)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Cl binding site</td>\n",
       "      <td>246</td>\n",
       "      <td>246</td>\n",
       "      <td>(246, 1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Ethandiol binding site</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>(16, 1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Ethandiol binding site</td>\n",
       "      <td>23</td>\n",
       "      <td>25</td>\n",
       "      <td>(23, 3)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Ethandiol binding site</td>\n",
       "      <td>38</td>\n",
       "      <td>38</td>\n",
       "      <td>(38, 1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Ethandiol binding site</td>\n",
       "      <td>50</td>\n",
       "      <td>50</td>\n",
       "      <td>(50, 1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Ethandiol binding site</td>\n",
       "      <td>99</td>\n",
       "      <td>99</td>\n",
       "      <td>(99, 1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Ethandiol binding site</td>\n",
       "      <td>107</td>\n",
       "      <td>109</td>\n",
       "      <td>(107, 3)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Ethandiol binding site</td>\n",
       "      <td>116</td>\n",
       "      <td>117</td>\n",
       "      <td>(116, 2)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Ethandiol binding site</td>\n",
       "      <td>260</td>\n",
       "      <td>260</td>\n",
       "      <td>(260, 1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>SO4 binding site</td>\n",
       "      <td>27</td>\n",
       "      <td>27</td>\n",
       "      <td>(27, 1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>SO4 binding site</td>\n",
       "      <td>74</td>\n",
       "      <td>74</td>\n",
       "      <td>(74, 1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>SO4 binding site</td>\n",
       "      <td>79</td>\n",
       "      <td>79</td>\n",
       "      <td>(79, 1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>SO4 binding site</td>\n",
       "      <td>145</td>\n",
       "      <td>145</td>\n",
       "      <td>(145, 1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>SO4 binding site</td>\n",
       "      <td>154</td>\n",
       "      <td>154</td>\n",
       "      <td>(154, 1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>SO4 binding site</td>\n",
       "      <td>162</td>\n",
       "      <td>162</td>\n",
       "      <td>(162, 1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>SO4 binding site</td>\n",
       "      <td>189</td>\n",
       "      <td>189</td>\n",
       "      <td>(189, 1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>SO4 binding site</td>\n",
       "      <td>193</td>\n",
       "      <td>193</td>\n",
       "      <td>(193, 1)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         feature start  end      range\n",
       "0                                          Chain     2  326   (2, 325)\n",
       "1    Glycine cleavage T-protein/YgfZ, C-terminal   244  325  (244, 82)\n",
       "2                          Folate-binding domain     6  242   (6, 237)\n",
       "3                            Folate binding site   189  189   (189, 1)\n",
       "4   Aminomethyltransferase folate-binding domain    31  135  (31, 105)\n",
       "5                            Folate binding site    27   27    (27, 1)\n",
       "6                    GcvT family signature motif   187  251  (187, 65)\n",
       "7                           Acetate binding site    27   27    (27, 1)\n",
       "8                           Acetate binding site    74   74    (74, 1)\n",
       "9                           Acetate binding site   162  162   (162, 1)\n",
       "10                          Acetate binding site   189  189   (189, 1)\n",
       "11                          Acetate binding site   193  193   (193, 1)\n",
       "12                               Ca binding site   286  286   (286, 1)\n",
       "13                               Cl binding site    68   68    (68, 1)\n",
       "14                               Cl binding site   209  210   (209, 2)\n",
       "15                               Cl binding site   246  246   (246, 1)\n",
       "16                        Ethandiol binding site    16   16    (16, 1)\n",
       "17                        Ethandiol binding site    23   25    (23, 3)\n",
       "18                        Ethandiol binding site    38   38    (38, 1)\n",
       "19                        Ethandiol binding site    50   50    (50, 1)\n",
       "20                        Ethandiol binding site    99   99    (99, 1)\n",
       "21                        Ethandiol binding site   107  109   (107, 3)\n",
       "22                        Ethandiol binding site   116  117   (116, 2)\n",
       "23                        Ethandiol binding site   260  260   (260, 1)\n",
       "24                              SO4 binding site    27   27    (27, 1)\n",
       "25                              SO4 binding site    74   74    (74, 1)\n",
       "26                              SO4 binding site    79   79    (79, 1)\n",
       "27                              SO4 binding site   145  145   (145, 1)\n",
       "28                              SO4 binding site   154  154   (154, 1)\n",
       "29                              SO4 binding site   162  162   (162, 1)\n",
       "30                              SO4 binding site   189  189   (189, 1)\n",
       "31                              SO4 binding site   193  193   (193, 1)"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "feats_df = pd.read_pickle('./annot_df.pkl')\n",
    "\n",
    "# filtering out features that aren't useful\n",
    "feats_to_remove = [\n",
    "    # 'Chain',\n",
    "    'Initiator methionine',\n",
    "    'tRNA-modifying protein YgfZ [ygfZ]',\n",
    "    'tRNA-modifying protein YgfZ',\n",
    "    'folate-binding protein YgfZ',\n",
    "    'UNCHARACTERIZED',\n",
    "    'Aminomethyltransferase, folate-binding domain',  # redundant\n",
    "    'YgfZ/GcvT conserved site',\n",
    "    'YgfZ/GcvT',\n",
    "    'Aminomethyltransferase beta-barrel domains',\n",
    "    'Aminomethyltransferase beta-barrel domain',\n",
    "]\n",
    "feats_df = feats_df[~(feats_df['feature'].isin(feats_to_remove))]\n",
    "feats_df.reset_index(inplace=True, drop=True)\n",
    "feats_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>range</th>\n",
       "      <th>layer</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chain</td>\n",
       "      <td>2</td>\n",
       "      <td>326</td>\n",
       "      <td>(2, 325)</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Glycine cleavage T-protein/YgfZ, C-terminal</td>\n",
       "      <td>244</td>\n",
       "      <td>325</td>\n",
       "      <td>(244, 82)</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Folate-binding domain</td>\n",
       "      <td>6</td>\n",
       "      <td>242</td>\n",
       "      <td>(6, 237)</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Folate binding site</td>\n",
       "      <td>189</td>\n",
       "      <td>189</td>\n",
       "      <td>(189, 1)</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Aminomethyltransferase folate-binding domain</td>\n",
       "      <td>31</td>\n",
       "      <td>135</td>\n",
       "      <td>(31, 105)</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        feature start  end      range  layer\n",
       "0                                         Chain     2  326   (2, 325)      0\n",
       "1   Glycine cleavage T-protein/YgfZ, C-terminal   244  325  (244, 82)      1\n",
       "2                         Folate-binding domain     6  242   (6, 237)      2\n",
       "3                           Folate binding site   189  189   (189, 1)      3\n",
       "4  Aminomethyltransferase folate-binding domain    31  135  (31, 105)      4"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "layer = 0\n",
    "layer_d = dict()\n",
    "for f in feats_df.feature.unique():\n",
    "    layer_d[f] = layer\n",
    "    layer += 1\n",
    "feats_df['layer'] = feats_df['feature'].apply(lambda f: layer_d[f])\n",
    "feats_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "11"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(feats_df.feature.unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>layer</th>\n",
       "      <th>color</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chain</td>\n",
       "      <td>2</td>\n",
       "      <td>326</td>\n",
       "      <td>0</td>\n",
       "      <td>#7F3C8D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Glycine cleavage T-protein/YgfZ, C-terminal</td>\n",
       "      <td>244</td>\n",
       "      <td>325</td>\n",
       "      <td>1</td>\n",
       "      <td>#11A579</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Folate-binding domain</td>\n",
       "      <td>6</td>\n",
       "      <td>242</td>\n",
       "      <td>2</td>\n",
       "      <td>#3969AC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Folate binding site</td>\n",
       "      <td>189</td>\n",
       "      <td>189</td>\n",
       "      <td>3</td>\n",
       "      <td>#F2B701</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Aminomethyltransferase folate-binding domain</td>\n",
       "      <td>31</td>\n",
       "      <td>135</td>\n",
       "      <td>4</td>\n",
       "      <td>#E73F74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Folate binding site</td>\n",
       "      <td>27</td>\n",
       "      <td>27</td>\n",
       "      <td>3</td>\n",
       "      <td>#F2B701</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>GcvT family signature motif</td>\n",
       "      <td>187</td>\n",
       "      <td>251</td>\n",
       "      <td>5</td>\n",
       "      <td>#80BA5A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Acetate binding site</td>\n",
       "      <td>27</td>\n",
       "      <td>27</td>\n",
       "      <td>6</td>\n",
       "      <td>#E68310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Acetate binding site</td>\n",
       "      <td>74</td>\n",
       "      <td>74</td>\n",
       "      <td>6</td>\n",
       "      <td>#E68310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Acetate binding site</td>\n",
       "      <td>162</td>\n",
       "      <td>162</td>\n",
       "      <td>6</td>\n",
       "      <td>#E68310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Acetate binding site</td>\n",
       "      <td>189</td>\n",
       "      <td>189</td>\n",
       "      <td>6</td>\n",
       "      <td>#E68310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Acetate binding site</td>\n",
       "      <td>193</td>\n",
       "      <td>193</td>\n",
       "      <td>6</td>\n",
       "      <td>#E68310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Ca binding site</td>\n",
       "      <td>286</td>\n",
       "      <td>286</td>\n",
       "      <td>7</td>\n",
       "      <td>#008695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Cl binding site</td>\n",
       "      <td>68</td>\n",
       "      <td>68</td>\n",
       "      <td>8</td>\n",
       "      <td>#CF1C90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Cl binding site</td>\n",
       "      <td>209</td>\n",
       "      <td>210</td>\n",
       "      <td>8</td>\n",
       "      <td>#CF1C90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Cl binding site</td>\n",
       "      <td>246</td>\n",
       "      <td>246</td>\n",
       "      <td>8</td>\n",
       "      <td>#CF1C90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Ethandiol binding site</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>9</td>\n",
       "      <td>#F97B72</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Ethandiol binding site</td>\n",
       "      <td>23</td>\n",
       "      <td>25</td>\n",
       "      <td>9</td>\n",
       "      <td>#F97B72</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Ethandiol binding site</td>\n",
       "      <td>38</td>\n",
       "      <td>38</td>\n",
       "      <td>9</td>\n",
       "      <td>#F97B72</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Ethandiol binding site</td>\n",
       "      <td>50</td>\n",
       "      <td>50</td>\n",
       "      <td>9</td>\n",
       "      <td>#F97B72</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Ethandiol binding site</td>\n",
       "      <td>99</td>\n",
       "      <td>99</td>\n",
       "      <td>9</td>\n",
       "      <td>#F97B72</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Ethandiol binding site</td>\n",
       "      <td>107</td>\n",
       "      <td>109</td>\n",
       "      <td>9</td>\n",
       "      <td>#F97B72</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Ethandiol binding site</td>\n",
       "      <td>116</td>\n",
       "      <td>117</td>\n",
       "      <td>9</td>\n",
       "      <td>#F97B72</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Ethandiol binding site</td>\n",
       "      <td>260</td>\n",
       "      <td>260</td>\n",
       "      <td>9</td>\n",
       "      <td>#F97B72</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>SO4 binding site</td>\n",
       "      <td>27</td>\n",
       "      <td>27</td>\n",
       "      <td>10</td>\n",
       "      <td>#A5AA99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>SO4 binding site</td>\n",
       "      <td>74</td>\n",
       "      <td>74</td>\n",
       "      <td>10</td>\n",
       "      <td>#A5AA99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>SO4 binding site</td>\n",
       "      <td>79</td>\n",
       "      <td>79</td>\n",
       "      <td>10</td>\n",
       "      <td>#A5AA99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>SO4 binding site</td>\n",
       "      <td>145</td>\n",
       "      <td>145</td>\n",
       "      <td>10</td>\n",
       "      <td>#A5AA99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>SO4 binding site</td>\n",
       "      <td>154</td>\n",
       "      <td>154</td>\n",
       "      <td>10</td>\n",
       "      <td>#A5AA99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>SO4 binding site</td>\n",
       "      <td>162</td>\n",
       "      <td>162</td>\n",
       "      <td>10</td>\n",
       "      <td>#A5AA99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>SO4 binding site</td>\n",
       "      <td>189</td>\n",
       "      <td>189</td>\n",
       "      <td>10</td>\n",
       "      <td>#A5AA99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>SO4 binding site</td>\n",
       "      <td>193</td>\n",
       "      <td>193</td>\n",
       "      <td>10</td>\n",
       "      <td>#A5AA99</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         feature start  end  layer    color\n",
       "0                                          Chain     2  326      0  #7F3C8D\n",
       "1    Glycine cleavage T-protein/YgfZ, C-terminal   244  325      1  #11A579\n",
       "2                          Folate-binding domain     6  242      2  #3969AC\n",
       "3                            Folate binding site   189  189      3  #F2B701\n",
       "4   Aminomethyltransferase folate-binding domain    31  135      4  #E73F74\n",
       "5                            Folate binding site    27   27      3  #F2B701\n",
       "6                    GcvT family signature motif   187  251      5  #80BA5A\n",
       "7                           Acetate binding site    27   27      6  #E68310\n",
       "8                           Acetate binding site    74   74      6  #E68310\n",
       "9                           Acetate binding site   162  162      6  #E68310\n",
       "10                          Acetate binding site   189  189      6  #E68310\n",
       "11                          Acetate binding site   193  193      6  #E68310\n",
       "12                               Ca binding site   286  286      7  #008695\n",
       "13                               Cl binding site    68   68      8  #CF1C90\n",
       "14                               Cl binding site   209  210      8  #CF1C90\n",
       "15                               Cl binding site   246  246      8  #CF1C90\n",
       "16                        Ethandiol binding site    16   16      9  #F97B72\n",
       "17                        Ethandiol binding site    23   25      9  #F97B72\n",
       "18                        Ethandiol binding site    38   38      9  #F97B72\n",
       "19                        Ethandiol binding site    50   50      9  #F97B72\n",
       "20                        Ethandiol binding site    99   99      9  #F97B72\n",
       "21                        Ethandiol binding site   107  109      9  #F97B72\n",
       "22                        Ethandiol binding site   116  117      9  #F97B72\n",
       "23                        Ethandiol binding site   260  260      9  #F97B72\n",
       "24                              SO4 binding site    27   27     10  #A5AA99\n",
       "25                              SO4 binding site    74   74     10  #A5AA99\n",
       "26                              SO4 binding site    79   79     10  #A5AA99\n",
       "27                              SO4 binding site   145  145     10  #A5AA99\n",
       "28                              SO4 binding site   154  154     10  #A5AA99\n",
       "29                              SO4 binding site   162  162     10  #A5AA99\n",
       "30                              SO4 binding site   189  189     10  #A5AA99\n",
       "31                              SO4 binding site   193  193     10  #A5AA99"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "{'Chain': '#7F3C8D',\n",
       " 'Glycine cleavage T-protein/YgfZ, C-terminal': '#11A579',\n",
       " 'Folate-binding domain': '#3969AC',\n",
       " 'Folate binding site': '#F2B701',\n",
       " 'Aminomethyltransferase folate-binding domain': '#E73F74',\n",
       " 'GcvT family signature motif': '#80BA5A',\n",
       " 'Acetate binding site': '#E68310',\n",
       " 'Ca binding site': '#008695',\n",
       " 'Cl binding site': '#CF1C90',\n",
       " 'Ethandiol binding site': '#F97B72',\n",
       " 'SO4 binding site': '#A5AA99'}"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "feat_color_d = dict()\n",
    "# muted=[\"#4878D0\", \"#EE854A\", \"#6ACC64\", \"#D65F5F\", \"#956CB4\", \"#8C613C\", \"#DC7EC0\", \"#797979\", \"#D5BB67\", \"#82C6E2\"]  # from seaborn\n",
    "clrs = ['#7F3C8D',\n",
    "'#11A579',\n",
    "'#3969AC',\n",
    "'#F2B701',\n",
    "'#E73F74',\n",
    "'#80BA5A',\n",
    "'#E68310',\n",
    "'#008695',\n",
    "'#CF1C90',\n",
    "'#F97B72',\n",
    "'#A5AA99']\n",
    "feats = list(feats_df.feature.unique())\n",
    "for i in range(0, len(feats_df.feature.unique())):\n",
    "    feat_color_d[feats[i]] = clrs[i]\n",
    "feat_color_d\n",
    "\n",
    "feats_df['color'] = feats_df['feature'].apply(lambda f: feat_color_d[f])\n",
    "\n",
    "feats_df.drop(columns=[\"range\"], inplace=True)\n",
    "feats_df.drop_duplicates(inplace=True)\n",
    "\n",
    "display(feats_df, feat_color_d)\n",
    "feats_df.to_csv('./ygfZ_feats.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filtering out some features that are not useful for the NGL visual 3D analysis\n",
    "ngl_feat_df = feats_df.copy()\n",
    "ngl_feat_df = ngl_feat_df[ngl_feat_df['feature'].isin(['Ethandiol binding site','GcvT family signature motif','Aminomethyltransferase folate-binding domain'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>color</th>\n",
       "      <th>size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Aminomethyltransferase folate-binding domain</td>\n",
       "      <td>31</td>\n",
       "      <td>135</td>\n",
       "      <td>#E73F74</td>\n",
       "      <td>105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GcvT family signature motif</td>\n",
       "      <td>187</td>\n",
       "      <td>251</td>\n",
       "      <td>#80BA5A</td>\n",
       "      <td>65</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Ethandiol binding site</td>\n",
       "      <td>23</td>\n",
       "      <td>25</td>\n",
       "      <td>#F97B72</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Ethandiol binding site</td>\n",
       "      <td>107</td>\n",
       "      <td>109</td>\n",
       "      <td>#F97B72</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Ethandiol binding site</td>\n",
       "      <td>116</td>\n",
       "      <td>117</td>\n",
       "      <td>#F97B72</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Ethandiol binding site</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>#F97B72</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Ethandiol binding site</td>\n",
       "      <td>38</td>\n",
       "      <td>38</td>\n",
       "      <td>#F97B72</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Ethandiol binding site</td>\n",
       "      <td>50</td>\n",
       "      <td>50</td>\n",
       "      <td>#F97B72</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Ethandiol binding site</td>\n",
       "      <td>99</td>\n",
       "      <td>99</td>\n",
       "      <td>#F97B72</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Ethandiol binding site</td>\n",
       "      <td>260</td>\n",
       "      <td>260</td>\n",
       "      <td>#F97B72</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        feature start  end    color  size\n",
       "0  Aminomethyltransferase folate-binding domain    31  135  #E73F74   105\n",
       "1                   GcvT family signature motif   187  251  #80BA5A    65\n",
       "2                        Ethandiol binding site    23   25  #F97B72     3\n",
       "3                        Ethandiol binding site   107  109  #F97B72     3\n",
       "4                        Ethandiol binding site   116  117  #F97B72     2\n",
       "5                        Ethandiol binding site    16   16  #F97B72     1\n",
       "6                        Ethandiol binding site    38   38  #F97B72     1\n",
       "7                        Ethandiol binding site    50   50  #F97B72     1\n",
       "8                        Ethandiol binding site    99   99  #F97B72     1\n",
       "9                        Ethandiol binding site   260  260  #F97B72     1"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# for NGL\n",
    "ngl_feat_df[\"size\"] = ngl_feat_df.apply(lambda r: r[\"end\"] - r[\"start\"] + 1, axis=1)\n",
    "ngl_feat_df = ngl_feat_df[[\"feature\", \"start\", \"end\", \"color\", \"size\"]]\n",
    "ngl_feat_df = ngl_feat_df.sort_values(\"size\", ascending=False)\n",
    "ngl_feat_df = ngl_feat_df.reset_index(drop=True)  # json explicitly uses the indexes\n",
    "\n",
    "with open('feats.json', 'w') as outfile:   # for NGL\n",
    "    outfile.write(ngl_feat_df.to_json())\n",
    "ngl_feat_df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "916dbcbb3f70747c44a77c7bcd40155683ae19c65e1c03b4aa3499c5328201f1"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
