{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Get sample set for iModulon analytics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "imoddb_ros_tale_samples = {'A1 F112 I1 R1',\n",
    " 'A11 F83 I1 R1',\n",
    " 'A13 F87 I1 R1',\n",
    " 'A14 F27 I1 R1',\n",
    " 'A16 F28 I1 R1',\n",
    " 'A16 F32 I1 R1',\n",
    " 'A17 F104 I1 R1',\n",
    " 'A18 F24 I1 R1',\n",
    " 'A18 F36 I1 R1',\n",
    " 'A19 F91 I1 R1',\n",
    " 'A21 F97 I0 R1',\n",
    " 'A25 F111 I1 R1',\n",
    " 'A29 F119 I1 R1',\n",
    " 'A3 F76 I1 R1',\n",
    " 'A4 F237 I1 R1',\n",
    " 'A7 F103 I1 R1',\n",
    " 'A9 F77 I1 R1'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# raw_mut_df = pd.read_pickle('../all_aledb_raw_mut_df.pkl')\n",
    "# raw_mut_df = raw_mut_df[(raw_mut_df['exp']=='ROS_eAMP')]\n",
    "# raw_mut_df['afir'] = raw_mut_df.apply(lambda r: \"A\" + str(r.ale) + \" F\" + str(r.flask) + \" I\" + str(r.isolate) + \" R\" + str(r.tech_rep), axis=1)\n",
    "# imoddb_ros_tale_mut_df = raw_mut_df[raw_mut_df['afir'].isin(imoddb_ros_tale_samples)]\n",
    "# imoddb_oxyr_mut_df = imoddb_ros_tale_mut_df[imoddb_ros_tale_mut_df['Gene']=='oxyR']\n",
    "# set(imoddb_oxyr_mut_df.afir)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Get variants for mutation 1D and 3D analytics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "102"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exp</th>\n",
       "      <th>sample</th>\n",
       "      <th>Position</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>Details</th>\n",
       "      <th>Gene</th>\n",
       "      <th>ALEdb source</th>\n",
       "      <th>genetic</th>\n",
       "      <th>clean genes</th>\n",
       "      <th>bnums</th>\n",
       "      <th>genes of interest</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>466615</th>\n",
       "      <td>10mM_Lactose_R1</td>\n",
       "      <td>8 47 1 1</td>\n",
       "      <td>4159160</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T→A</td>\n",
       "      <td>L224Q (CTG→CAG)</td>\n",
       "      <td>oxyR</td>\n",
       "      <td>private</td>\n",
       "      <td>True</td>\n",
       "      <td>{oxyR}</td>\n",
       "      <td>{b3961}</td>\n",
       "      <td>{b3961}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>303555</th>\n",
       "      <td>10mM_Lactose_R2</td>\n",
       "      <td>9 104 1 2</td>\n",
       "      <td>4158805</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C→A</td>\n",
       "      <td>L106I (CTA→ATA)</td>\n",
       "      <td>oxyR</td>\n",
       "      <td>private</td>\n",
       "      <td>True</td>\n",
       "      <td>{oxyR}</td>\n",
       "      <td>{b3961}</td>\n",
       "      <td>{b3961}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>303558</th>\n",
       "      <td>10mM_Lactose_R2</td>\n",
       "      <td>9 46 1 2</td>\n",
       "      <td>4159070</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T→A</td>\n",
       "      <td>L194Q (CTG→CAG)</td>\n",
       "      <td>oxyR</td>\n",
       "      <td>private</td>\n",
       "      <td>True</td>\n",
       "      <td>{oxyR}</td>\n",
       "      <td>{b3961}</td>\n",
       "      <td>{b3961}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>366871</th>\n",
       "      <td>10mM_Lactose_R3</td>\n",
       "      <td>10 6 1 3</td>\n",
       "      <td>4159121</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C→A</td>\n",
       "      <td>A211D (GCC→GAC)</td>\n",
       "      <td>oxyR</td>\n",
       "      <td>private</td>\n",
       "      <td>True</td>\n",
       "      <td>{oxyR}</td>\n",
       "      <td>{b3961}</td>\n",
       "      <td>{b3961}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>353677</th>\n",
       "      <td>1mM_Gluc_10mM_Lactose_R1</td>\n",
       "      <td>4 49 1 1</td>\n",
       "      <td>4159087</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T→G</td>\n",
       "      <td>L200V (TTG→GTG)</td>\n",
       "      <td>oxyR</td>\n",
       "      <td>private</td>\n",
       "      <td>True</td>\n",
       "      <td>{oxyR}</td>\n",
       "      <td>{b3961}</td>\n",
       "      <td>{b3961}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             exp     sample  Position Mutation Type  \\\n",
       "466615           10mM_Lactose_R1   8 47 1 1   4159160           SNP   \n",
       "303555           10mM_Lactose_R2  9 104 1 2   4158805           SNP   \n",
       "303558           10mM_Lactose_R2   9 46 1 2   4159070           SNP   \n",
       "366871           10mM_Lactose_R3   10 6 1 3   4159121           SNP   \n",
       "353677  1mM_Gluc_10mM_Lactose_R1   4 49 1 1   4159087           SNP   \n",
       "\n",
       "       Sequence Change          Details  Gene ALEdb source  genetic  \\\n",
       "466615             T→A  L224Q (CTG→CAG)  oxyR      private     True   \n",
       "303555             C→A  L106I (CTA→ATA)  oxyR      private     True   \n",
       "303558             T→A  L194Q (CTG→CAG)  oxyR      private     True   \n",
       "366871             C→A  A211D (GCC→GAC)  oxyR      private     True   \n",
       "353677             T→G  L200V (TTG→GTG)  oxyR      private     True   \n",
       "\n",
       "       clean genes    bnums genes of interest  \n",
       "466615      {oxyR}  {b3961}           {b3961}  \n",
       "303555      {oxyR}  {b3961}           {b3961}  \n",
       "303558      {oxyR}  {b3961}           {b3961}  \n",
       "366871      {oxyR}  {b3961}           {b3961}  \n",
       "353677      {oxyR}  {b3961}           {b3961}  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "meta_muts = pd.read_pickle('./targ_mut_df.pkl')\n",
    "display(len(meta_muts), meta_muts.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exp</th>\n",
       "      <th>sample</th>\n",
       "      <th>Position</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>Details</th>\n",
       "      <th>Gene</th>\n",
       "      <th>ALEdb source</th>\n",
       "      <th>genetic</th>\n",
       "      <th>clean genes</th>\n",
       "      <th>bnums</th>\n",
       "      <th>genes of interest</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>132903</th>\n",
       "      <td>ROS_eAMP</td>\n",
       "      <td>5 84 0 1</td>\n",
       "      <td>4158922</td>\n",
       "      <td>DEL</td>\n",
       "      <td>Δ6 bp</td>\n",
       "      <td>coding (433‐438/918 nt)</td>\n",
       "      <td>oxyR</td>\n",
       "      <td>public</td>\n",
       "      <td>True</td>\n",
       "      <td>{oxyR}</td>\n",
       "      <td>{b3961}</td>\n",
       "      <td>{b3961}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>131820</th>\n",
       "      <td>ROS_eAMP</td>\n",
       "      <td>11 20 1 1</td>\n",
       "      <td>4159147</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C→G</td>\n",
       "      <td>R220G (CGC→GGC)</td>\n",
       "      <td>oxyR</td>\n",
       "      <td>public</td>\n",
       "      <td>True</td>\n",
       "      <td>{oxyR}</td>\n",
       "      <td>{b3961}</td>\n",
       "      <td>{b3961}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>131986</th>\n",
       "      <td>ROS_eAMP</td>\n",
       "      <td>13 87 1 1</td>\n",
       "      <td>4159126</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G→A</td>\n",
       "      <td>A213T (GCG→ACG)</td>\n",
       "      <td>oxyR</td>\n",
       "      <td>public</td>\n",
       "      <td>True</td>\n",
       "      <td>{oxyR}</td>\n",
       "      <td>{b3961}</td>\n",
       "      <td>{b3961}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132121</th>\n",
       "      <td>ROS_eAMP</td>\n",
       "      <td>17 104 0 1</td>\n",
       "      <td>4158785</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C→T</td>\n",
       "      <td>P99L (CCC→CTC)</td>\n",
       "      <td>oxyR</td>\n",
       "      <td>public</td>\n",
       "      <td>True</td>\n",
       "      <td>{oxyR}</td>\n",
       "      <td>{b3961}</td>\n",
       "      <td>{b3961}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132385</th>\n",
       "      <td>ROS_eAMP</td>\n",
       "      <td>21 97 0 1</td>\n",
       "      <td>4159079</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G→T</td>\n",
       "      <td>G197V (GGT→GTT)</td>\n",
       "      <td>oxyR</td>\n",
       "      <td>public</td>\n",
       "      <td>True</td>\n",
       "      <td>{oxyR}</td>\n",
       "      <td>{b3961}</td>\n",
       "      <td>{b3961}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             exp      sample  Position Mutation Type Sequence Change  \\\n",
       "132903  ROS_eAMP    5 84 0 1   4158922           DEL           Δ6 bp   \n",
       "131820  ROS_eAMP   11 20 1 1   4159147           SNP             C→G   \n",
       "131986  ROS_eAMP   13 87 1 1   4159126           SNP             G→A   \n",
       "132121  ROS_eAMP  17 104 0 1   4158785           SNP             C→T   \n",
       "132385  ROS_eAMP   21 97 0 1   4159079           SNP             G→T   \n",
       "\n",
       "                        Details  Gene ALEdb source  genetic clean genes  \\\n",
       "132903  coding (433‐438/918 nt)  oxyR       public     True      {oxyR}   \n",
       "131820          R220G (CGC→GGC)  oxyR       public     True      {oxyR}   \n",
       "131986          A213T (GCG→ACG)  oxyR       public     True      {oxyR}   \n",
       "132121           P99L (CCC→CTC)  oxyR       public     True      {oxyR}   \n",
       "132385          G197V (GGT→GTT)  oxyR       public     True      {oxyR}   \n",
       "\n",
       "          bnums genes of interest  \n",
       "132903  {b3961}           {b3961}  \n",
       "131820  {b3961}           {b3961}  \n",
       "131986  {b3961}           {b3961}  \n",
       "132121  {b3961}           {b3961}  \n",
       "132385  {b3961}           {b3961}  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta_muts[meta_muts.exp=='ROS_eAMP']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# !!! Temporary; need to reprocess all ALE experiment data for this project with the now removed ERS project experiments which were being duplicated by the newer ROS_eAMP ALE project/experiment.\n",
    "\n",
    "meta_muts = meta_muts[meta_muts.exp != 'GLU_16321-M9-NC_000913_3gb-stationary-37-Glucose2-NH4Cl1-KH2PO43-MgSO4024-O2']\n",
    "meta_muts = meta_muts[meta_muts.exp != 'GLU_18361-M9-NC_000913_3gb-stationary-37-Glucose2-NH4Cl1-KH2PO43-MgSO4024-O2']\n",
    "meta_muts = meta_muts[meta_muts.exp != 'GLU_4721-M9-NC_000913_3gb-stationary-37-Glucose2-NH4Cl1-KH2PO43-MgSO4024-O2']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, sys\n",
    "module_path = os.path.abspath(os.path.join('..'))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "from aledbmutil.mut import get_mut_size, get_mob_size, is_genetic_mut, is_coding_mut, predict_mutation_effect_on_feature, get_SNP_aa_pos, get_DEL_INS_MOB_aa_start_pos, get_DEL_AA_range\n",
    "\n",
    "\n",
    "meta_muts[\"genetic\"] = meta_muts[\"Details\"].apply(is_genetic_mut)\n",
    "meta_muts[\"coding\"] = meta_muts[\"Details\"].apply(is_coding_mut)\n",
    "meta_muts[\"mutation size\"] = meta_muts.apply(get_mut_size, axis=1)  # get_mut_size currently doesn't return MOB mut sizes. Isn't something currently necessary."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exp</th>\n",
       "      <th>sample</th>\n",
       "      <th>Position</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>Details</th>\n",
       "      <th>Gene</th>\n",
       "      <th>ALEdb source</th>\n",
       "      <th>genetic</th>\n",
       "      <th>clean genes</th>\n",
       "      <th>bnums</th>\n",
       "      <th>genes of interest</th>\n",
       "      <th>coding</th>\n",
       "      <th>mutation size</th>\n",
       "      <th>AA range</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>466615</th>\n",
       "      <td>10mM_Lactose_R1</td>\n",
       "      <td>8 47 1 1</td>\n",
       "      <td>4159160</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T→A</td>\n",
       "      <td>L224Q (CTG→CAG)</td>\n",
       "      <td>oxyR</td>\n",
       "      <td>private</td>\n",
       "      <td>True</td>\n",
       "      <td>{oxyR}</td>\n",
       "      <td>{b3961}</td>\n",
       "      <td>{b3961}</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>(224, 224)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>303555</th>\n",
       "      <td>10mM_Lactose_R2</td>\n",
       "      <td>9 104 1 2</td>\n",
       "      <td>4158805</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C→A</td>\n",
       "      <td>L106I (CTA→ATA)</td>\n",
       "      <td>oxyR</td>\n",
       "      <td>private</td>\n",
       "      <td>True</td>\n",
       "      <td>{oxyR}</td>\n",
       "      <td>{b3961}</td>\n",
       "      <td>{b3961}</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>(106, 106)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>303558</th>\n",
       "      <td>10mM_Lactose_R2</td>\n",
       "      <td>9 46 1 2</td>\n",
       "      <td>4159070</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T→A</td>\n",
       "      <td>L194Q (CTG→CAG)</td>\n",
       "      <td>oxyR</td>\n",
       "      <td>private</td>\n",
       "      <td>True</td>\n",
       "      <td>{oxyR}</td>\n",
       "      <td>{b3961}</td>\n",
       "      <td>{b3961}</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>(194, 194)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>366871</th>\n",
       "      <td>10mM_Lactose_R3</td>\n",
       "      <td>10 6 1 3</td>\n",
       "      <td>4159121</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C→A</td>\n",
       "      <td>A211D (GCC→GAC)</td>\n",
       "      <td>oxyR</td>\n",
       "      <td>private</td>\n",
       "      <td>True</td>\n",
       "      <td>{oxyR}</td>\n",
       "      <td>{b3961}</td>\n",
       "      <td>{b3961}</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>(211, 211)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>353677</th>\n",
       "      <td>1mM_Gluc_10mM_Lactose_R1</td>\n",
       "      <td>4 49 1 1</td>\n",
       "      <td>4159087</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T→G</td>\n",
       "      <td>L200V (TTG→GTG)</td>\n",
       "      <td>oxyR</td>\n",
       "      <td>private</td>\n",
       "      <td>True</td>\n",
       "      <td>{oxyR}</td>\n",
       "      <td>{b3961}</td>\n",
       "      <td>{b3961}</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>(200, 200)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             exp     sample  Position Mutation Type  \\\n",
       "466615           10mM_Lactose_R1   8 47 1 1   4159160           SNP   \n",
       "303555           10mM_Lactose_R2  9 104 1 2   4158805           SNP   \n",
       "303558           10mM_Lactose_R2   9 46 1 2   4159070           SNP   \n",
       "366871           10mM_Lactose_R3   10 6 1 3   4159121           SNP   \n",
       "353677  1mM_Gluc_10mM_Lactose_R1   4 49 1 1   4159087           SNP   \n",
       "\n",
       "       Sequence Change          Details  Gene ALEdb source  genetic  \\\n",
       "466615             T→A  L224Q (CTG→CAG)  oxyR      private     True   \n",
       "303555             C→A  L106I (CTA→ATA)  oxyR      private     True   \n",
       "303558             T→A  L194Q (CTG→CAG)  oxyR      private     True   \n",
       "366871             C→A  A211D (GCC→GAC)  oxyR      private     True   \n",
       "353677             T→G  L200V (TTG→GTG)  oxyR      private     True   \n",
       "\n",
       "       clean genes    bnums genes of interest  coding  mutation size  \\\n",
       "466615      {oxyR}  {b3961}           {b3961}    True              1   \n",
       "303555      {oxyR}  {b3961}           {b3961}    True              1   \n",
       "303558      {oxyR}  {b3961}           {b3961}    True              1   \n",
       "366871      {oxyR}  {b3961}           {b3961}    True              1   \n",
       "353677      {oxyR}  {b3961}           {b3961}    True              1   \n",
       "\n",
       "          AA range  \n",
       "466615  (224, 224)  \n",
       "303555  (106, 106)  \n",
       "303558  (194, 194)  \n",
       "366871  (211, 211)  \n",
       "353677  (200, 200)  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from aledbmutil.mut import get_SNP_aa_pos, get_DEL_AA_range\n",
    "\n",
    "def _get_SNP_AA_range(mut_details_str):\n",
    "    SNP_AA_range = ()\n",
    "    AA_pos = get_SNP_aa_pos(mut_details_str)\n",
    "    SNP_AA_range = (AA_pos, AA_pos)\n",
    "    return SNP_AA_range\n",
    "\n",
    "meta_muts[\"AA range\"] = None\n",
    "meta_muts[\"AA range\"] = meta_muts.apply(lambda r:\n",
    "                                        _get_SNP_AA_range(r[\"Details\"]) if ((r[\"Mutation Type\"] == \"SNP\") & (r[\"coding\"]==1)) \\\n",
    "                                        else (get_DEL_AA_range(r[\"Details\"]) if ((r[\"Mutation Type\"] == \"DEL\") & (r[\"coding\"]==1)) \\\n",
    "                                        else r[\"AA range\"]),\n",
    "                                        axis=1)\n",
    "meta_muts[meta_muts[\"Mutation Type\"].isin([\"SNP\"])].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exp</th>\n",
       "      <th>sample</th>\n",
       "      <th>Position</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>Details</th>\n",
       "      <th>Gene</th>\n",
       "      <th>ALEdb source</th>\n",
       "      <th>genetic</th>\n",
       "      <th>clean genes</th>\n",
       "      <th>bnums</th>\n",
       "      <th>genes of interest</th>\n",
       "      <th>coding</th>\n",
       "      <th>mutation size</th>\n",
       "      <th>AA range</th>\n",
       "      <th>AA ref seq</th>\n",
       "      <th>AA seq change</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>466615</th>\n",
       "      <td>10mM_Lactose_R1</td>\n",
       "      <td>8 47 1 1</td>\n",
       "      <td>4159160</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T→A</td>\n",
       "      <td>L224Q (CTG→CAG)</td>\n",
       "      <td>oxyR</td>\n",
       "      <td>private</td>\n",
       "      <td>True</td>\n",
       "      <td>{oxyR}</td>\n",
       "      <td>{b3961}</td>\n",
       "      <td>{b3961}</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>(224, 224)</td>\n",
       "      <td>L</td>\n",
       "      <td>Q</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>303555</th>\n",
       "      <td>10mM_Lactose_R2</td>\n",
       "      <td>9 104 1 2</td>\n",
       "      <td>4158805</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C→A</td>\n",
       "      <td>L106I (CTA→ATA)</td>\n",
       "      <td>oxyR</td>\n",
       "      <td>private</td>\n",
       "      <td>True</td>\n",
       "      <td>{oxyR}</td>\n",
       "      <td>{b3961}</td>\n",
       "      <td>{b3961}</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>(106, 106)</td>\n",
       "      <td>L</td>\n",
       "      <td>I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>303558</th>\n",
       "      <td>10mM_Lactose_R2</td>\n",
       "      <td>9 46 1 2</td>\n",
       "      <td>4159070</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T→A</td>\n",
       "      <td>L194Q (CTG→CAG)</td>\n",
       "      <td>oxyR</td>\n",
       "      <td>private</td>\n",
       "      <td>True</td>\n",
       "      <td>{oxyR}</td>\n",
       "      <td>{b3961}</td>\n",
       "      <td>{b3961}</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>(194, 194)</td>\n",
       "      <td>L</td>\n",
       "      <td>Q</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>366871</th>\n",
       "      <td>10mM_Lactose_R3</td>\n",
       "      <td>10 6 1 3</td>\n",
       "      <td>4159121</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C→A</td>\n",
       "      <td>A211D (GCC→GAC)</td>\n",
       "      <td>oxyR</td>\n",
       "      <td>private</td>\n",
       "      <td>True</td>\n",
       "      <td>{oxyR}</td>\n",
       "      <td>{b3961}</td>\n",
       "      <td>{b3961}</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>(211, 211)</td>\n",
       "      <td>A</td>\n",
       "      <td>D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>353677</th>\n",
       "      <td>1mM_Gluc_10mM_Lactose_R1</td>\n",
       "      <td>4 49 1 1</td>\n",
       "      <td>4159087</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T→G</td>\n",
       "      <td>L200V (TTG→GTG)</td>\n",
       "      <td>oxyR</td>\n",
       "      <td>private</td>\n",
       "      <td>True</td>\n",
       "      <td>{oxyR}</td>\n",
       "      <td>{b3961}</td>\n",
       "      <td>{b3961}</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>(200, 200)</td>\n",
       "      <td>L</td>\n",
       "      <td>V</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             exp     sample  Position Mutation Type  \\\n",
       "466615           10mM_Lactose_R1   8 47 1 1   4159160           SNP   \n",
       "303555           10mM_Lactose_R2  9 104 1 2   4158805           SNP   \n",
       "303558           10mM_Lactose_R2   9 46 1 2   4159070           SNP   \n",
       "366871           10mM_Lactose_R3   10 6 1 3   4159121           SNP   \n",
       "353677  1mM_Gluc_10mM_Lactose_R1   4 49 1 1   4159087           SNP   \n",
       "\n",
       "       Sequence Change          Details  Gene ALEdb source  genetic  \\\n",
       "466615             T→A  L224Q (CTG→CAG)  oxyR      private     True   \n",
       "303555             C→A  L106I (CTA→ATA)  oxyR      private     True   \n",
       "303558             T→A  L194Q (CTG→CAG)  oxyR      private     True   \n",
       "366871             C→A  A211D (GCC→GAC)  oxyR      private     True   \n",
       "353677             T→G  L200V (TTG→GTG)  oxyR      private     True   \n",
       "\n",
       "       clean genes    bnums genes of interest  coding  mutation size  \\\n",
       "466615      {oxyR}  {b3961}           {b3961}    True              1   \n",
       "303555      {oxyR}  {b3961}           {b3961}    True              1   \n",
       "303558      {oxyR}  {b3961}           {b3961}    True              1   \n",
       "366871      {oxyR}  {b3961}           {b3961}    True              1   \n",
       "353677      {oxyR}  {b3961}           {b3961}    True              1   \n",
       "\n",
       "          AA range AA ref seq AA seq change  \n",
       "466615  (224, 224)          L             Q  \n",
       "303555  (106, 106)          L             I  \n",
       "303558  (194, 194)          L             Q  \n",
       "366871  (211, 211)          A             D  \n",
       "353677  (200, 200)          L             V  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# currently only tracking seq changes with SNPs for ALEdb mutations for later processing with AA property change trends\n",
    "meta_muts['AA ref seq'] = meta_muts.apply(lambda m: m['Details'][0] if m[\"Mutation Type\"]=='SNP' else '', axis=1)\n",
    "meta_muts['AA seq change'] = meta_muts.apply(lambda m: m['Details'].split(' (')[0][-1] if m[\"Mutation Type\"]=='SNP' else '', axis=1)\n",
    "meta_muts.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "exp_sel_pres_d = {\n",
    "    'C321.∆A': \"recoded stop codon\",\n",
    "    'C321.∆A.earlyfix': \"recoded stop codon\",\n",
    "    'C321_A_earlyfix': \"recoded stop codon\",\n",
    "    'Iron-ALE': \"FeSO4\",\n",
    "    'JCL260__isobutanol_production_host-M9-MG1655gbk': 'isobutanol',\n",
    "    'ROS_eAMP': \"paraquat\",  # starting strains were also pre-evolved on glucose\n",
    "    'pgiPAE': 'Pyrobaculum aerophilum pgi',\n",
    "    'pgiBME': 'Brucella melitensis pgi',\n",
    "    'tpiHSA': 'Homo sapiens tpi',\n",
    "    'tpiPAE': 'Pyrobaculum aerophilum tpi',\n",
    "    'tpiVCH': 'Vibrio cholerae tpi',\n",
    " }\n",
    "\n",
    "\n",
    "def _get_sel_pres(mut):\n",
    "    sel_pres = ''\n",
    "    if mut['ALEdb source'] == 'private':\n",
    "        sel_pres = 'unpublished'\n",
    "    else:\n",
    "        if mut.exp in exp_sel_pres_d.keys():\n",
    "            sel_pres = exp_sel_pres_d[mut.exp]        \n",
    "    return sel_pres\n",
    "\n",
    "meta_muts['selection pressure'] = meta_muts.apply(lambda r: _get_sel_pres(r), axis=1)\n",
    "assert('' not in set(meta_muts['selection pressure']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exp</th>\n",
       "      <th>sample</th>\n",
       "      <th>Position</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Sequence Change</th>\n",
       "      <th>Details</th>\n",
       "      <th>Gene</th>\n",
       "      <th>ALEdb source</th>\n",
       "      <th>genetic</th>\n",
       "      <th>clean genes</th>\n",
       "      <th>bnums</th>\n",
       "      <th>genes of interest</th>\n",
       "      <th>coding</th>\n",
       "      <th>mutation size</th>\n",
       "      <th>AA range</th>\n",
       "      <th>AA ref seq</th>\n",
       "      <th>AA seq change</th>\n",
       "      <th>selection pressure</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>29421</th>\n",
       "      <td>C321.∆A.earlyfix</td>\n",
       "      <td>1 1 1 1</td>\n",
       "      <td>4158776</td>\n",
       "      <td>DEL</td>\n",
       "      <td>Δ1 bp</td>\n",
       "      <td>coding (287/918 nt)</td>\n",
       "      <td>oxyR</td>\n",
       "      <td>public</td>\n",
       "      <td>True</td>\n",
       "      <td>{oxyR}</td>\n",
       "      <td>{b3961}</td>\n",
       "      <td>{b3961}</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>(96, 96)</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>recoded stop codon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>305292</th>\n",
       "      <td>C321_A_earlyfix</td>\n",
       "      <td>1 1 1 1</td>\n",
       "      <td>4158776</td>\n",
       "      <td>DEL</td>\n",
       "      <td>Δ1 bp</td>\n",
       "      <td>coding (287/918 nt)</td>\n",
       "      <td>oxyR</td>\n",
       "      <td>private</td>\n",
       "      <td>True</td>\n",
       "      <td>{oxyR}</td>\n",
       "      <td>{b3961}</td>\n",
       "      <td>{b3961}</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>(96, 96)</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>unpublished</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132903</th>\n",
       "      <td>ROS_eAMP</td>\n",
       "      <td>5 84 0 1</td>\n",
       "      <td>4158922</td>\n",
       "      <td>DEL</td>\n",
       "      <td>Δ6 bp</td>\n",
       "      <td>coding (433‐438/918 nt)</td>\n",
       "      <td>oxyR</td>\n",
       "      <td>public</td>\n",
       "      <td>True</td>\n",
       "      <td>{oxyR}</td>\n",
       "      <td>{b3961}</td>\n",
       "      <td>{b3961}</td>\n",
       "      <td>True</td>\n",
       "      <td>6</td>\n",
       "      <td>(145, 146)</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>paraquat</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     exp    sample  Position Mutation Type Sequence Change  \\\n",
       "29421   C321.∆A.earlyfix   1 1 1 1   4158776           DEL           Δ1 bp   \n",
       "305292   C321_A_earlyfix   1 1 1 1   4158776           DEL           Δ1 bp   \n",
       "132903          ROS_eAMP  5 84 0 1   4158922           DEL           Δ6 bp   \n",
       "\n",
       "                        Details  Gene ALEdb source  genetic clean genes  \\\n",
       "29421       coding (287/918 nt)  oxyR       public     True      {oxyR}   \n",
       "305292      coding (287/918 nt)  oxyR      private     True      {oxyR}   \n",
       "132903  coding (433‐438/918 nt)  oxyR       public     True      {oxyR}   \n",
       "\n",
       "          bnums genes of interest  coding  mutation size    AA range  \\\n",
       "29421   {b3961}           {b3961}    True              1    (96, 96)   \n",
       "305292  {b3961}           {b3961}    True              1    (96, 96)   \n",
       "132903  {b3961}           {b3961}    True              6  (145, 146)   \n",
       "\n",
       "       AA ref seq AA seq change  selection pressure  \n",
       "29421                            recoded stop codon  \n",
       "305292                                  unpublished  \n",
       "132903                                     paraquat  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "98"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AA pos</th>\n",
       "      <th>mutation type</th>\n",
       "      <th>mutation effect</th>\n",
       "      <th>mutation source</th>\n",
       "      <th>AA ref seq</th>\n",
       "      <th>AA seq change</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>96</td>\n",
       "      <td>DEL</td>\n",
       "      <td>truncation</td>\n",
       "      <td>recoded stop codon</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>96</td>\n",
       "      <td>DEL</td>\n",
       "      <td>truncation</td>\n",
       "      <td>unpublished</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>145</td>\n",
       "      <td>DEL</td>\n",
       "      <td>deletion</td>\n",
       "      <td>paraquat</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>146</td>\n",
       "      <td>DEL</td>\n",
       "      <td>deletion</td>\n",
       "      <td>paraquat</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>224</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>L</td>\n",
       "      <td>Q</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   AA pos mutation type             mutation effect     mutation source  \\\n",
       "0      96           DEL                  truncation  recoded stop codon   \n",
       "1      96           DEL                  truncation         unpublished   \n",
       "2     145           DEL                    deletion            paraquat   \n",
       "3     146           DEL                    deletion            paraquat   \n",
       "4     224           SUB  nonsynonymous substitution         unpublished   \n",
       "\n",
       "  AA ref seq AA seq change  \n",
       "0                           \n",
       "1                           \n",
       "2                           \n",
       "3                           \n",
       "4          L             Q  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# TODO: fix warnings with append\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "\n",
    "\n",
    "aa_var_df = pd.DataFrame()\n",
    "\n",
    "from aledbmutil.mut import get_mut_size  #, get_DEL_INS_MOB_aa_start_pos\n",
    "\n",
    "\n",
    "for mt, mut_type_mut_df in meta_muts.groupby(\"Mutation Type\"):\n",
    "    assert mt != \"MOB\", 'Currently not handilng MOB. The below INS logic might be able to handle this.'\n",
    "    if mt == \"SNP\":\n",
    "        for _, m in mut_type_mut_df.iterrows():\n",
    "            mut_eff = \"\"\n",
    "            details_str = m[\"Details\"]\n",
    "            details_str = details_str[0:details_str.find(\" (\")]\n",
    "            \n",
    "            if details_str[-1] == '*':  # Could have them show the effected downstream positions like with ACS SynBio paper. \n",
    "                mut_eff = \"truncation\"\n",
    "            elif details_str[0] == details_str[-1]:\n",
    "                mut_eff = \"synonymous substitution\"\n",
    "            elif details_str[0] != details_str[-1]:\n",
    "                mut_eff = \"nonsynonymous substitution\"\n",
    "            else:\n",
    "                mut_eff = \"unknown\"\n",
    "            \n",
    "            aa_pos = int(details_str[1:-1])\n",
    "            aa_var_df= aa_var_df.append({\n",
    "                \"AA pos\": aa_pos,\n",
    "                \"mutation type\": \"SUB\",\n",
    "                \"mutation effect\": mut_eff,\n",
    "                \"mutation source\": m['selection pressure'],\n",
    "                'AA ref seq': m['AA ref seq'],\n",
    "                'AA seq change': m['AA seq change'],\n",
    "            }, ignore_index=True)\n",
    "    if (mt == 'DEL'):  # Unravel (unless truncation) since downstream analysis just looks at individual mutated AAs no matter if they come from the same allele.\n",
    "        display(mut_type_mut_df)\n",
    "        for _, m in mut_type_mut_df.iterrows():\n",
    "            mut_eff = \"deletion\"\n",
    "            mut_size = m['mutation size']\n",
    "            if mut_size % 3 != 0:  # Don't want to record all positions involved in the truncation, just the first.\n",
    "                aa_var_df= aa_var_df.append({\n",
    "                    \"AA pos\": m['AA range'][0],\n",
    "                    \"mutation type\": mt,\n",
    "                    \"mutation effect\": \"truncation\",\n",
    "                    \"mutation source\": m['selection pressure'],\n",
    "                    'AA ref seq': '',\n",
    "                    'AA seq change': '',\n",
    "                },ignore_index=True)\n",
    "            else:                \n",
    "                for aa_pos in range(m['AA range'][0], m['AA range'][1]+1):\n",
    "                    aa_var_df= aa_var_df.append({\n",
    "                        \"AA pos\": aa_pos,\n",
    "                        \"mutation type\": mt,\n",
    "                        \"mutation effect\": mut_eff,\n",
    "                        \"mutation source\": m['selection pressure'],\n",
    "                        'AA ref seq': '',\n",
    "                        'AA seq change': '',\n",
    "                    },ignore_index=True)\n",
    "    if (mt == 'INS'):\n",
    "        for _, m in mut_type_mut_df.iterrows():\n",
    "            mut_eff = \"insertion\"  # currently assuming only insertions remain\n",
    "            mut_size = m['mutation size']\n",
    "            if mut_size % 3 != 0:\n",
    "                mut_eff = \"truncation\"\n",
    "            aa_var_df= aa_var_df.append({\n",
    "                \"AA pos\": get_DEL_INS_MOB_aa_start_pos(m[\"Details\"]),\n",
    "                \"mutation type\": mt,\n",
    "                \"mutation effect\": mut_eff,\n",
    "                \"mutation source\": m['selection pressure'],\n",
    "                'AA ref seq': '',\n",
    "                'AA seq change': '',\n",
    "            },ignore_index=True)\n",
    "    if (mt == 'SUB'):  # Unravel since downstream analysis just looks at individual mutated AAs no matter if they come from the same allele.\n",
    "        for _, m in mut_type_mut_df.iterrows():\n",
    "            for aa_pos in range(m['AA range'][0], m['AA range'][1]+1):\n",
    "                aa_var_df= aa_var_df.append({\n",
    "                    \"AA pos\": aa_pos,\n",
    "                    \"mutation type\": mt,\n",
    "                    \"mutation effect\": \"nonsynonymous substitution\",\n",
    "                    \"mutation source\": m['selection pressure'],\n",
    "                    'AA ref seq': '',\n",
    "                    'AA seq change': '',\n",
    "                },ignore_index=True)\n",
    "\n",
    "\n",
    "display(len(aa_var_df), aa_var_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AA pos</th>\n",
       "      <th>mutation type</th>\n",
       "      <th>mutation effect</th>\n",
       "      <th>mutation source</th>\n",
       "      <th>AA ref seq</th>\n",
       "      <th>AA seq change</th>\n",
       "      <th>mutation count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>8</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>recoded stop codon</td>\n",
       "      <td>Y</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>13</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>recoded stop codon</td>\n",
       "      <td>A</td>\n",
       "      <td>V</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>29</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>recoded stop codon</td>\n",
       "      <td>Q</td>\n",
       "      <td>R</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>29</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>Q</td>\n",
       "      <td>R</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>31</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>recoded stop codon</td>\n",
       "      <td>T</td>\n",
       "      <td>M</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   AA pos mutation type             mutation effect     mutation source  \\\n",
       "0       8           SUB  nonsynonymous substitution  recoded stop codon   \n",
       "1      13           SUB  nonsynonymous substitution  recoded stop codon   \n",
       "2      29           SUB  nonsynonymous substitution  recoded stop codon   \n",
       "3      29           SUB  nonsynonymous substitution         unpublished   \n",
       "4      31           SUB  nonsynonymous substitution  recoded stop codon   \n",
       "\n",
       "  AA ref seq AA seq change  mutation count  \n",
       "0          Y             C               1  \n",
       "1          A             V               1  \n",
       "2          Q             R               1  \n",
       "3          Q             R               1  \n",
       "4          T             M               1  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aa_var_cnt_df = pd.DataFrame()\n",
    "gcols = ['AA pos', 'mutation type', 'mutation effect', 'mutation source', 'AA ref seq', 'AA seq change']\n",
    "for g, gdf in aa_var_df.groupby(gcols):  # Essentially just finding duplicates\n",
    "    d = dict(zip(gcols, g))\n",
    "    d['mutation count'] = len(gdf)\n",
    "    aa_var_cnt_df = aa_var_cnt_df.append(d, ignore_index=True)\n",
    "aa_var_cnt_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AA pos</th>\n",
       "      <th>mutation type</th>\n",
       "      <th>mutation effect</th>\n",
       "      <th>mutation source</th>\n",
       "      <th>AA ref seq</th>\n",
       "      <th>AA seq change</th>\n",
       "      <th>mutation count</th>\n",
       "      <th>mutation effect color</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>8</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>recoded stop codon</td>\n",
       "      <td>Y</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>13</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>recoded stop codon</td>\n",
       "      <td>A</td>\n",
       "      <td>V</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>29</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>recoded stop codon</td>\n",
       "      <td>Q</td>\n",
       "      <td>R</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>29</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>Q</td>\n",
       "      <td>R</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>31</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>recoded stop codon</td>\n",
       "      <td>T</td>\n",
       "      <td>M</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   AA pos mutation type             mutation effect     mutation source  \\\n",
       "0       8           SUB  nonsynonymous substitution  recoded stop codon   \n",
       "1      13           SUB  nonsynonymous substitution  recoded stop codon   \n",
       "2      29           SUB  nonsynonymous substitution  recoded stop codon   \n",
       "3      29           SUB  nonsynonymous substitution         unpublished   \n",
       "4      31           SUB  nonsynonymous substitution  recoded stop codon   \n",
       "\n",
       "  AA ref seq AA seq change  mutation count mutation effect color  \n",
       "0          Y             C               1               #DD8452  \n",
       "1          A             V               1               #DD8452  \n",
       "2          Q             R               1               #DD8452  \n",
       "3          Q             R               1               #DD8452  \n",
       "4          T             M               1               #DD8452  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os, sys\n",
    "module_path = os.path.abspath(os.path.join('..'))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "import const\n",
    "\n",
    "\n",
    "aa_var_cnt_df['mutation effect color'] = aa_var_cnt_df['mutation effect'].apply(lambda e: const.MUTATION_TYPE_COLOR_PALETTE[e])\n",
    "aa_var_cnt_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Brucella melitensis pgi',\n",
       " 'FeSO4',\n",
       " 'Homo sapiens tpi',\n",
       " 'Pyrobaculum aerophilum pgi',\n",
       " 'Pyrobaculum aerophilum tpi',\n",
       " 'Vibrio cholerae tpi',\n",
       " 'paraquat',\n",
       " 'recoded stop codon',\n",
       " 'unpublished'}"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(aa_var_cnt_df['mutation source'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AA pos</th>\n",
       "      <th>mutation type</th>\n",
       "      <th>mutation effect</th>\n",
       "      <th>mutation source</th>\n",
       "      <th>AA ref seq</th>\n",
       "      <th>AA seq change</th>\n",
       "      <th>mutation count</th>\n",
       "      <th>mutation effect color</th>\n",
       "      <th>selection pressure color</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>8</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>recoded stop codon</td>\n",
       "      <td>Y</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#A8786E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>13</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>recoded stop codon</td>\n",
       "      <td>A</td>\n",
       "      <td>V</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#A8786E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>29</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>recoded stop codon</td>\n",
       "      <td>Q</td>\n",
       "      <td>R</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#A8786E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>29</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>Q</td>\n",
       "      <td>R</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#AD8BC9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>31</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>recoded stop codon</td>\n",
       "      <td>T</td>\n",
       "      <td>M</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#A8786E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>234</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>G</td>\n",
       "      <td>V</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#AD8BC9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>236</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>Pyrobaculum aerophilum tpi</td>\n",
       "      <td>G</td>\n",
       "      <td>E</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#ED665D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>238</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>Homo sapiens tpi</td>\n",
       "      <td>T</td>\n",
       "      <td>N</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#FF9E4A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>246</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>P</td>\n",
       "      <td>Q</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#AD8BC9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>279</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>R</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#AD8BC9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>71 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    AA pos mutation type             mutation effect  \\\n",
       "0        8           SUB  nonsynonymous substitution   \n",
       "1       13           SUB  nonsynonymous substitution   \n",
       "2       29           SUB  nonsynonymous substitution   \n",
       "3       29           SUB  nonsynonymous substitution   \n",
       "4       31           SUB  nonsynonymous substitution   \n",
       "..     ...           ...                         ...   \n",
       "66     234           SUB  nonsynonymous substitution   \n",
       "67     236           SUB  nonsynonymous substitution   \n",
       "68     238           SUB  nonsynonymous substitution   \n",
       "69     246           SUB  nonsynonymous substitution   \n",
       "70     279           SUB  nonsynonymous substitution   \n",
       "\n",
       "               mutation source AA ref seq AA seq change  mutation count  \\\n",
       "0           recoded stop codon          Y             C               1   \n",
       "1           recoded stop codon          A             V               1   \n",
       "2           recoded stop codon          Q             R               1   \n",
       "3                  unpublished          Q             R               1   \n",
       "4           recoded stop codon          T             M               1   \n",
       "..                         ...        ...           ...             ...   \n",
       "66                 unpublished          G             V               1   \n",
       "67  Pyrobaculum aerophilum tpi          G             E               1   \n",
       "68            Homo sapiens tpi          T             N               1   \n",
       "69                 unpublished          P             Q               1   \n",
       "70                 unpublished          R             C               1   \n",
       "\n",
       "   mutation effect color selection pressure color  \n",
       "0                #DD8452                  #A8786E  \n",
       "1                #DD8452                  #A8786E  \n",
       "2                #DD8452                  #A8786E  \n",
       "3                #DD8452                  #AD8BC9  \n",
       "4                #DD8452                  #A8786E  \n",
       "..                   ...                      ...  \n",
       "66               #DD8452                  #AD8BC9  \n",
       "67               #DD8452                  #ED665D  \n",
       "68               #DD8452                  #FF9E4A  \n",
       "69               #DD8452                  #AD8BC9  \n",
       "70               #DD8452                  #AD8BC9  \n",
       "\n",
       "[71 rows x 9 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sel_pres_clr_d = {\n",
    "    'FeSO4': '#729ECE',\n",
    "    'Homo sapiens tpi': '#FF9E4A',\n",
    "    'Pyrobaculum aerophilum pgi': '#67BF5C',\n",
    "    'Brucella melitensis pgi': '#797979',\n",
    "    'Pyrobaculum aerophilum tpi': '#ED665D',\n",
    "    'Vibrio cholerae tpi': '#D5BB67',\n",
    "    'isobutanol': '#ED97CA',\n",
    "    'paraquat': '#6DCCDA',\n",
    "    'recoded stop codon': '#A8786E',\n",
    "    'unpublished': '#AD8BC9'\n",
    " }\n",
    "\n",
    "aa_var_cnt_df['selection pressure color'] = aa_var_cnt_df['mutation source'].apply(lambda e: sel_pres_clr_d[e])\n",
    "aa_var_cnt_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AA pos</th>\n",
       "      <th>mutation type</th>\n",
       "      <th>mutation effect</th>\n",
       "      <th>mutation source</th>\n",
       "      <th>AA ref seq</th>\n",
       "      <th>AA seq change</th>\n",
       "      <th>mutation count</th>\n",
       "      <th>mutation effect color</th>\n",
       "      <th>selection pressure color</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>8</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>recoded stop codon</td>\n",
       "      <td>Y</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#A8786E</td>\n",
       "      <td>SUB Y8C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>13</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>recoded stop codon</td>\n",
       "      <td>A</td>\n",
       "      <td>V</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#A8786E</td>\n",
       "      <td>SUB A13V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>29</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>recoded stop codon</td>\n",
       "      <td>Q</td>\n",
       "      <td>R</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#A8786E</td>\n",
       "      <td>SUB Q29R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>29</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>Q</td>\n",
       "      <td>R</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#AD8BC9</td>\n",
       "      <td>SUB Q29R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>31</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>recoded stop codon</td>\n",
       "      <td>T</td>\n",
       "      <td>M</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#A8786E</td>\n",
       "      <td>SUB T31M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>234</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>G</td>\n",
       "      <td>V</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#AD8BC9</td>\n",
       "      <td>SUB G234V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>236</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>Pyrobaculum aerophilum tpi</td>\n",
       "      <td>G</td>\n",
       "      <td>E</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#ED665D</td>\n",
       "      <td>SUB G236E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>238</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>Homo sapiens tpi</td>\n",
       "      <td>T</td>\n",
       "      <td>N</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#FF9E4A</td>\n",
       "      <td>SUB T238N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>246</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>P</td>\n",
       "      <td>Q</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#AD8BC9</td>\n",
       "      <td>SUB P246Q</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>279</td>\n",
       "      <td>SUB</td>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>unpublished</td>\n",
       "      <td>R</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>#DD8452</td>\n",
       "      <td>#AD8BC9</td>\n",
       "      <td>SUB R279C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>71 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    AA pos mutation type             mutation effect  \\\n",
       "0        8           SUB  nonsynonymous substitution   \n",
       "1       13           SUB  nonsynonymous substitution   \n",
       "2       29           SUB  nonsynonymous substitution   \n",
       "3       29           SUB  nonsynonymous substitution   \n",
       "4       31           SUB  nonsynonymous substitution   \n",
       "..     ...           ...                         ...   \n",
       "66     234           SUB  nonsynonymous substitution   \n",
       "67     236           SUB  nonsynonymous substitution   \n",
       "68     238           SUB  nonsynonymous substitution   \n",
       "69     246           SUB  nonsynonymous substitution   \n",
       "70     279           SUB  nonsynonymous substitution   \n",
       "\n",
       "               mutation source AA ref seq AA seq change  mutation count  \\\n",
       "0           recoded stop codon          Y             C               1   \n",
       "1           recoded stop codon          A             V               1   \n",
       "2           recoded stop codon          Q             R               1   \n",
       "3                  unpublished          Q             R               1   \n",
       "4           recoded stop codon          T             M               1   \n",
       "..                         ...        ...           ...             ...   \n",
       "66                 unpublished          G             V               1   \n",
       "67  Pyrobaculum aerophilum tpi          G             E               1   \n",
       "68            Homo sapiens tpi          T             N               1   \n",
       "69                 unpublished          P             Q               1   \n",
       "70                 unpublished          R             C               1   \n",
       "\n",
       "   mutation effect color selection pressure color       name  \n",
       "0                #DD8452                  #A8786E    SUB Y8C  \n",
       "1                #DD8452                  #A8786E   SUB A13V  \n",
       "2                #DD8452                  #A8786E   SUB Q29R  \n",
       "3                #DD8452                  #AD8BC9   SUB Q29R  \n",
       "4                #DD8452                  #A8786E   SUB T31M  \n",
       "..                   ...                      ...        ...  \n",
       "66               #DD8452                  #AD8BC9  SUB G234V  \n",
       "67               #DD8452                  #ED665D  SUB G236E  \n",
       "68               #DD8452                  #FF9E4A  SUB T238N  \n",
       "69               #DD8452                  #AD8BC9  SUB P246Q  \n",
       "70               #DD8452                  #AD8BC9  SUB R279C  \n",
       "\n",
       "[71 rows x 10 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def _get_variant_name(variant):\n",
    "    var_name = ''\n",
    "    if variant['mutation type'] == 'SUB':\n",
    "        var_name = variant['mutation type'] + \" \" + variant['AA ref seq'] + str(variant['AA pos']) + variant['AA seq change']\n",
    "    if variant['mutation type'] == 'DEL':\n",
    "        var_name = variant['mutation type'] + \" \" + str(variant['AA pos'])\n",
    "    return var_name\n",
    "\n",
    "\n",
    "aa_var_cnt_df['name'] = aa_var_cnt_df.apply(lambda r: _get_variant_name(r), axis=1)\n",
    "aa_var_cnt_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "aa_var_cnt_df.to_csv('oxyR_ale_muts.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>fill</th>\n",
       "      <th>edge</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>recoded stop codon</td>\n",
       "      <td>#A8786E</td>\n",
       "      <td>white</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>unpublished</td>\n",
       "      <td>#AD8BC9</td>\n",
       "      <td>white</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Brucella melitensis pgi</td>\n",
       "      <td>#797979</td>\n",
       "      <td>white</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>paraquat</td>\n",
       "      <td>#6DCCDA</td>\n",
       "      <td>white</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Pyrobaculum aerophilum tpi</td>\n",
       "      <td>#ED665D</td>\n",
       "      <td>white</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>FeSO4</td>\n",
       "      <td>#729ECE</td>\n",
       "      <td>white</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Pyrobaculum aerophilum pgi</td>\n",
       "      <td>#67BF5C</td>\n",
       "      <td>white</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Homo sapiens tpi</td>\n",
       "      <td>#FF9E4A</td>\n",
       "      <td>white</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Vibrio cholerae tpi</td>\n",
       "      <td>#D5BB67</td>\n",
       "      <td>white</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>nonsynonymous substitution</td>\n",
       "      <td>white</td>\n",
       "      <td>#DD8452</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>truncation</td>\n",
       "      <td>white</td>\n",
       "      <td>#C44E52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>deletion</td>\n",
       "      <td>white</td>\n",
       "      <td>#937860</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          name     fill     edge\n",
       "0           recoded stop codon  #A8786E    white\n",
       "1                  unpublished  #AD8BC9    white\n",
       "2      Brucella melitensis pgi  #797979    white\n",
       "3                     paraquat  #6DCCDA    white\n",
       "4   Pyrobaculum aerophilum tpi  #ED665D    white\n",
       "5                        FeSO4  #729ECE    white\n",
       "6   Pyrobaculum aerophilum pgi  #67BF5C    white\n",
       "7             Homo sapiens tpi  #FF9E4A    white\n",
       "8          Vibrio cholerae tpi  #D5BB67    white\n",
       "9   nonsynonymous substitution    white  #DD8452\n",
       "10                  truncation    white  #C44E52\n",
       "11                    deletion    white  #937860"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "mut_needle_plot_legend_df = pd.DataFrame()\n",
    "mut_needle_plot_legend_df = aa_var_cnt_df.copy()\n",
    "mut_needle_plot_legend_df = mut_needle_plot_legend_df[['mutation source', 'selection pressure color', ]]\n",
    "mut_needle_plot_legend_df.rename(columns={'mutation source': 'name', 'selection pressure color': 'fill'}, inplace=True)\n",
    "mut_needle_plot_legend_df['edge'] = 'white'\n",
    "mut_needle_plot_legend_df.drop_duplicates(inplace=True)\n",
    "\n",
    "df = aa_var_cnt_df.copy()\n",
    "df = df[['mutation effect', 'mutation effect color']]\n",
    "df.rename(columns={'mutation effect': 'name', 'mutation effect color': 'edge'}, inplace=True)\n",
    "df.drop_duplicates(inplace=True)\n",
    "df['fill'] = 'white'\n",
    "mut_needle_plot_legend_df = mut_needle_plot_legend_df.append(df, ignore_index=True)\n",
    "display(mut_needle_plot_legend_df)\n",
    "\n",
    "mut_needle_plot_legend_df.to_csv('oxyR_ale_muts_legend.csv', index=False)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Creating functional feature annotation output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Chain', 'Substrate binding domain', 'Disulfide bond region',\n",
       "       'OxyR subunit interface', 'Modified residue', 'DNA binding domain',\n",
       "       'HTH domain'], dtype=object)"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "feats_df = pd.read_pickle('./annot_df.pkl')\n",
    "\n",
    "def _rename_feat(f):\n",
    "    new_name = f\n",
    "    if f == 'The C-terminal substrate-binding domain of the LysR-type transcriptional regulator OxyR, a member of the type 2 periplasmic binding fold protein superfamily':\n",
    "        new_name = 'C-terminal substrate-binding domain'\n",
    "    if f == 'Winged helix DNA-binding domain superfamily':\n",
    "        new_name = 'DNA binding domain'\n",
    "    if f == 'LysR substrate binding domain':\n",
    "        new_name = 'Substrate binding domain'\n",
    "    return new_name\n",
    "\n",
    "feats_df['feature'] = feats_df['feature'].apply(\n",
    "    lambda f: _rename_feat(f)\n",
    ")\n",
    "# filtering out features that aren't useful\n",
    "feats_to_remove = [\n",
    "    'Initiator methionine',\n",
    "    # 'Chain',\n",
    "    'TRANSCRIPTIONAL DUAL REGULATOR HCAR-RELATED',\n",
    "    'HYDROGEN PEROXIDE-INDUCIBLE GENES ACTIVATOR',\n",
    "    'LysR, substrate-binding',\n",
    "    '\"Winged helix\" DNA-binding domain',\n",
    "    'Winged helix-like DNA-binding domain superfamily/Winged helix DNA-binding domain',\n",
    "    'Winged helix-like DNA-binding domain superfamily',\n",
    "    'Transcription regulator HTH, LysR',\n",
    "    'Bacterial regulatory helix-turn-helix protein, lysR family',\n",
    "    'LysR-type HTH domain profile',\n",
    "    'C-terminal substrate-binding domain',\n",
    "    'Periplasmic binding protein-like II',\n",
    "    'HTHLYSR',\n",
    "    'DNA binding',\n",
    "    # 'HTH domain',\n",
    "]\n",
    "feats_df = feats_df[~(feats_df['feature'].isin(feats_to_remove))]\n",
    "feats_df.reset_index(inplace=True, drop=True)\n",
    "feats_df.feature.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>range</th>\n",
       "      <th>layer</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chain</td>\n",
       "      <td>1</td>\n",
       "      <td>305</td>\n",
       "      <td>(1, 305)</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Substrate binding domain</td>\n",
       "      <td>89</td>\n",
       "      <td>293</td>\n",
       "      <td>(89, 205)</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Disulfide bond region</td>\n",
       "      <td>180</td>\n",
       "      <td>259</td>\n",
       "      <td>(180, 80)</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>OxyR subunit interface</td>\n",
       "      <td>252</td>\n",
       "      <td>253</td>\n",
       "      <td>(252, 2)</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>OxyR subunit interface</td>\n",
       "      <td>232</td>\n",
       "      <td>235</td>\n",
       "      <td>(232, 4)</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    feature start  end      range  layer\n",
       "0                     Chain     1  305   (1, 305)      0\n",
       "1  Substrate binding domain    89  293  (89, 205)      1\n",
       "2     Disulfide bond region   180  259  (180, 80)      2\n",
       "3    OxyR subunit interface   252  253   (252, 2)      3\n",
       "4    OxyR subunit interface   232  235   (232, 4)      3"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "layer = 0\n",
    "layer_d = dict()\n",
    "for f in feats_df.feature.unique():\n",
    "    layer_d[f] = layer\n",
    "    layer += 1\n",
    "feats_df['layer'] = feats_df['feature'].apply(lambda f: layer_d[f])\n",
    "feats_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>layer</th>\n",
       "      <th>color</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chain</td>\n",
       "      <td>1</td>\n",
       "      <td>305</td>\n",
       "      <td>0</td>\n",
       "      <td>#4878D0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Substrate binding domain</td>\n",
       "      <td>89</td>\n",
       "      <td>293</td>\n",
       "      <td>1</td>\n",
       "      <td>#6ACC64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Disulfide bond region</td>\n",
       "      <td>180</td>\n",
       "      <td>259</td>\n",
       "      <td>2</td>\n",
       "      <td>#D65F5F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>OxyR subunit interface</td>\n",
       "      <td>252</td>\n",
       "      <td>253</td>\n",
       "      <td>3</td>\n",
       "      <td>#956CB4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>OxyR subunit interface</td>\n",
       "      <td>232</td>\n",
       "      <td>235</td>\n",
       "      <td>3</td>\n",
       "      <td>#956CB4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>OxyR subunit interface</td>\n",
       "      <td>228</td>\n",
       "      <td>230</td>\n",
       "      <td>3</td>\n",
       "      <td>#956CB4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>OxyR subunit interface</td>\n",
       "      <td>225</td>\n",
       "      <td>226</td>\n",
       "      <td>3</td>\n",
       "      <td>#956CB4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>OxyR subunit interface</td>\n",
       "      <td>221</td>\n",
       "      <td>223</td>\n",
       "      <td>3</td>\n",
       "      <td>#956CB4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>OxyR subunit interface</td>\n",
       "      <td>219</td>\n",
       "      <td>219</td>\n",
       "      <td>3</td>\n",
       "      <td>#956CB4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Disulfide bond region</td>\n",
       "      <td>199</td>\n",
       "      <td>208</td>\n",
       "      <td>2</td>\n",
       "      <td>#D65F5F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Modified residue</td>\n",
       "      <td>199</td>\n",
       "      <td>199</td>\n",
       "      <td>4</td>\n",
       "      <td>#8C613C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>OxyR subunit interface</td>\n",
       "      <td>172</td>\n",
       "      <td>172</td>\n",
       "      <td>3</td>\n",
       "      <td>#956CB4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>OxyR subunit interface</td>\n",
       "      <td>170</td>\n",
       "      <td>170</td>\n",
       "      <td>3</td>\n",
       "      <td>#956CB4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>OxyR subunit interface</td>\n",
       "      <td>120</td>\n",
       "      <td>126</td>\n",
       "      <td>3</td>\n",
       "      <td>#956CB4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>OxyR subunit interface</td>\n",
       "      <td>114</td>\n",
       "      <td>114</td>\n",
       "      <td>3</td>\n",
       "      <td>#956CB4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>OxyR subunit interface</td>\n",
       "      <td>110</td>\n",
       "      <td>111</td>\n",
       "      <td>3</td>\n",
       "      <td>#956CB4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>OxyR subunit interface</td>\n",
       "      <td>106</td>\n",
       "      <td>107</td>\n",
       "      <td>3</td>\n",
       "      <td>#956CB4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>OxyR subunit interface</td>\n",
       "      <td>103</td>\n",
       "      <td>103</td>\n",
       "      <td>3</td>\n",
       "      <td>#956CB4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>OxyR subunit interface</td>\n",
       "      <td>94</td>\n",
       "      <td>94</td>\n",
       "      <td>3</td>\n",
       "      <td>#956CB4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>DNA binding domain</td>\n",
       "      <td>1</td>\n",
       "      <td>86</td>\n",
       "      <td>5</td>\n",
       "      <td>#DC7EC0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>HTH domain</td>\n",
       "      <td>1</td>\n",
       "      <td>58</td>\n",
       "      <td>6</td>\n",
       "      <td>#797979</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>OxyR subunit interface</td>\n",
       "      <td>110</td>\n",
       "      <td>110</td>\n",
       "      <td>3</td>\n",
       "      <td>#956CB4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>OxyR subunit interface</td>\n",
       "      <td>122</td>\n",
       "      <td>126</td>\n",
       "      <td>3</td>\n",
       "      <td>#956CB4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>OxyR subunit interface</td>\n",
       "      <td>222</td>\n",
       "      <td>223</td>\n",
       "      <td>3</td>\n",
       "      <td>#956CB4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>OxyR subunit interface</td>\n",
       "      <td>226</td>\n",
       "      <td>226</td>\n",
       "      <td>3</td>\n",
       "      <td>#956CB4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>OxyR subunit interface</td>\n",
       "      <td>229</td>\n",
       "      <td>230</td>\n",
       "      <td>3</td>\n",
       "      <td>#956CB4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>OxyR subunit interface</td>\n",
       "      <td>232</td>\n",
       "      <td>233</td>\n",
       "      <td>3</td>\n",
       "      <td>#956CB4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     feature start  end  layer    color\n",
       "0                      Chain     1  305      0  #4878D0\n",
       "1   Substrate binding domain    89  293      1  #6ACC64\n",
       "2      Disulfide bond region   180  259      2  #D65F5F\n",
       "3     OxyR subunit interface   252  253      3  #956CB4\n",
       "4     OxyR subunit interface   232  235      3  #956CB4\n",
       "5     OxyR subunit interface   228  230      3  #956CB4\n",
       "6     OxyR subunit interface   225  226      3  #956CB4\n",
       "7     OxyR subunit interface   221  223      3  #956CB4\n",
       "8     OxyR subunit interface   219  219      3  #956CB4\n",
       "9      Disulfide bond region   199  208      2  #D65F5F\n",
       "10          Modified residue   199  199      4  #8C613C\n",
       "13    OxyR subunit interface   172  172      3  #956CB4\n",
       "14    OxyR subunit interface   170  170      3  #956CB4\n",
       "15    OxyR subunit interface   120  126      3  #956CB4\n",
       "16    OxyR subunit interface   114  114      3  #956CB4\n",
       "17    OxyR subunit interface   110  111      3  #956CB4\n",
       "18    OxyR subunit interface   106  107      3  #956CB4\n",
       "19    OxyR subunit interface   103  103      3  #956CB4\n",
       "20    OxyR subunit interface    94   94      3  #956CB4\n",
       "21        DNA binding domain     1   86      5  #DC7EC0\n",
       "22                HTH domain     1   58      6  #797979\n",
       "25    OxyR subunit interface   110  110      3  #956CB4\n",
       "27    OxyR subunit interface   122  126      3  #956CB4\n",
       "29    OxyR subunit interface   222  223      3  #956CB4\n",
       "30    OxyR subunit interface   226  226      3  #956CB4\n",
       "31    OxyR subunit interface   229  230      3  #956CB4\n",
       "32    OxyR subunit interface   232  233      3  #956CB4"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "{'Chain': '#4878D0',\n",
       " 'Substrate binding domain': '#6ACC64',\n",
       " 'Disulfide bond region': '#D65F5F',\n",
       " 'OxyR subunit interface': '#956CB4',\n",
       " 'Modified residue': '#8C613C',\n",
       " 'DNA binding domain': '#DC7EC0',\n",
       " 'HTH domain': '#797979'}"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "feat_color_d = dict()\n",
    "muted=[\"#4878D0\", \"#6ACC64\", \"#D65F5F\", \"#956CB4\", \"#8C613C\", \"#DC7EC0\", \"#797979\", \"#D5BB67\", \"#82C6E2\", \"#EE854A\"]  # from seaborn\n",
    "feats = list(feats_df.feature.unique())\n",
    "for i in range(0, len(feats_df.feature.unique())):\n",
    "    feat_color_d[feats[i]] = muted[i]\n",
    "\n",
    "feats_df['color'] = feats_df['feature'].apply(lambda f: feat_color_d[f])\n",
    "\n",
    "feats_df.drop(columns=[\"range\"], inplace=True)\n",
    "feats_df.drop_duplicates(inplace=True)\n",
    "\n",
    "display(feats_df, feat_color_d)\n",
    "feats_df.to_csv('./oxyR_feats.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>color</th>\n",
       "      <th>size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chain</td>\n",
       "      <td>1</td>\n",
       "      <td>305</td>\n",
       "      <td>#4878D0</td>\n",
       "      <td>305</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Substrate binding domain</td>\n",
       "      <td>89</td>\n",
       "      <td>293</td>\n",
       "      <td>#6ACC64</td>\n",
       "      <td>205</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DNA binding domain</td>\n",
       "      <td>1</td>\n",
       "      <td>86</td>\n",
       "      <td>#DC7EC0</td>\n",
       "      <td>86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Disulfide bond region</td>\n",
       "      <td>180</td>\n",
       "      <td>259</td>\n",
       "      <td>#D65F5F</td>\n",
       "      <td>80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HTH domain</td>\n",
       "      <td>1</td>\n",
       "      <td>58</td>\n",
       "      <td>#797979</td>\n",
       "      <td>58</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Disulfide bond region</td>\n",
       "      <td>199</td>\n",
       "      <td>208</td>\n",
       "      <td>#D65F5F</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>OxyR subunit interface</td>\n",
       "      <td>120</td>\n",
       "      <td>126</td>\n",
       "      <td>#956CB4</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>OxyR subunit interface</td>\n",
       "      <td>122</td>\n",
       "      <td>126</td>\n",
       "      <td>#956CB4</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>OxyR subunit interface</td>\n",
       "      <td>232</td>\n",
       "      <td>235</td>\n",
       "      <td>#956CB4</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>OxyR subunit interface</td>\n",
       "      <td>221</td>\n",
       "      <td>223</td>\n",
       "      <td>#956CB4</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>OxyR subunit interface</td>\n",
       "      <td>228</td>\n",
       "      <td>230</td>\n",
       "      <td>#956CB4</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>OxyR subunit interface</td>\n",
       "      <td>106</td>\n",
       "      <td>107</td>\n",
       "      <td>#956CB4</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>OxyR subunit interface</td>\n",
       "      <td>229</td>\n",
       "      <td>230</td>\n",
       "      <td>#956CB4</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>OxyR subunit interface</td>\n",
       "      <td>222</td>\n",
       "      <td>223</td>\n",
       "      <td>#956CB4</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>OxyR subunit interface</td>\n",
       "      <td>252</td>\n",
       "      <td>253</td>\n",
       "      <td>#956CB4</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>OxyR subunit interface</td>\n",
       "      <td>232</td>\n",
       "      <td>233</td>\n",
       "      <td>#956CB4</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>OxyR subunit interface</td>\n",
       "      <td>110</td>\n",
       "      <td>111</td>\n",
       "      <td>#956CB4</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>OxyR subunit interface</td>\n",
       "      <td>225</td>\n",
       "      <td>226</td>\n",
       "      <td>#956CB4</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>OxyR subunit interface</td>\n",
       "      <td>219</td>\n",
       "      <td>219</td>\n",
       "      <td>#956CB4</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>OxyR subunit interface</td>\n",
       "      <td>94</td>\n",
       "      <td>94</td>\n",
       "      <td>#956CB4</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>OxyR subunit interface</td>\n",
       "      <td>114</td>\n",
       "      <td>114</td>\n",
       "      <td>#956CB4</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>OxyR subunit interface</td>\n",
       "      <td>170</td>\n",
       "      <td>170</td>\n",
       "      <td>#956CB4</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>OxyR subunit interface</td>\n",
       "      <td>110</td>\n",
       "      <td>110</td>\n",
       "      <td>#956CB4</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>OxyR subunit interface</td>\n",
       "      <td>172</td>\n",
       "      <td>172</td>\n",
       "      <td>#956CB4</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Modified residue</td>\n",
       "      <td>199</td>\n",
       "      <td>199</td>\n",
       "      <td>#8C613C</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>OxyR subunit interface</td>\n",
       "      <td>226</td>\n",
       "      <td>226</td>\n",
       "      <td>#956CB4</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>OxyR subunit interface</td>\n",
       "      <td>103</td>\n",
       "      <td>103</td>\n",
       "      <td>#956CB4</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     feature start  end    color  size\n",
       "0                      Chain     1  305  #4878D0   305\n",
       "1   Substrate binding domain    89  293  #6ACC64   205\n",
       "2         DNA binding domain     1   86  #DC7EC0    86\n",
       "3      Disulfide bond region   180  259  #D65F5F    80\n",
       "4                 HTH domain     1   58  #797979    58\n",
       "5      Disulfide bond region   199  208  #D65F5F    10\n",
       "6     OxyR subunit interface   120  126  #956CB4     7\n",
       "7     OxyR subunit interface   122  126  #956CB4     5\n",
       "8     OxyR subunit interface   232  235  #956CB4     4\n",
       "9     OxyR subunit interface   221  223  #956CB4     3\n",
       "10    OxyR subunit interface   228  230  #956CB4     3\n",
       "11    OxyR subunit interface   106  107  #956CB4     2\n",
       "12    OxyR subunit interface   229  230  #956CB4     2\n",
       "13    OxyR subunit interface   222  223  #956CB4     2\n",
       "14    OxyR subunit interface   252  253  #956CB4     2\n",
       "15    OxyR subunit interface   232  233  #956CB4     2\n",
       "16    OxyR subunit interface   110  111  #956CB4     2\n",
       "17    OxyR subunit interface   225  226  #956CB4     2\n",
       "18    OxyR subunit interface   219  219  #956CB4     1\n",
       "19    OxyR subunit interface    94   94  #956CB4     1\n",
       "20    OxyR subunit interface   114  114  #956CB4     1\n",
       "21    OxyR subunit interface   170  170  #956CB4     1\n",
       "22    OxyR subunit interface   110  110  #956CB4     1\n",
       "23    OxyR subunit interface   172  172  #956CB4     1\n",
       "24          Modified residue   199  199  #8C613C     1\n",
       "25    OxyR subunit interface   226  226  #956CB4     1\n",
       "26    OxyR subunit interface   103  103  #956CB4     1"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# for NGL\n",
    "ngl_feat_df = feats_df.copy()\n",
    "ngl_feat_df[\"size\"] = ngl_feat_df.apply(lambda r: r[\"end\"] - r[\"start\"] + 1, axis=1)\n",
    "ngl_feat_df = ngl_feat_df[[\"feature\", \"start\", \"end\", \"color\", \"size\"]]\n",
    "ngl_feat_df = ngl_feat_df.sort_values(\"size\", ascending=False)\n",
    "ngl_feat_df = ngl_feat_df.reset_index(drop=True)  # json explicitly uses the indexes\n",
    "\n",
    "with open('feats.json', 'w') as outfile:   # for NGL\n",
    "    outfile.write(ngl_feat_df.to_json())\n",
    "ngl_feat_df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "7ebccac1095285a8f3a7365f16cbaa2bb02a7e2674096928d95b1b0e9894f90a"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
